{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\" style=\"margin-top: 1em;\"><ul class=\"toc-item\"><li><ul class=\"toc-item\"><li><span><a href=\"#Task-1.-Compiling-Ebola-Data\" data-toc-modified-id=\"Task-1.-Compiling-Ebola-Data-0.1\"><span class=\"toc-item-num\">0.1&nbsp;&nbsp;</span>Task 1. Compiling Ebola Data</a></span><ul class=\"toc-item\"><li><span><a href=\"#1---Cleaning-Guinea\" data-toc-modified-id=\"1---Cleaning-Guinea-0.1.1\"><span class=\"toc-item-num\">0.1.1&nbsp;&nbsp;</span>1 - Cleaning Guinea</a></span><ul class=\"toc-item\"><li><span><a href=\"#model-0\" data-toc-modified-id=\"model-0-0.1.1.1\"><span class=\"toc-item-num\">0.1.1.1&nbsp;&nbsp;</span>model 0</a></span></li><li><span><a href=\"#model-1\" data-toc-modified-id=\"model-1-0.1.1.2\"><span class=\"toc-item-num\">0.1.1.2&nbsp;&nbsp;</span>model 1</a></span></li></ul></li><li><span><a href=\"#2---Cleaning-Liberia\" data-toc-modified-id=\"2---Cleaning-Liberia-0.1.2\"><span class=\"toc-item-num\">0.1.2&nbsp;&nbsp;</span>2 - Cleaning Liberia</a></span><ul class=\"toc-item\"><li><span><a href=\"#model-0\" data-toc-modified-id=\"model-0-0.1.2.1\"><span class=\"toc-item-num\">0.1.2.1&nbsp;&nbsp;</span>model 0</a></span></li><li><span><a href=\"#model-1\" data-toc-modified-id=\"model-1-0.1.2.2\"><span class=\"toc-item-num\">0.1.2.2&nbsp;&nbsp;</span>model 1</a></span></li><li><span><a href=\"#model-2---5\" data-toc-modified-id=\"model-2---5-0.1.2.3\"><span class=\"toc-item-num\">0.1.2.3&nbsp;&nbsp;</span>model 2 - 5</a></span></li><li><span><a href=\"#Finally-we-can-concat-all-the-different-model-dataframe\" data-toc-modified-id=\"Finally-we-can-concat-all-the-different-model-dataframe-0.1.2.4\"><span class=\"toc-item-num\">0.1.2.4&nbsp;&nbsp;</span>Finally we can concat all the different model dataframe</a></span></li></ul></li><li><span><a href=\"#3---Cleaning-Sierra-Leone\" data-toc-modified-id=\"3---Cleaning-Sierra-Leone-0.1.3\"><span class=\"toc-item-num\">0.1.3&nbsp;&nbsp;</span>3 - Cleaning Sierra Leone</a></span><ul class=\"toc-item\"><li><span><a href=\"#model-0\" data-toc-modified-id=\"model-0-0.1.3.1\"><span class=\"toc-item-num\">0.1.3.1&nbsp;&nbsp;</span>model 0</a></span></li><li><span><a href=\"#model-1\" data-toc-modified-id=\"model-1-0.1.3.2\"><span class=\"toc-item-num\">0.1.3.2&nbsp;&nbsp;</span>model 1</a></span></li></ul></li><li><span><a href=\"#4---Result\" data-toc-modified-id=\"4---Result-0.1.4\"><span class=\"toc-item-num\">0.1.4&nbsp;&nbsp;</span>4 - Result</a></span></li></ul></li><li><span><a href=\"#Task-2.-RNA-Sequences\" data-toc-modified-id=\"Task-2.-RNA-Sequences-0.2\"><span class=\"toc-item-num\">0.2&nbsp;&nbsp;</span>Task 2. RNA Sequences</a></span><ul class=\"toc-item\"><li><span><a href=\"#1--Importing-the-data\" data-toc-modified-id=\"1--Importing-the-data-0.2.1\"><span class=\"toc-item-num\">0.2.1&nbsp;&nbsp;</span>1- Importing the data</a></span></li><li><span><a href=\"#2---Merging-and-keeping-a-unique-index\" data-toc-modified-id=\"2---Merging-and-keeping-a-unique-index-0.2.2\"><span class=\"toc-item-num\">0.2.2&nbsp;&nbsp;</span>2 - Merging and keeping a unique index</a></span></li></ul></li><li><span><a href=\"#Task-3.-Class-War-in-Titanic\" data-toc-modified-id=\"Task-3.-Class-War-in-Titanic-0.3\"><span class=\"toc-item-num\">0.3&nbsp;&nbsp;</span>Task 3. Class War in Titanic</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import glob\n",
    "from datetime import date\n",
    "from datetime import timedelta  \n",
    "sns.set_context('notebook')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pd.options.mode.chained_assignment = None  # default='warn', Mutes warnings when copying a slice from a DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = './Data' \n",
    "EBOLA_FOLDER = DATA_FOLDER + '/ebola'\n",
    "\n",
    "GUINEA_FOLDER = EBOLA_FOLDER + '/guinea_data'\n",
    "LIBERIA_FOLDER = EBOLA_FOLDER + '/liberia_data'\n",
    "SIERRA_LEONE_FOLDER = EBOLA_FOLDER + '/sl_data'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "<p>\n",
    "    <div class=\"lev1\">\n",
    "        <a href=\"#1---Cleaning-Guinea\">1 - Cleaning Guinea</a>\n",
    "    </div>\n",
    "    <div class=\"lev2\">\n",
    "        <a href=\"#2---Cleaning-Liberia\">2 - Cleaning Liberia</a>\n",
    "    </div>\n",
    "    <div class=\"lev3\">\n",
    "        <a href=\"#3---Cleaning-Sierra-Leone\">3 - Cleaning Sierra Leone</a>\n",
    "    </div>\n",
    "    <div class=\"lev3\">\n",
    "        <a href=\"#4---Result\">4 - Result</a>\n",
    "    </div>\n",
    " </p>\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Nota Bene : because we didn't know what exactly the exercise asked for when asking for new deaths and cases, we decided to be consistent along the homework and to always focus on confirmed values (as opposed to the sum of ```confirmed + supected + probable```)**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1 - Cleaning Guinea"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We want to identify the models. We will call model the different kind of ```Description``` values that are present in files. To be able to clean them as a whole. We will drop the columns of each region as we are only interested in national numbers. \n",
    "\n",
    "We use ```glob``` in order to get all UNIX style pathname pattern expansion to be able to import all files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# The following functions were written after a first analysis on panda, and therefore we knew what we needed\n",
    "def find_model(distinct_model_variables,current_model):\n",
    "    \"\"\" Utility function to find current_model in distinct_model_variables. \"\"\"\n",
    "    for i,d in enumerate(distinct_model_variables):\n",
    "        if d == current_model:\n",
    "            return i\n",
    "    return -1\n",
    "\n",
    "def print_distinct_model_repartition(n_distinct_models, belongs_to):\n",
    "    \"\"\" Pretty printing the number of files that share the same models for each model \"\"\"\n",
    "    for model_num in range(0,n_distinct_models):\n",
    "        n_model = len([x for x in belongs_to if x == model_num])\n",
    "        print(\"{} file(s) in model {}\".format(n_model,model_num))\n",
    "\n",
    "def get_distinct_model(path_to_folder,desc_col_name,date_col_name,total_col_name):\n",
    "    \"\"\" \n",
    "    Will look for each file and try to cluster them into 'models', \n",
    "    that is the file that have the same Description/variable values.\n",
    "    \"\"\"\n",
    "    country_df = pd.DataFrame()\n",
    "    list_ = []\n",
    "    distinct_model_variables = []\n",
    "    belongs_to_model = []\n",
    "    \n",
    "    allFiles = sorted(glob.glob(path_to_folder + \"/*.csv\"))\n",
    "    cols_to_use = [desc_col_name,date_col_name,total_col_name]\n",
    "    \n",
    "\n",
    "    for file_ in allFiles:\n",
    "        df = pd.read_csv(file_,index_col=None, header=0,usecols=cols_to_use,parse_dates=[date_col_name],thousands=',')\n",
    "        list_.append(df)\n",
    "    \n",
    "    for df in list_:\n",
    "        current_model = set(df[desc_col_name].values)\n",
    "        index = find_model(distinct_model_variables,current_model)\n",
    "        if(index == -1):\n",
    "            distinct_model_variables.append(current_model)\n",
    "            belongs_to_model.append(len(distinct_model_variables)-1)\n",
    "        else:\n",
    "            belongs_to_model.append(index)\n",
    "    \n",
    "    n_distinct_models = len(distinct_model_variables)\n",
    "    print('We have found {} different model(s).'.format(n_distinct_models))\n",
    "    print_distinct_model_repartition(n_distinct_models,belongs_to_model)\n",
    "    \n",
    "    return (list_,distinct_model_variables,belongs_to_model)\n",
    "\n",
    "def get_grouped_dataframes(path_to_folder,desc_col_name,date_col_name,total_col_name):\n",
    "    \"\"\" \n",
    "    Returns grouped dataframes, each dataframe contains \n",
    "    the data of the files sharing the same model.\n",
    "    \"\"\"\n",
    "    (dfs,distinct_model_variables,belongs_to_model) = get_distinct_model(path_to_folder,desc_col_name,date_col_name,total_col_name)\n",
    "    \n",
    "    n_distinct_models = len(distinct_model_variables)\n",
    "    model_dataframes = []\n",
    "    for i in range(0,n_distinct_models):\n",
    "        list_ = []\n",
    "        for j,m in enumerate(belongs_to_model):\n",
    "            if(m == i):\n",
    "                list_.append(dfs[j])\n",
    "        \n",
    "        # We also parse the date and get rid of the 'Date' column\n",
    "        df = pd.concat(list_)\n",
    "        df['Month'] = df[date_col_name].map(lambda x:x.month)\n",
    "        df['Day'] = df[date_col_name].map(lambda x:x.day)\n",
    "        df = df.drop(date_col_name,axis=1)\n",
    "        \n",
    "        df.set_index(['Month','Day',desc_col_name],inplace=True)\n",
    "        \n",
    "        if df.index.is_unique:\n",
    "            df = df[[total_col_name]].unstack(desc_col_name)\n",
    "            \n",
    "            # To avoid hierarchical columns, we drop the first level\n",
    "            df.columns = df.columns.droplevel(0)\n",
    "            df.columns.name = ''\n",
    "\n",
    "        model_dataframes.append(df)\n",
    "    return model_dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have found 2 different model(s).\n",
      "1 file(s) in model 0\n",
      "21 file(s) in model 1\n"
     ]
    }
   ],
   "source": [
    "path = GUINEA_FOLDER\n",
    "model_dataframes = get_grouped_dataframes(GUINEA_FOLDER,'Description','Date','Totals')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### model 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_0 = model_dataframes[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have found 2 different 'models' of description values. Therefore we select the columns of interest for each of them. To do so we list the columns and look at those that talk about cases and deaths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# We check for deaths\n",
    "cols = df_model_0.columns.values.tolist()\n",
    "rel_cols = [c for c in cols if 'deaths' in c]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New deaths registered today</th>\n",
       "      <th>New deaths registered today (confirmed)</th>\n",
       "      <th>New deaths registered today (probables)</th>\n",
       "      <th>New deaths registered today (suspects)</th>\n",
       "      <th>Number of deaths of confirmed cases among health workers</th>\n",
       "      <th>Number of deaths of probables cases among health workers</th>\n",
       "      <th>Total deaths (confirmed + probables + suspects)</th>\n",
       "      <th>Total deaths of confirmed</th>\n",
       "      <th>Total deaths of probables</th>\n",
       "      <th>Total deaths of suspects</th>\n",
       "      <th>Total of deaths in confirmed cases in CTE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>363</td>\n",
       "      <td>228</td>\n",
       "      <td>133</td>\n",
       "      <td>2</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           New deaths registered today  \\\n",
       "Month Day                                \n",
       "8     4                              2   \n",
       "\n",
       "           New deaths registered today (confirmed)  \\\n",
       "Month Day                                            \n",
       "8     4                                          2   \n",
       "\n",
       "           New deaths registered today (probables)  \\\n",
       "Month Day                                            \n",
       "8     4                                          0   \n",
       "\n",
       "           New deaths registered today (suspects)  \\\n",
       "Month Day                                           \n",
       "8     4                                         0   \n",
       "\n",
       "           Number of deaths of confirmed cases among health workers  \\\n",
       "Month Day                                                             \n",
       "8     4                                                   12          \n",
       "\n",
       "           Number of deaths of probables cases among health workers  \\\n",
       "Month Day                                                             \n",
       "8     4                                                    8          \n",
       "\n",
       "           Total deaths (confirmed + probables + suspects)  \\\n",
       "Month Day                                                    \n",
       "8     4                                                363   \n",
       "\n",
       "           Total deaths of confirmed  Total deaths of probables  \\\n",
       "Month Day                                                         \n",
       "8     4                          228                        133   \n",
       "\n",
       "           Total deaths of suspects  Total of deaths in confirmed cases in CTE  \n",
       "Month Day                                                                       \n",
       "8     4                           2                                        138  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_0.loc[:,rel_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will only consider the column containing new, as the ```Total``` and ```Number``` columns seem to be cumulative. To be consistent with our **nota bene** we decide to take only the confimed deaths in ```New deaths registered today (confirmed)```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_model_0['New Deaths'] = df_model_0[['New deaths registered today (confirmed)']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New cases of confirmed</th>\n",
       "      <th>New cases of probables</th>\n",
       "      <th>New cases of suspects</th>\n",
       "      <th>Total new cases registered so far</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           New cases of confirmed  New cases of probables  \\\n",
       "Month Day                                                   \n",
       "8     4                         4                       0   \n",
       "\n",
       "           New cases of suspects  Total new cases registered so far  \n",
       "Month Day                                                            \n",
       "8     4                        5                                  9  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We check for cases\n",
    "cols = df_model_0.columns.values.tolist()\n",
    "rel_cols = [c for c in cols if ('cases' in c and ('new' in c or 'New' in c))]\n",
    "df_model_0.loc[:,rel_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again we only keep the headers that contain both ``new`` (with or without caps) and ```cases```. Again we only keep the confirmed cases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_0['New Cases'] =  df_model_0['New cases of confirmed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           New Deaths  New Cases\n",
       "Month Day                       \n",
       "8     4             2          4"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_0 = df_model_0[['New Deaths','New Cases']]\n",
    "model_dataframes[0] = df_model_0\n",
    "df_model_0.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### model 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we do the same pipeline for the second model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New deaths registered</th>\n",
       "      <th>New deaths registered among health workers</th>\n",
       "      <th>Total deaths (confirmed + probables + suspects)</th>\n",
       "      <th>Total deaths of confirmed</th>\n",
       "      <th>Total deaths of probables</th>\n",
       "      <th>Total deaths of suspects</th>\n",
       "      <th>Total deaths registered among health workers</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">8</th>\n",
       "      <th>26</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>958</td>\n",
       "      <td>292</td>\n",
       "      <td>141</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>438</td>\n",
       "      <td>294</td>\n",
       "      <td>142</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>468</td>\n",
       "      <td>324</td>\n",
       "      <td>142</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>489</td>\n",
       "      <td>337</td>\n",
       "      <td>150</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>501</td>\n",
       "      <td>349</td>\n",
       "      <td>150</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          New deaths registered New deaths registered among health workers  \\\n",
       "Month Day                                                                    \n",
       "8     26                      5                                          0   \n",
       "      27                      2                                          0   \n",
       "      30                      5                                          0   \n",
       "      31                      3                                          0   \n",
       "9     2                       5                                          0   \n",
       "\n",
       "          Total deaths (confirmed + probables + suspects)  \\\n",
       "Month Day                                                   \n",
       "8     26                                              958   \n",
       "      27                                              438   \n",
       "      30                                              468   \n",
       "      31                                              489   \n",
       "9     2                                               501   \n",
       "\n",
       "          Total deaths of confirmed Total deaths of probables  \\\n",
       "Month Day                                                       \n",
       "8     26                        292                       141   \n",
       "      27                        294                       142   \n",
       "      30                        324                       142   \n",
       "      31                        337                       150   \n",
       "9     2                         349                       150   \n",
       "\n",
       "          Total deaths of suspects  \\\n",
       "Month Day                            \n",
       "8     26                         2   \n",
       "      27                         2   \n",
       "      30                         2   \n",
       "      31                         2   \n",
       "9     2                          2   \n",
       "\n",
       "          Total deaths registered among health workers  \n",
       "Month Day                                               \n",
       "8     26                                            26  \n",
       "      27                                            26  \n",
       "      30                                            26  \n",
       "      31                                            26  \n",
       "9     2                                             26  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_1 = model_dataframes[1]\n",
    "# We check for deaths\n",
    "cols = df_model_1.columns.values.tolist()\n",
    "rel_cols = [c for c in cols if 'death' in c]\n",
    "df_model_1.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again it seems like :\n",
    "* ```New deaths registered among health workers``` is a particular case of the new deaths registerd and is therefore already included in there.\n",
    "* ```Total ...``` seem to be stricly increasing (appart for some weird outliers on the 26 of july for total deaths) and must therefore be cumulative.\n",
    "* ``` New deaths registered ``` is the only column that interests us, **we will assume that it contains the confirmed deaths**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_1['New Deaths'] = df_model_1[['New deaths registered']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New cases of confirmed</th>\n",
       "      <th>New cases of confirmed among health workers</th>\n",
       "      <th>New cases of probables</th>\n",
       "      <th>New cases of suspects</th>\n",
       "      <th>Total cases of confirmed</th>\n",
       "      <th>Total cases of probables</th>\n",
       "      <th>Total cases of suspects</th>\n",
       "      <th>Total new cases registered so far</th>\n",
       "      <th>Total number of hospitalized cases in CTE</th>\n",
       "      <th>Total suspected non-class cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">8</th>\n",
       "      <th>26</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18</td>\n",
       "      <td>490</td>\n",
       "      <td>141</td>\n",
       "      <td>30</td>\n",
       "      <td>28</td>\n",
       "      <td>87</td>\n",
       "      <td>518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>499</td>\n",
       "      <td>142</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>80</td>\n",
       "      <td>518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>533</td>\n",
       "      <td>142</td>\n",
       "      <td>32</td>\n",
       "      <td>24</td>\n",
       "      <td>82</td>\n",
       "      <td>560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>563</td>\n",
       "      <td>150</td>\n",
       "      <td>36</td>\n",
       "      <td>46</td>\n",
       "      <td>99</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>591</td>\n",
       "      <td>150</td>\n",
       "      <td>49</td>\n",
       "      <td>25</td>\n",
       "      <td>105</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          New cases of confirmed New cases of confirmed among health workers  \\\n",
       "Month Day                                                                      \n",
       "8     26                      10                                           0   \n",
       "      27                      10                                           0   \n",
       "      30                       9                                           0   \n",
       "      31                      29                                           0   \n",
       "9     2                       14                                           0   \n",
       "\n",
       "          New cases of probables New cases of suspects  \\\n",
       "Month Day                                                \n",
       "8     26                     NaN                    18   \n",
       "      27                       0                    12   \n",
       "      30                       0                    15   \n",
       "      31                       8                     9   \n",
       "9     2                        0                    11   \n",
       "\n",
       "          Total cases of confirmed Total cases of probables  \\\n",
       "Month Day                                                     \n",
       "8     26                       490                      141   \n",
       "      27                       499                      142   \n",
       "      30                       533                      142   \n",
       "      31                       563                      150   \n",
       "9     2                        591                      150   \n",
       "\n",
       "          Total cases of suspects Total new cases registered so far  \\\n",
       "Month Day                                                             \n",
       "8     26                       30                                28   \n",
       "      27                       22                                22   \n",
       "      30                       32                                24   \n",
       "      31                       36                                46   \n",
       "9     2                        49                                25   \n",
       "\n",
       "          Total number of hospitalized cases in CTE  \\\n",
       "Month Day                                             \n",
       "8     26                                         87   \n",
       "      27                                         80   \n",
       "      30                                         82   \n",
       "      31                                         99   \n",
       "9     2                                         105   \n",
       "\n",
       "          Total suspected non-class cases  \n",
       "Month Day                                  \n",
       "8     26                              518  \n",
       "      27                              518  \n",
       "      30                              560  \n",
       "      31                              642  \n",
       "9     2                               642  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We check for cases\n",
    "cols = df_model_1.columns.values.tolist()\n",
    "rel_cols = [c for c in cols if ('cases' in c )]\n",
    "df_model_1.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we can make the same usual observation, the only column of interest seems to be ```New cases of confirmed``` (to remain consistent with our ** Nota Bene **). The rest are cumulative values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_1['New Cases'] = df_model_1[['New cases of confirmed']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_model_1 = df_model_1[['New Deaths','New Cases']]\n",
    "model_dataframes[1] = df_model_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we concatenate the two models to obtain a dataframe for the whole country. And compute the mean."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_guinea = pd.concat(model_dataframes,axis=0)\n",
    "df_guinea['Country'] = 'Guinea'\n",
    "df_guinea['New Deaths'] = df_guinea['New Deaths'].astype(int)\n",
    "df_guinea['New Cases'] = df_guinea['New Cases'].astype(int)\n",
    "df_guinea = df_guinea.reset_index().groupby(['Country','Month']).mean()[['New Deaths','New Cases']].round(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2 - Cleaning Liberia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have found 6 different model(s).\n",
      "55 file(s) in model 0\n",
      "1 file(s) in model 1\n",
      "11 file(s) in model 2\n",
      "1 file(s) in model 3\n",
      "8 file(s) in model 4\n",
      "24 file(s) in model 5\n"
     ]
    }
   ],
   "source": [
    "path = LIBERIA_FOLDER\n",
    "model_dataframes = get_grouped_dataframes(path,'Variable','Date','National')\n",
    "allFiles = glob.glob(path + \"/*.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here again, we have found 5 different 'models' of description values. Therefore we select the columns of interest for each of them. To do so we list the columns and look at those that talk about cases and deaths."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### model 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Cumulative deaths among HCW', 'Newly Reported deaths in HCW', 'Newly reported deaths']\n"
     ]
    }
   ],
   "source": [
    "df_model_0 = model_dataframes[0]\n",
    "\n",
    "# We check for deaths\n",
    "cols = df_model_0.columns.values.tolist()\n",
    "rel_cols = [c for c in cols if 'deaths' in c]\n",
    "print(rel_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Cumulative deaths among HCW</th>\n",
       "      <th>Newly Reported deaths in HCW</th>\n",
       "      <th>Newly reported deaths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">6</th>\n",
       "      <th>16</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Cumulative deaths among HCW  Newly Reported deaths in HCW  \\\n",
       "Month Day                                                              \n",
       "6     16                           1.0                           0.0   \n",
       "      17                           1.0                           0.0   \n",
       "      22                           1.0                           0.0   \n",
       "      24                           2.0                           0.0   \n",
       "      25                           3.0                           1.0   \n",
       "\n",
       "           Newly reported deaths  \n",
       "Month Day                         \n",
       "6     16                     2.0  \n",
       "      17                     0.0  \n",
       "      22                     4.0  \n",
       "      24                     4.0  \n",
       "      25                     3.0  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_0.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we will only consider the column ```Newly reported deaths``` because we are interested in the total number of death for each day and not the one among HCW."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Case Fatality Rate (CFR) - Confirmed & Probable Cases', 'Contacts lost to follow-up', 'Contacts seen', 'Contacts who completed 21 day follow-up', 'Cumulative admission/isolation', 'Cumulative cases among HCW', 'Cumulative deaths among HCW', 'Currently under follow-up', 'New Case/s (Probable)', 'New Case/s (Suspected)', 'New admissions', 'New case/s (confirmed)', 'Newly Reported Cases in HCW', 'Newly Reported deaths in HCW', 'Newly reported contacts', 'Newly reported deaths', 'Specimens collected', 'Specimens pending for testing', 'Total Number of Confirmed Cases of Guinean Nationality', 'Total Number of Confirmed Cases of Sierra Leonean Nationality', 'Total confirmed cases', 'Total contacts listed', 'Total death/s in confirmed cases', 'Total death/s in confirmed, probable, suspected cases', 'Total death/s in probable cases', 'Total death/s in suspected cases', 'Total discharges', 'Total no. currently in Treatment Units', 'Total probable cases', 'Total specimens tested', 'Total suspected cases', 'New Deaths']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Cumulative cases among HCW</th>\n",
       "      <th>New case/s (confirmed)</th>\n",
       "      <th>Total confirmed cases</th>\n",
       "      <th>Total death/s in confirmed cases</th>\n",
       "      <th>Total death/s in confirmed, probable, suspected cases</th>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "      <th>Total probable cases</th>\n",
       "      <th>Total suspected cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">6</th>\n",
       "      <th>16</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Cumulative cases among HCW  New case/s (confirmed)  \\\n",
       "Month Day                                                       \n",
       "6     16                          1.0                     1.0   \n",
       "      17                          1.0                     0.0   \n",
       "      22                          1.0                     5.0   \n",
       "      24                          2.0                     4.0   \n",
       "      25                          4.0                     2.0   \n",
       "\n",
       "           Total confirmed cases  Total death/s in confirmed cases  \\\n",
       "Month Day                                                            \n",
       "6     16                    12.0                               8.0   \n",
       "      17                    12.0                               8.0   \n",
       "      22                    28.0                              16.0   \n",
       "      24                    33.0                              18.0   \n",
       "      25                    35.0                              20.0   \n",
       "\n",
       "           Total death/s in confirmed, probable, suspected cases  \\\n",
       "Month Day                                                          \n",
       "6     16                                                16.0       \n",
       "      17                                                16.0       \n",
       "      22                                                25.0       \n",
       "      24                                                32.0       \n",
       "      25                                                37.0       \n",
       "\n",
       "           Total death/s in probable cases  Total death/s in suspected cases  \\\n",
       "Month Day                                                                      \n",
       "6     16                               6.0                               2.0   \n",
       "      17                               6.0                               2.0   \n",
       "      22                               8.0                               1.0   \n",
       "      24                               8.0                               6.0   \n",
       "      25                               8.0                               9.0   \n",
       "\n",
       "           Total probable cases  Total suspected cases  \n",
       "Month Day                                               \n",
       "6     16                    6.0                    4.0  \n",
       "      17                    6.0                    8.0  \n",
       "      22                    8.0                    6.0  \n",
       "      24                    8.0                   13.0  \n",
       "      25                    9.0                   17.0  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_0['New Deaths'] = df_model_0[['Newly reported deaths']]\n",
    "\n",
    "# We check for deaths\n",
    "cols = df_model_0.columns.values.tolist()\n",
    "print(cols);\n",
    "rel_cols = [c for c in cols if ('cases' in c or 'case' in c)]\n",
    "df_model_0.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here it seems that we should select the New case/s (confirmed) column as it seems to correspond to the daily new cases and is not cumulative."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">6</th>\n",
       "      <th>16</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           New Deaths  New Cases\n",
       "Month Day                       \n",
       "6     16          2.0        1.0\n",
       "      17          0.0        0.0\n",
       "      22          4.0        5.0\n",
       "      24          4.0        4.0\n",
       "      25          3.0        2.0"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add the new cases to the model df\n",
    "df_model_0['New Cases'] =  df_model_0['New case/s (confirmed)']\n",
    "# return a df with only the New Death and New Cases\n",
    "df_model_0 = df_model_0[['New Deaths','New Cases']]\n",
    "model_dataframes[0] = df_model_0\n",
    "df_model_0.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### model 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>National</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th>Variable</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">4</th>\n",
       "      <th>New Case/s (Suspected)</th>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New Case/s (Probable)</th>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New case/s (confirmed)</th>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total suspected cases</th>\n",
       "      <td>1190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total probable cases</th>\n",
       "      <td>1796.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  National\n",
       "Month Day Variable                        \n",
       "10    4   New Case/s (Suspected)      23.0\n",
       "          New Case/s (Probable)       14.0\n",
       "          New case/s (confirmed)       5.0\n",
       "          Total suspected cases     1190.0\n",
       "          Total probable cases      1796.0"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_1 = model_dataframes[1]\n",
    "df_model_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df has unique index : False\n",
      "                                                         National\n",
      "Month Day Variable                                               \n",
      "10    4   Cumulative (confirmed + probable + suspected)    3921.0\n",
      "          Cumulative (confirmed + probable + suspected)    3929.0\n",
      "                                      National\n",
      "Month Day Variable                            \n",
      "10    4   Cumulative cases among HCW     190.0\n",
      "          Cumulative cases among HCW     192.0\n",
      "                                       National\n",
      "Month Day Variable                             \n",
      "10    4   Cumulative deaths among HCW      92.0\n",
      "          Cumulative deaths among HCW      94.0\n",
      "                                            National\n",
      "Month Day Variable                                  \n",
      "10    4   Total death/s in confirmed cases    1012.0\n",
      "          Total death/s in confirmed cases    1018.0\n",
      "                                           National\n",
      "Month Day Variable                                 \n",
      "10    4   Total death/s in probable cases     699.0\n",
      "          Total death/s in probable cases     701.0\n",
      "                                            National\n",
      "Month Day Variable                                  \n",
      "10    4   Total death/s in suspected cases     488.0\n",
      "          Total death/s in suspected cases     491.0\n",
      "                                                              National\n",
      "Month Day Variable                                                    \n",
      "10    4   Total death/s in confirmed, probable, suspected...    2199.0\n",
      "          Total death/s in confirmed, probable, suspected...    2210.0\n",
      "                                                         National\n",
      "Month Day Variable                                               \n",
      "10    4   Cumulative (confirmed + probable + suspected)    3921.0\n",
      "          Cumulative (confirmed + probable + suspected)    3929.0\n",
      "                                      National\n",
      "Month Day Variable                            \n",
      "10    4   Cumulative cases among HCW     190.0\n",
      "          Cumulative cases among HCW     192.0\n",
      "                                       National\n",
      "Month Day Variable                             \n",
      "10    4   Cumulative deaths among HCW      92.0\n",
      "          Cumulative deaths among HCW      94.0\n",
      "                                            National\n",
      "Month Day Variable                                  \n",
      "10    4   Total death/s in confirmed cases    1012.0\n",
      "          Total death/s in confirmed cases    1018.0\n",
      "                                           National\n",
      "Month Day Variable                                 \n",
      "10    4   Total death/s in probable cases     699.0\n",
      "          Total death/s in probable cases     701.0\n",
      "                                            National\n",
      "Month Day Variable                                  \n",
      "10    4   Total death/s in suspected cases     488.0\n",
      "          Total death/s in suspected cases     491.0\n",
      "                                                              National\n",
      "Month Day Variable                                                    \n",
      "10    4   Total death/s in confirmed, probable, suspected...    2199.0\n",
      "          Total death/s in confirmed, probable, suspected...    2210.0\n"
     ]
    }
   ],
   "source": [
    "print('df has unique index :', df_model_1.index.is_unique)\n",
    "# here we observe that the model can't be unstack because the index is not unique \n",
    "#(several time the same Variable in the file)\n",
    "\n",
    "# compute a list of the index appearing multiple times:\n",
    "index_list = df_model_1.index.get_level_values(2).tolist()\n",
    "multiple_index_list = [i for i in index_list if index_list.count(i)>1]\n",
    "\n",
    "for index in multiple_index_list:\n",
    "    print(df_model_1.query('Variable == @index'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We observe only a small difference between the value that are in double here, hence we choose to merge them and take the mean in order to be able to unstack the dataframe and see better what Variable we should choose for this particular model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for index in multiple_index_list:\n",
    "    # get the mean of the index values\n",
    "    mean = df_model_1.query('Variable == @index').values.mean()\n",
    "    # drop the multiple index from the df\n",
    "    df_model_1.drop(index, level=2, inplace=True)\n",
    "    # add the mean instead in the df\n",
    "    df_model_1.loc[(df_model_1.index.get_level_values(0)[0], df_model_1.index.get_level_values(1)[0], index)] = [mean]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "my index are unique: True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Case Fatality Rate (CFR) - Confirmed &amp; Probable Cases</th>\n",
       "      <th>Contacts lost to follow-up</th>\n",
       "      <th>Contacts seen</th>\n",
       "      <th>Contacts who completed 21 day follow-up</th>\n",
       "      <th>Cumulative (confirmed + probable + suspected)</th>\n",
       "      <th>Cumulative CFR</th>\n",
       "      <th>Cumulative cases among HCW</th>\n",
       "      <th>Cumulative deaths among HCW</th>\n",
       "      <th>Currently under follow-up</th>\n",
       "      <th>New Case/s (Probable)</th>\n",
       "      <th>...</th>\n",
       "      <th>Total confirmed cases</th>\n",
       "      <th>Total contacts listed</th>\n",
       "      <th>Total death/s in confirmed cases</th>\n",
       "      <th>Total death/s in confirmed, probable, suspected cases</th>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "      <th>Total discharges</th>\n",
       "      <th>Total no. currently in Treatment Units</th>\n",
       "      <th>Total probable cases</th>\n",
       "      <th>Total suspected cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <th>4</th>\n",
       "      <td>62.7</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3101.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>3925.0</td>\n",
       "      <td>62.8</td>\n",
       "      <td>191.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>6689.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>935.0</td>\n",
       "      <td>15271.0</td>\n",
       "      <td>1015.0</td>\n",
       "      <td>2204.5</td>\n",
       "      <td>700.0</td>\n",
       "      <td>489.5</td>\n",
       "      <td>25.0</td>\n",
       "      <td>448.0</td>\n",
       "      <td>1796.0</td>\n",
       "      <td>1190.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows  32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Case Fatality Rate (CFR) - Confirmed & Probable Cases  \\\n",
       "Month Day                                                          \n",
       "10    4                                                 62.7       \n",
       "\n",
       "           Contacts lost to follow-up  Contacts seen  \\\n",
       "Month Day                                              \n",
       "10    4                           2.0         3101.0   \n",
       "\n",
       "           Contacts who completed 21 day follow-up  \\\n",
       "Month Day                                            \n",
       "10    4                                      132.0   \n",
       "\n",
       "           Cumulative (confirmed + probable + suspected)  Cumulative CFR  \\\n",
       "Month Day                                                                  \n",
       "10    4                                           3925.0            62.8   \n",
       "\n",
       "           Cumulative cases among HCW  Cumulative deaths among HCW  \\\n",
       "Month Day                                                            \n",
       "10    4                         191.0                         93.0   \n",
       "\n",
       "           Currently under follow-up  New Case/s (Probable)  \\\n",
       "Month Day                                                     \n",
       "10    4                       6689.0                   14.0   \n",
       "\n",
       "                   ...            Total confirmed cases  \\\n",
       "Month Day          ...                                    \n",
       "10    4            ...                            935.0   \n",
       "\n",
       "           Total contacts listed  Total death/s in confirmed cases  \\\n",
       "Month Day                                                            \n",
       "10    4                  15271.0                            1015.0   \n",
       "\n",
       "           Total death/s in confirmed, probable, suspected cases  \\\n",
       "Month Day                                                          \n",
       "10    4                                               2204.5       \n",
       "\n",
       "           Total death/s in probable cases  Total death/s in suspected cases  \\\n",
       "Month Day                                                                      \n",
       "10    4                              700.0                             489.5   \n",
       "\n",
       "           Total discharges  Total no. currently in Treatment Units  \\\n",
       "Month Day                                                             \n",
       "10    4                25.0                                   448.0   \n",
       "\n",
       "           Total probable cases  Total suspected cases  \n",
       "Month Day                                               \n",
       "10    4                  1796.0                 1190.0  \n",
       "\n",
       "[1 rows x 32 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now our index should be unique:\n",
    "print(\"my index are unique:\", df_model_1.index.is_unique)\n",
    "\n",
    "# so I can unstack and process like the other models:\n",
    "df_model_1 = df_model_1[['National']].unstack('Variable')\n",
    "            \n",
    "# To avoid hierarchical columns, we drop the first level\n",
    "df_model_1.columns = df_model_1.columns.droplevel(0)\n",
    "df_model_1.columns.name = ''\n",
    "\n",
    "df_model_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Case Fatality Rate (CFR) - Confirmed & Probable Cases', 'Contacts lost to follow-up', 'Contacts seen', 'Contacts who completed 21 day follow-up', 'Cumulative (confirmed + probable + suspected)', 'Cumulative CFR', 'Cumulative cases among HCW', 'Cumulative deaths among HCW', 'Currently under follow-up', 'New Case/s (Probable)', 'New Case/s (Suspected)', 'New admissions', 'New case/s (confirmed)', 'Newly Reported Cases in HCW', 'Newly Reported deaths in HCW', 'Newly reported contacts', 'Newly reported deaths', 'Total Case/s (Probable)', 'Total Case/s (Suspected)', 'Total Number of Confirmed Cases of Guinean Nationality', 'Total Number of Confirmed Cases of Sierra Leonean Nationality', 'Total case/s (confirmed)', 'Total confirmed cases', 'Total contacts listed', 'Total death/s in confirmed cases', 'Total death/s in confirmed, probable, suspected cases', 'Total death/s in probable cases', 'Total death/s in suspected cases', 'Total discharges', 'Total no. currently in Treatment Units', 'Total probable cases', 'Total suspected cases']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Cumulative deaths among HCW</th>\n",
       "      <th>Newly Reported deaths in HCW</th>\n",
       "      <th>Newly reported deaths</th>\n",
       "      <th>Total death/s in confirmed cases</th>\n",
       "      <th>Total death/s in confirmed, probable, suspected cases</th>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <th>4</th>\n",
       "      <td>93.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>1015.0</td>\n",
       "      <td>2204.5</td>\n",
       "      <td>700.0</td>\n",
       "      <td>489.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Cumulative deaths among HCW  Newly Reported deaths in HCW  \\\n",
       "Month Day                                                              \n",
       "10    4                           93.0                           0.0   \n",
       "\n",
       "           Newly reported deaths  Total death/s in confirmed cases  \\\n",
       "Month Day                                                            \n",
       "10    4                     43.0                            1015.0   \n",
       "\n",
       "           Total death/s in confirmed, probable, suspected cases  \\\n",
       "Month Day                                                          \n",
       "10    4                                               2204.5       \n",
       "\n",
       "           Total death/s in probable cases  Total death/s in suspected cases  \n",
       "Month Day                                                                     \n",
       "10    4                              700.0                             489.5  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lets check for death:\n",
    "cols = df_model_1.columns.values.tolist()\n",
    "print(cols);\n",
    "rel_cols = [c for c in cols if ('death' in c)]\n",
    "df_model_1.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```Newly reported``` death seems to be the good columns to pick here for this model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_1['New Deaths'] = df_model_1[['Newly reported deaths']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Case Fatality Rate (CFR) - Confirmed & Probable Cases', 'Contacts lost to follow-up', 'Contacts seen', 'Contacts who completed 21 day follow-up', 'Cumulative (confirmed + probable + suspected)', 'Cumulative CFR', 'Cumulative cases among HCW', 'Cumulative deaths among HCW', 'Currently under follow-up', 'New Case/s (Probable)', 'New Case/s (Suspected)', 'New admissions', 'New case/s (confirmed)', 'Newly Reported Cases in HCW', 'Newly Reported deaths in HCW', 'Newly reported contacts', 'Newly reported deaths', 'Total Case/s (Probable)', 'Total Case/s (Suspected)', 'Total Number of Confirmed Cases of Guinean Nationality', 'Total Number of Confirmed Cases of Sierra Leonean Nationality', 'Total case/s (confirmed)', 'Total confirmed cases', 'Total contacts listed', 'Total death/s in confirmed cases', 'Total death/s in confirmed, probable, suspected cases', 'Total death/s in probable cases', 'Total death/s in suspected cases', 'Total discharges', 'Total no. currently in Treatment Units', 'Total probable cases', 'Total suspected cases', 'New Deaths']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Case Fatality Rate (CFR) - Confirmed &amp; Probable Cases</th>\n",
       "      <th>Cumulative cases among HCW</th>\n",
       "      <th>New Case/s (Probable)</th>\n",
       "      <th>New Case/s (Suspected)</th>\n",
       "      <th>New case/s (confirmed)</th>\n",
       "      <th>Newly Reported Cases in HCW</th>\n",
       "      <th>Total Case/s (Probable)</th>\n",
       "      <th>Total Case/s (Suspected)</th>\n",
       "      <th>Total Number of Confirmed Cases of Guinean Nationality</th>\n",
       "      <th>Total Number of Confirmed Cases of Sierra Leonean Nationality</th>\n",
       "      <th>Total case/s (confirmed)</th>\n",
       "      <th>Total confirmed cases</th>\n",
       "      <th>Total death/s in confirmed cases</th>\n",
       "      <th>Total death/s in confirmed, probable, suspected cases</th>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "      <th>Total probable cases</th>\n",
       "      <th>Total suspected cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <th>4</th>\n",
       "      <td>62.7</td>\n",
       "      <td>191.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1798.0</td>\n",
       "      <td>1190.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>941.0</td>\n",
       "      <td>935.0</td>\n",
       "      <td>1015.0</td>\n",
       "      <td>2204.5</td>\n",
       "      <td>700.0</td>\n",
       "      <td>489.5</td>\n",
       "      <td>1796.0</td>\n",
       "      <td>1190.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Case Fatality Rate (CFR) - Confirmed & Probable Cases  \\\n",
       "Month Day                                                          \n",
       "10    4                                                 62.7       \n",
       "\n",
       "           Cumulative cases among HCW  New Case/s (Probable)  \\\n",
       "Month Day                                                      \n",
       "10    4                         191.0                   14.0   \n",
       "\n",
       "           New Case/s (Suspected)  New case/s (confirmed)  \\\n",
       "Month Day                                                   \n",
       "10    4                      23.0                     5.0   \n",
       "\n",
       "           Newly Reported Cases in HCW  Total Case/s (Probable)  \\\n",
       "Month Day                                                         \n",
       "10    4                            1.0                   1798.0   \n",
       "\n",
       "           Total Case/s (Suspected)  \\\n",
       "Month Day                             \n",
       "10    4                      1190.0   \n",
       "\n",
       "           Total Number of Confirmed Cases of Guinean Nationality  \\\n",
       "Month Day                                                           \n",
       "10    4                                                  4.0        \n",
       "\n",
       "           Total Number of Confirmed Cases of Sierra Leonean Nationality  \\\n",
       "Month Day                                                                  \n",
       "10    4                                                 13.0               \n",
       "\n",
       "           Total case/s (confirmed)  Total confirmed cases  \\\n",
       "Month Day                                                    \n",
       "10    4                       941.0                  935.0   \n",
       "\n",
       "           Total death/s in confirmed cases  \\\n",
       "Month Day                                     \n",
       "10    4                              1015.0   \n",
       "\n",
       "           Total death/s in confirmed, probable, suspected cases  \\\n",
       "Month Day                                                          \n",
       "10    4                                               2204.5       \n",
       "\n",
       "           Total death/s in probable cases  Total death/s in suspected cases  \\\n",
       "Month Day                                                                      \n",
       "10    4                              700.0                             489.5   \n",
       "\n",
       "           Total probable cases  Total suspected cases  \n",
       "Month Day                                               \n",
       "10    4                  1796.0                 1190.0  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now lets check for the new case of the day\n",
    "cols = df_model_1.columns.values.tolist()\n",
    "print(cols);\n",
    "rel_cols = [c for c in cols if ('cases' in c or 'case' in c or 'Case' in c)]\n",
    "df_model_1.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The value that seems to make the more sense here is ```New case/s (confirmed)``` so we are going to keep that in our df."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_1['New Cases'] = df_model_1[['New case/s (confirmed)']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <th>4</th>\n",
       "      <td>43.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           New Deaths  New Cases\n",
       "Month Day                       \n",
       "10    4          43.0        5.0"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now we merge and we are done with this model:\n",
    "# return a df with only the New Death and New Cases\n",
    "df_model_1 = df_model_1[['New Deaths','New Cases']]\n",
    "model_dataframes[1] = df_model_1\n",
    "df_model_1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### model 2 - 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this point we realise that even though it was different models of variables, we took at the end the same variable name for the new death and the new cases: ```New case/s (confirmed)``` and ```Newly reported deaths```to not do the same work again and again we chose to first check if this variable exist in the remaining models and just use them if they do. Then directly double check for each model that the value are making sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "New case/s (confirmed) was found in this model with the following value:\n",
      "[[4.0], [3.0], [2.0], [0.0], [0.0], [0.0], [0.0], [1.0], [7.0], [0.0], [1.0]]\n",
      "New case/s (confirmed) was found in this model with the following value:\n",
      "[[0.0]]\n",
      "New case/s (confirmed) was found in this model with the following value:\n",
      "[[0.0], [6.0], [0.0], [1.0], [nan], [nan], [nan], [nan]]\n",
      "New case/s (confirmed) was found in this model with the following value:\n",
      "[[nan], [nan], [nan], [0.0], [nan], [nan], [nan], [nan], [nan], [nan], [nan], [12.0], [7.0], [10.0], [10.0], [1.0], [9.0], [nan], [2867.0], [2867.0], [2869.0], [2869.0], [2927.0], [2946.0]]\n"
     ]
    }
   ],
   "source": [
    "for df_model in model_dataframes[2:6]:\n",
    "    try:\n",
    "        df_model[['New case/s (confirmed)']]\n",
    "        print('New case/s (confirmed) was found in this model with the following value:' )\n",
    "        print(df_model[['New case/s (confirmed)']].values.tolist())\n",
    "    except KeyError as e:\n",
    "        print(str(e) + ' in this model')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Newly reported deaths was found in this model with the following value:\n",
      "[[39.0], [45.0], [25.0], [33.0], [4.0], [22.0], [30.0], [29.0], [16.0], [8.0], [28.0]]\n",
      "Newly reported deaths was found in this model with the following value:\n",
      "[[23.0]]\n",
      "Newly reported deaths was found in this model with the following value:\n",
      "[[19.0], [41.0], [40.0], [24.0], [30.0], [20.0], [29.0], [8.0]]\n",
      "Newly reported deaths was found in this model with the following value:\n",
      "[[41.0], [24.0], [14.0], [20.0], [12.0], [21.0], [30.0], [14.0], [0.0], [15.0], [21.0], [16.0], [nan], [7.0], [nan], [nan], [nan], [nan], [nan], [nan], [nan], [nan], [nan], [nan]]\n"
     ]
    }
   ],
   "source": [
    "for df_model in model_dataframes[2:6]:\n",
    "    try:\n",
    "        df_model[['Newly reported deaths']]\n",
    "        print('Newly reported deaths was found in this model with the following value:' )\n",
    "        print(df_model[['Newly reported deaths']].values.tolist())\n",
    "    except KeyError as e:\n",
    "        print(str(e) + ' in this model')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the 2 last code cells we can see that the ```Newly reported deaths``` seems to make sense in the model 2, 3, 4 but not in the last model (model 5) we need to investigate a little more as we have a lot of NaN values. Regarding the ```New case/s (confirmed)``` the model 2 and 3 seems to work but we need to investigate on model 4 and 5 as we have multiple NaN values (check if there is better alternative)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10</th>\n",
       "      <th>8</th>\n",
       "      <td>39.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>45.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>25.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           New Deaths  New Cases\n",
       "Month Day                       \n",
       "10    8          39.0        4.0\n",
       "      9          45.0        3.0\n",
       "      10         25.0        2.0\n",
       "      11         33.0        0.0\n",
       "      12          4.0        0.0"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## lets deal first with model 2 and 3 with are straight forward:\n",
    "df_model_2 = model_dataframes[2]\n",
    "df_model_2['New Deaths'] = df_model_2[['Newly reported deaths']]\n",
    "df_model_2['New Cases'] = df_model_2[['New case/s (confirmed)']]\n",
    "df_model_2 = df_model_2[['New Deaths','New Cases']]\n",
    "model_dataframes[2] = df_model_2\n",
    "df_model_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <th>21</th>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           New Deaths  New Cases\n",
       "Month Day                       \n",
       "10    21         23.0        0.0"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## lets deal first with model 2 and 3 with are straight forward:\n",
    "df_model_3 = model_dataframes[3]\n",
    "df_model_3['New Deaths'] = df_model_3[['Newly reported deaths']]\n",
    "df_model_3['New Cases'] = df_model_3[['New case/s (confirmed)']]\n",
    "df_model_3 = df_model_3[['New Deaths','New Cases']]\n",
    "model_dataframes[3] = df_model_3\n",
    "df_model_3.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now for model 4 it seems that we can use the ```Newly reported deaths``` but we are not sure about the ```New case/s (confirmed)``` so let's take look at the whole dataframe for this one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Case Fatality Rate (CFR) - Confirmed & Probable Cases', 'Contacts lost to follow-up', 'Contacts seen', 'Contacts who completed 21 day follow-up', 'Cumulative admission/isolation', 'Cumulative cases among HCW', 'Cumulative confirmed, probable and suspected cases', 'Cumulative deaths among HCW', 'Currently under follow-up', 'New Case/s (Probable)', 'New Case/s (Suspected)', 'New admissions', 'New case/s (confirmed)', 'Newly Reported Cases in HCW', 'Newly Reported deaths in HCW', 'Newly reported contacts', 'Newly reported deaths', 'Specimens collected', 'Specimens pending for testing', 'Total Number of Confirmed Cases of Guinean Nationality', 'Total Number of Confirmed Cases of Sierra Leonean Nationality', 'Total confirmed cases', 'Total contacts listed', 'Total death/s in confirmed cases', 'Total death/s in confirmed, probable, suspected cases', 'Total death/s in probable cases', 'Total death/s in suspected cases', 'Total discharges', 'Total no. currently in Treatment Units', 'Total probable cases', 'Total specimens tested', 'Total suspected cases']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Case Fatality Rate (CFR) - Confirmed &amp; Probable Cases</th>\n",
       "      <th>Cumulative cases among HCW</th>\n",
       "      <th>Cumulative confirmed, probable and suspected cases</th>\n",
       "      <th>New Case/s (Probable)</th>\n",
       "      <th>New Case/s (Suspected)</th>\n",
       "      <th>New case/s (confirmed)</th>\n",
       "      <th>Newly Reported Cases in HCW</th>\n",
       "      <th>Total Number of Confirmed Cases of Guinean Nationality</th>\n",
       "      <th>Total Number of Confirmed Cases of Sierra Leonean Nationality</th>\n",
       "      <th>Total confirmed cases</th>\n",
       "      <th>Total death/s in confirmed cases</th>\n",
       "      <th>Total death/s in confirmed, probable, suspected cases</th>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "      <th>Total probable cases</th>\n",
       "      <th>Total suspected cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10</th>\n",
       "      <th>22</th>\n",
       "      <td>NaN</td>\n",
       "      <td>296.0</td>\n",
       "      <td>6166.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2218.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2168.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1556.0</td>\n",
       "      <td>2392.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>299.0</td>\n",
       "      <td>6248.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2229.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2104.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1591.0</td>\n",
       "      <td>2428.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>299.0</td>\n",
       "      <td>6253.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2232.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2104.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1592.0</td>\n",
       "      <td>2429.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NaN</td>\n",
       "      <td>299.0</td>\n",
       "      <td>6267.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2240.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2106.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1596.0</td>\n",
       "      <td>2431.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NaN</td>\n",
       "      <td>305.0</td>\n",
       "      <td>6398.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2304.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2471.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1642.0</td>\n",
       "      <td>2452.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Case Fatality Rate (CFR) - Confirmed & Probable Cases  \\\n",
       "Month Day                                                          \n",
       "10    22                                                 NaN       \n",
       "      23                                                 NaN       \n",
       "      24                                                 NaN       \n",
       "      25                                                 NaN       \n",
       "      28                                                 NaN       \n",
       "\n",
       "           Cumulative cases among HCW  \\\n",
       "Month Day                               \n",
       "10    22                        296.0   \n",
       "      23                        299.0   \n",
       "      24                        299.0   \n",
       "      25                        299.0   \n",
       "      28                        305.0   \n",
       "\n",
       "           Cumulative confirmed, probable and suspected cases  \\\n",
       "Month Day                                                       \n",
       "10    22                                              6166.0    \n",
       "      23                                              6248.0    \n",
       "      24                                              6253.0    \n",
       "      25                                              6267.0    \n",
       "      28                                              6398.0    \n",
       "\n",
       "           New Case/s (Probable)  New Case/s (Suspected)  \\\n",
       "Month Day                                                  \n",
       "10    22                    10.0                    20.0   \n",
       "      23                     5.0                    35.0   \n",
       "      24                    12.0                    23.0   \n",
       "      25                    18.0                    12.0   \n",
       "      28                    25.0                    30.0   \n",
       "\n",
       "           New case/s (confirmed)  Newly Reported Cases in HCW  \\\n",
       "Month Day                                                        \n",
       "10    22                      0.0                          0.0   \n",
       "      23                      6.0                          0.0   \n",
       "      24                      0.0                          0.0   \n",
       "      25                      1.0                          0.0   \n",
       "      28                      NaN                          0.0   \n",
       "\n",
       "           Total Number of Confirmed Cases of Guinean Nationality  \\\n",
       "Month Day                                                           \n",
       "10    22                                                 NaN        \n",
       "      23                                                 NaN        \n",
       "      24                                                 NaN        \n",
       "      25                                                 NaN        \n",
       "      28                                                 NaN        \n",
       "\n",
       "           Total Number of Confirmed Cases of Sierra Leonean Nationality  \\\n",
       "Month Day                                                                  \n",
       "10    22                                                 NaN               \n",
       "      23                                                 NaN               \n",
       "      24                                                 NaN               \n",
       "      25                                                 NaN               \n",
       "      28                                                 NaN               \n",
       "\n",
       "           Total confirmed cases  Total death/s in confirmed cases  \\\n",
       "Month Day                                                            \n",
       "10    22                  2218.0                               NaN   \n",
       "      23                  2229.0                               NaN   \n",
       "      24                  2232.0                               NaN   \n",
       "      25                  2240.0                               NaN   \n",
       "      28                  2304.0                               NaN   \n",
       "\n",
       "           Total death/s in confirmed, probable, suspected cases  \\\n",
       "Month Day                                                          \n",
       "10    22                                              2168.0       \n",
       "      23                                              2104.0       \n",
       "      24                                              2104.0       \n",
       "      25                                              2106.0       \n",
       "      28                                              2471.0       \n",
       "\n",
       "           Total death/s in probable cases  Total death/s in suspected cases  \\\n",
       "Month Day                                                                      \n",
       "10    22                               NaN                               NaN   \n",
       "      23                               NaN                               NaN   \n",
       "      24                               NaN                               NaN   \n",
       "      25                               NaN                               NaN   \n",
       "      28                               NaN                               NaN   \n",
       "\n",
       "           Total probable cases  Total suspected cases  \n",
       "Month Day                                               \n",
       "10    22                 1556.0                 2392.0  \n",
       "      23                 1591.0                 2428.0  \n",
       "      24                 1592.0                 2429.0  \n",
       "      25                 1596.0                 2431.0  \n",
       "      28                 1642.0                 2452.0  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_4 = model_dataframes[4]\n",
    "cols = df_model_4.columns.values.tolist()\n",
    "print(cols);\n",
    "rel_cols = [c for c in cols if ('case' in c or 'Case' in c)]\n",
    "df_model_4.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We clearly see that there is a lot of missing values in the end of this DataFrame unfortunately we can't take or other columns which are globally much higher or use them to approximate the missing cases. So we decided here to simply keep the NaN value that will be igore in our mean calculation at the end in order to not biase the monthly average at the end."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10</th>\n",
       "      <th>22</th>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>41.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>30.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           New Deaths  New Cases\n",
       "Month Day                       \n",
       "10    22         19.0        0.0\n",
       "      23         41.0        6.0\n",
       "      24         40.0        0.0\n",
       "      25         24.0        1.0\n",
       "      28         30.0        NaN"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# since there is no better alterative we go for the straight forward approach:\n",
    "df_model_4['New Deaths'] = df_model_4[['Newly reported deaths']]\n",
    "df_model_4['New Cases'] = df_model_4[['New case/s (confirmed)']]\n",
    "df_model_4 = df_model_4[['New Deaths','New Cases']]\n",
    "model_dataframes[4] = df_model_4\n",
    "df_model_4.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now for the last model we saw a lot of NaN values as well so lets check the whole dataframe again to see if we can find a better alternative than ```Newly reported deaths``` and ```New case/s (confirmed)```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Case Fatality Rate (CFR) - \\n Confirmed & Probable Cases', 'Contacts lost to follow-up', 'Contacts seen', 'Contacts who completed 21 day \\n follow-up', 'Cumulative admission/isolation', 'Cumulative cases among HCW', 'Cumulative confirmed, probable and suspected cases', 'Cumulative deaths among HCW', 'Currently under follow-up', 'New Case/s (Probable)', 'New Case/s (Suspected)', 'New admissions', 'New case/s (confirmed)', 'Newly Reported Cases in HCW', 'Newly Reported deaths in HCW', 'Newly reported contacts', 'Newly reported deaths', 'Specimens collected', 'Specimens pending for testing', 'Total Number of Confirmed Cases \\n of Guinean Nationality', 'Total Number of Confirmed Cases \\n of Sierra Leonean Nationality', 'Total confirmed cases', 'Total contacts listed', 'Total death/s in confirmed cases', 'Total death/s in confirmed, \\n probable, suspected cases', 'Total death/s in probable cases', 'Total death/s in suspected cases', 'Total discharges', 'Total no. currently in Treatment \\n Units', 'Total probable cases', 'Total specimens tested', 'Total suspected cases']\n"
     ]
    }
   ],
   "source": [
    "df_model_5 = model_dataframes[5]\n",
    "cols = df_model_5.columns.values.tolist()\n",
    "print(cols);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Cumulative deaths among HCW</th>\n",
       "      <th>Newly Reported deaths in HCW</th>\n",
       "      <th>Newly reported deaths</th>\n",
       "      <th>Total death/s in confirmed cases</th>\n",
       "      <th>Total death/s in confirmed, \n",
       " probable, suspected cases</th>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <th>31</th>\n",
       "      <td>157.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2697.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">11</th>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>162.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>170.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2960.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>170.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2964.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Cumulative deaths among HCW  Newly Reported deaths in HCW  \\\n",
       "Month Day                                                              \n",
       "10    31                         157.0                           0.0   \n",
       "11    4                            NaN                           0.0   \n",
       "      8                          162.0                           0.0   \n",
       "      14                         170.0                           0.0   \n",
       "      15                         170.0                           0.0   \n",
       "\n",
       "           Newly reported deaths  Total death/s in confirmed cases  \\\n",
       "Month Day                                                            \n",
       "10    31                    41.0                               NaN   \n",
       "11    4                     24.0                               NaN   \n",
       "      8                     14.0                               NaN   \n",
       "      14                    20.0                               0.0   \n",
       "      15                    12.0                               NaN   \n",
       "\n",
       "           Total death/s in confirmed, \\n probable, suspected cases  \\\n",
       "Month Day                                                             \n",
       "10    31                                              2697.0          \n",
       "11    4                                               1179.0          \n",
       "      8                                                  NaN          \n",
       "      14                                              2960.0          \n",
       "      15                                              2964.0          \n",
       "\n",
       "           Total death/s in probable cases  Total death/s in suspected cases  \n",
       "Month Day                                                                     \n",
       "10    31                               NaN                               NaN  \n",
       "11    4                                NaN                               NaN  \n",
       "      8                                NaN                               NaN  \n",
       "      14                               0.0                               0.0  \n",
       "      15                               NaN                               NaN  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rel_cols = [c for c in cols if ('death' in c or 'Death' in c)]\n",
    "df_model_5.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unfortunately we see for the death that we can't find any other columns with value that could help us fill the NaN values so we will stick with Newly reported deaths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_5['New Deaths'] = df_model_5[['Newly reported deaths']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Case Fatality Rate (CFR) - \n",
       " Confirmed &amp; Probable Cases</th>\n",
       "      <th>Cumulative cases among HCW</th>\n",
       "      <th>Cumulative confirmed, probable and suspected cases</th>\n",
       "      <th>New Case/s (Probable)</th>\n",
       "      <th>New Case/s (Suspected)</th>\n",
       "      <th>New case/s (confirmed)</th>\n",
       "      <th>Newly Reported Cases in HCW</th>\n",
       "      <th>Total Number of Confirmed Cases \n",
       " of Guinean Nationality</th>\n",
       "      <th>Total Number of Confirmed Cases \n",
       " of Sierra Leonean Nationality</th>\n",
       "      <th>Total confirmed cases</th>\n",
       "      <th>Total death/s in confirmed cases</th>\n",
       "      <th>Total death/s in confirmed, \n",
       " probable, suspected cases</th>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "      <th>Total probable cases</th>\n",
       "      <th>Total suspected cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <th>31</th>\n",
       "      <td>NaN</td>\n",
       "      <td>315.0</td>\n",
       "      <td>6525.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2697.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1627.0</td>\n",
       "      <td>2447.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">11</th>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>318.0</td>\n",
       "      <td>6619.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2766.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2514.0</td>\n",
       "      <td>1629.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>329.0</td>\n",
       "      <td>6822.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2553.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1687.0</td>\n",
       "      <td>2582.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>340.0</td>\n",
       "      <td>7060.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2642.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2960.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1757.0</td>\n",
       "      <td>2661.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NaN</td>\n",
       "      <td>341.0</td>\n",
       "      <td>7069.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2643.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2964.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1762.0</td>\n",
       "      <td>2664.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Case Fatality Rate (CFR) - \\n Confirmed & Probable Cases  \\\n",
       "Month Day                                                             \n",
       "10    31                                                 NaN          \n",
       "11    4                                                  NaN          \n",
       "      8                                                  NaN          \n",
       "      14                                                 0.0          \n",
       "      15                                                 NaN          \n",
       "\n",
       "           Cumulative cases among HCW  \\\n",
       "Month Day                               \n",
       "10    31                        315.0   \n",
       "11    4                         318.0   \n",
       "      8                         329.0   \n",
       "      14                        340.0   \n",
       "      15                        341.0   \n",
       "\n",
       "           Cumulative confirmed, probable and suspected cases  \\\n",
       "Month Day                                                       \n",
       "10    31                                              6525.0    \n",
       "11    4                                               6619.0    \n",
       "      8                                               6822.0    \n",
       "      14                                              7060.0    \n",
       "      15                                              7069.0    \n",
       "\n",
       "           New Case/s (Probable)  New Case/s (Suspected)  \\\n",
       "Month Day                                                  \n",
       "10    31                    14.0                    18.0   \n",
       "11    4                      3.0                    23.0   \n",
       "      8                     13.0                    15.0   \n",
       "      14                     5.0                    15.0   \n",
       "      15                    11.0                    28.0   \n",
       "\n",
       "           New case/s (confirmed)  Newly Reported Cases in HCW  \\\n",
       "Month Day                                                        \n",
       "10    31                      NaN                          0.0   \n",
       "11    4                       NaN                          0.0   \n",
       "      8                       NaN                          0.0   \n",
       "      14                      0.0                          1.0   \n",
       "      15                      NaN                          0.0   \n",
       "\n",
       "           Total Number of Confirmed Cases \\n of Guinean Nationality  \\\n",
       "Month Day                                                              \n",
       "10    31                                                 NaN           \n",
       "11    4                                                  NaN           \n",
       "      8                                                  NaN           \n",
       "      14                                                 0.0           \n",
       "      15                                                 NaN           \n",
       "\n",
       "           Total Number of Confirmed Cases \\n of Sierra Leonean Nationality  \\\n",
       "Month Day                                                                     \n",
       "10    31                                                 NaN                  \n",
       "11    4                                                  NaN                  \n",
       "      8                                                  NaN                  \n",
       "      14                                                 0.0                  \n",
       "      15                                                 NaN                  \n",
       "\n",
       "           Total confirmed cases  Total death/s in confirmed cases  \\\n",
       "Month Day                                                            \n",
       "10    31                    73.0                               NaN   \n",
       "11    4                   2766.0                               NaN   \n",
       "      8                   2553.0                               NaN   \n",
       "      14                  2642.0                               0.0   \n",
       "      15                  2643.0                               NaN   \n",
       "\n",
       "           Total death/s in confirmed, \\n probable, suspected cases  \\\n",
       "Month Day                                                             \n",
       "10    31                                              2697.0          \n",
       "11    4                                               1179.0          \n",
       "      8                                                  NaN          \n",
       "      14                                              2960.0          \n",
       "      15                                              2964.0          \n",
       "\n",
       "           Total death/s in probable cases  Total death/s in suspected cases  \\\n",
       "Month Day                                                                      \n",
       "10    31                               NaN                               NaN   \n",
       "11    4                                NaN                               NaN   \n",
       "      8                                NaN                               NaN   \n",
       "      14                               0.0                               0.0   \n",
       "      15                               NaN                               NaN   \n",
       "\n",
       "           Total probable cases  Total suspected cases  \n",
       "Month Day                                               \n",
       "10    31                 1627.0                 2447.0  \n",
       "11    4                  2514.0                 1629.0  \n",
       "      8                  1687.0                 2582.0  \n",
       "      14                 1757.0                 2661.0  \n",
       "      15                 1762.0                 2664.0  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# let's do the same with the new cases:\n",
    "rel_cols = [c for c in cols if ('case' in c or 'Case' in c)]\n",
    "df_model_5.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here again we dont have many other option to estimate the value than keep the NaN value however we can observe that the few last value seems to be out of range and doesn't make sense with the rest of the data so we decided to take them out by putting a treshhold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Month  Day\n",
       "10     31     NaN\n",
       "11     4      NaN\n",
       "       8      NaN\n",
       "       14     0.0\n",
       "       15     NaN\n",
       "Name: New Cases, dtype: float64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_5['New Cases'] = df_model_5[['New case/s (confirmed)']]\n",
    "# treshhold filter\n",
    "df_model_5[df_model_5['New Cases'] > 1500] = np.nan\n",
    "df_model_5['New Cases'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <th>31</th>\n",
       "      <td>41.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">11</th>\n",
       "      <th>4</th>\n",
       "      <td>24.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>14.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           New Deaths  New Cases\n",
       "Month Day                       \n",
       "10    31         41.0        NaN\n",
       "11    4          24.0        NaN\n",
       "      8          14.0        NaN\n",
       "      14         20.0        0.0\n",
       "      15         12.0        NaN"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_5 = df_model_5[['New Deaths','New Cases']]\n",
    "model_dataframes[5] = df_model_5\n",
    "df_model_5.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finally we can concat all the different model dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <th>Month</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">Liberia</th>\n",
       "      <th>6</th>\n",
       "      <td>2.000</td>\n",
       "      <td>2.143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4.273</td>\n",
       "      <td>1.818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>23.222</td>\n",
       "      <td>5.444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>37.609</td>\n",
       "      <td>6.167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>28.040</td>\n",
       "      <td>1.619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>15.538</td>\n",
       "      <td>7.800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               New Deaths  New Cases\n",
       "Country Month                       \n",
       "Liberia 6           2.000      2.143\n",
       "        7           4.273      1.818\n",
       "        8          23.222      5.444\n",
       "        9          37.609      6.167\n",
       "        10         28.040      1.619\n",
       "        11         15.538      7.800\n",
       "        12            NaN      5.000"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_liberia = pd.concat(model_dataframes,axis=0)\n",
    "df_liberia['Country'] = 'Liberia'\n",
    "df_liberia = df_liberia.reset_index().groupby(['Country','Month']).mean()[['New Deaths','New Cases']].round(3)\n",
    "df_liberia"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We observe that we don't have any consistant info for the 12th month in terms of death."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3 - Cleaning Sierra Leone"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we will work on Sierra Leone datas. As we did before, we identify the models. To be able to clean them as a whole, we will drop the columns of each region as we are only interested in national numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have found 2 different model(s).\n",
      "69 file(s) in model 0\n",
      "34 file(s) in model 1\n"
     ]
    }
   ],
   "source": [
    "path = SIERRA_LEONE_FOLDER\n",
    "model_dataframes = get_grouped_dataframes(SIERRA_LEONE_FOLDER,'variable','date','National')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have found 2 different 'models' of description values. Therefore we select the columns of interest for each of them. To do so we list the columns and look at those that talk about cases and deaths."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### model 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>death_confirmed</th>\n",
       "      <th>death_probable</th>\n",
       "      <th>death_suspected</th>\n",
       "      <th>etc_cum_deaths</th>\n",
       "      <th>etc_new_deaths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">8</th>\n",
       "      <th>12</th>\n",
       "      <td>264</td>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>273</td>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>280</td>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>287</td>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>297</td>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          death_confirmed death_probable death_suspected etc_cum_deaths  \\\n",
       "Month Day                                                                 \n",
       "8     12              264             34               5            NaN   \n",
       "      13              273             34               5            NaN   \n",
       "      14              280             34               5            NaN   \n",
       "      15              287             34               5            NaN   \n",
       "      16              297             34               5            NaN   \n",
       "\n",
       "          etc_new_deaths  \n",
       "Month Day                 \n",
       "8     12             NaN  \n",
       "      13             NaN  \n",
       "      14             NaN  \n",
       "      15             NaN  \n",
       "      16             NaN  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_0 = model_dataframes[0]\n",
    "\n",
    "# We check for deaths\n",
    "cols = df_model_0.columns.values.tolist()\n",
    "rel_cols = [c for c in cols if ('death' in c or 'Death' in c)]\n",
    "df_model_0.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "By observing the data frame of our model, we see that the column that interest us is ```death_confirmed``` because we don't want to take into account deaths that were not caused for sure by Ebola. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>death_confirmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">8</th>\n",
       "      <th>12</th>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          death_confirmed\n",
       "Month Day                \n",
       "8     12              264\n",
       "      13              273\n",
       "      14              280\n",
       "      15              287\n",
       "      16              297"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df_model_0.loc[:,['death_confirmed']]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that it is a cumulative value. We have to find the correct number of death for each day.\n",
    "We decide to proceed like this :\n",
    "- We set the first value to NaN since we don't know how many death before \n",
    "- We compute the difference between each row and we divide by the number of days between the rows because sometimes a day is missing in the dataset. (this will result in having floats but since we want the daily average in the end, we cannot account for one day the number of deaths that happenned in all the past days since last measure)  \n",
    "- If the difference is negative, we set the value to NaN because it's not possible. It would mean less death after than before. \n",
    "\n",
    "We have not found interesting datas in other columns to help us managing theses cases. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>death_confirmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">8</th>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          death_confirmed\n",
       "Month Day                \n",
       "8     12              NaN\n",
       "      13                9\n",
       "      14                7\n",
       "      15                7\n",
       "      16               10"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_non_cumulative(df,column_name):\n",
    "    to_return = df.loc[:,[column_name]]\n",
    "\n",
    "    i = 0;\n",
    "    for index in df.itertuples():    \n",
    "        if i == 0:\n",
    "            to_return.iloc[i].death_confirmed = float('NaN')\n",
    "        else:\n",
    "            date_now = date(2014, index[0][0], index[0][1])\n",
    "            date_previous = date(2014, df.iloc[i-1].name[0], df.iloc[i-1].name[1])\n",
    "            diff_days = float((date_now - date_previous).days)\n",
    "            diff_deaths = float(index.death_confirmed) - float(df.iloc[i-1][column_name])\n",
    "\n",
    "            if diff_deaths < 0:\n",
    "                to_return.iloc[i].death_confirmed = float('NaN')\n",
    "            else:\n",
    "                to_return.iloc[i].death_confirmed = (diff_deaths)/diff_days\n",
    "        i = i+1\n",
    "    return to_return\n",
    "    \n",
    "new_df = get_non_cumulative(df_model_0,'death_confirmed')\n",
    "new_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_0['New Deaths'] =  new_df['death_confirmed']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These values correspond now to the deaths registered each day. We save it on our model dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>etc_new_admission</th>\n",
       "      <th>etc_new_deaths</th>\n",
       "      <th>etc_new_discharges</th>\n",
       "      <th>new_completed_contacts</th>\n",
       "      <th>new_confirmed</th>\n",
       "      <th>new_contacts</th>\n",
       "      <th>new_noncase</th>\n",
       "      <th>new_probable</th>\n",
       "      <th>new_suspected</th>\n",
       "      <th>New Deaths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">8</th>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73</td>\n",
       "      <td>11</td>\n",
       "      <td>743</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68</td>\n",
       "      <td>15</td>\n",
       "      <td>85</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>58</td>\n",
       "      <td>13</td>\n",
       "      <td>179</td>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>52</td>\n",
       "      <td>10</td>\n",
       "      <td>79</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86</td>\n",
       "      <td>18</td>\n",
       "      <td>72</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          etc_new_admission etc_new_deaths etc_new_discharges  \\\n",
       "Month Day                                                       \n",
       "8     12                NaN            NaN                NaN   \n",
       "      13                NaN            NaN                NaN   \n",
       "      14                NaN            NaN                NaN   \n",
       "      15                NaN            NaN                NaN   \n",
       "      16                NaN            NaN                NaN   \n",
       "\n",
       "          new_completed_contacts new_confirmed new_contacts new_noncase  \\\n",
       "Month Day                                                                 \n",
       "8     12                      73            11          743           4   \n",
       "      13                      68            15           85          13   \n",
       "      14                      58            13          179          26   \n",
       "      15                      52            10           79          15   \n",
       "      16                      86            18           72           9   \n",
       "\n",
       "          new_probable new_suspected New Deaths  \n",
       "Month Day                                        \n",
       "8     12             1            10        NaN  \n",
       "      13             1             3          9  \n",
       "      14             2             0          7  \n",
       "      15             1             6          7  \n",
       "      16             0             3         10  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We check now for cases\n",
    "cols = df_model_0.columns.values.tolist()\n",
    "rel_cols = [c for c in cols if ('cases' in c or 'New' in c or 'new' in c)]\n",
    "df_model_0.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that the column that interest us to represent the daily new cases is the ```new confirmed```."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_0['New Cases'] =  df_model_0['new_confirmed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">8</th>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>9</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>10</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          New Deaths New Cases\n",
       "Month Day                     \n",
       "8     12         NaN        11\n",
       "      13           9        15\n",
       "      14           7        13\n",
       "      15           7        10\n",
       "      16          10        18"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_0 = df_model_0[['New Deaths','New Cases']]\n",
    "model_dataframes[0] = df_model_0\n",
    "df_model_0.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we do the same pipeline for the second model."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### model 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>National</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th>variable</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">23</th>\n",
       "      <th>population</th>\n",
       "      <td>6348350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>new_noncase</th>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>new_suspected</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>new_probable</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>new_confirmed</th>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        National\n",
       "Month Day variable              \n",
       "10    23  population     6348350\n",
       "          new_noncase         76\n",
       "          new_suspected        7\n",
       "          new_probable         0\n",
       "          new_confirmed       60"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_1 = model_dataframes[1]\n",
    "df_model_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "variable\n",
      "cfr                       34\n",
      "contacts_followed         34\n",
      "contacts_healthy          34\n",
      "contacts_ill              34\n",
      "contacts_not_seen         34\n",
      "cum_completed_contacts    34\n",
      "cum_confirmed             34\n",
      "cum_contacts              34\n",
      "cum_noncase               34\n",
      "cum_probable              34\n",
      "cum_suspected             34\n",
      "death_confirmed           34\n",
      "death_probable            34\n",
      "death_suspected           34\n",
      "etc_cum_admission         34\n",
      "etc_cum_deaths            34\n",
      "etc_cum_discharges        34\n",
      "etc_currently_admitted    34\n",
      "etc_new_admission         34\n",
      "etc_new_deaths            34\n",
      "etc_new_discharges        34\n",
      "negative_corpse           35\n",
      "new_completed_contacts    34\n",
      "new_confirmed             34\n",
      "new_contacts              34\n",
      "new_negative              34\n",
      "new_noncase               34\n",
      "new_positive              34\n",
      "new_probable              34\n",
      "new_samples               34\n",
      "new_suspected             34\n",
      "pending                   35\n",
      "percent_seen              34\n",
      "population                34\n",
      "positive_corpse           35\n",
      "repeat_samples            34\n",
      "total_lab_samples         34\n",
      "Name: Day, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "variable_counts = df_model_1.reset_index().groupby('variable').count()['Day']\n",
    "print(variable_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One can see that the majority number of variable occurence is 34, therefore, the one with 35 must be duplicated for some days. We just want to check if they are of any interest, because if they are not we can simply drop them. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variable\n",
       "negative_corpse    35\n",
       "pending            35\n",
       "positive_corpse    35\n",
       "Name: Day, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variable_counts[variable_counts == 35]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see from the name of those variables that we don't really care about the ```negative_corpse, pending, positive_corpse```. Therefore we get rid of them to be able to unstack the variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df = df_model_1.reset_index()\n",
    "df_model_1 = df[(df['variable'] != 'negative_corpse') & (df['variable'] != 'positive_corpse') & (df['variable'] != 'pending')]\n",
    "df_model_1.set_index(['Month','Day','variable'],inplace=True)\n",
    "df_model_1 = df_model_1[['National']].unstack('variable')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>cfr</th>\n",
       "      <th>contacts_followed</th>\n",
       "      <th>contacts_healthy</th>\n",
       "      <th>contacts_ill</th>\n",
       "      <th>contacts_not_seen</th>\n",
       "      <th>cum_completed_contacts</th>\n",
       "      <th>cum_confirmed</th>\n",
       "      <th>cum_contacts</th>\n",
       "      <th>cum_noncase</th>\n",
       "      <th>cum_probable</th>\n",
       "      <th>...</th>\n",
       "      <th>new_negative</th>\n",
       "      <th>new_noncase</th>\n",
       "      <th>new_positive</th>\n",
       "      <th>new_probable</th>\n",
       "      <th>new_samples</th>\n",
       "      <th>new_suspected</th>\n",
       "      <th>percent_seen</th>\n",
       "      <th>population</th>\n",
       "      <th>repeat_samples</th>\n",
       "      <th>total_lab_samples</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10</th>\n",
       "      <th>23</th>\n",
       "      <td>29</td>\n",
       "      <td>11673</td>\n",
       "      <td>11384</td>\n",
       "      <td>20</td>\n",
       "      <td>199</td>\n",
       "      <td>13154</td>\n",
       "      <td>3449</td>\n",
       "      <td>24855</td>\n",
       "      <td>3049</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>143</td>\n",
       "      <td>7</td>\n",
       "      <td>98%</td>\n",
       "      <td>6348350</td>\n",
       "      <td>36</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>29</td>\n",
       "      <td>12786</td>\n",
       "      <td>11912</td>\n",
       "      <td>1578</td>\n",
       "      <td>53</td>\n",
       "      <td>13260</td>\n",
       "      <td>3490</td>\n",
       "      <td>25074</td>\n",
       "      <td>3118</td>\n",
       "      <td>47</td>\n",
       "      <td>...</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>41</td>\n",
       "      <td>10</td>\n",
       "      <td>113</td>\n",
       "      <td>3</td>\n",
       "      <td>98%</td>\n",
       "      <td>6348350</td>\n",
       "      <td>24</td>\n",
       "      <td>137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3241</td>\n",
       "      <td>82</td>\n",
       "      <td>...</td>\n",
       "      <td>123</td>\n",
       "      <td>123</td>\n",
       "      <td>70</td>\n",
       "      <td>9</td>\n",
       "      <td>201</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.34835e+06</td>\n",
       "      <td>39</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>29</td>\n",
       "      <td>11718</td>\n",
       "      <td>12383</td>\n",
       "      <td>21</td>\n",
       "      <td>104</td>\n",
       "      <td>13593</td>\n",
       "      <td>3622</td>\n",
       "      <td>25374</td>\n",
       "      <td>3264</td>\n",
       "      <td>79</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>96</td>\n",
       "      <td>11</td>\n",
       "      <td>98%</td>\n",
       "      <td>6348350</td>\n",
       "      <td>26</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>28</td>\n",
       "      <td>12131</td>\n",
       "      <td>11936</td>\n",
       "      <td>48</td>\n",
       "      <td>109</td>\n",
       "      <td>13918</td>\n",
       "      <td>3713</td>\n",
       "      <td>26177</td>\n",
       "      <td>3357</td>\n",
       "      <td>84</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>98%</td>\n",
       "      <td>6348350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cfr contacts_followed contacts_healthy contacts_ill  \\\n",
       "Month Day                                                       \n",
       "10    23   29             11673            11384           20   \n",
       "      24   29             12786            11912         1578   \n",
       "      25   29               NaN              NaN          NaN   \n",
       "      26   29             11718            12383           21   \n",
       "      27   28             12131            11936           48   \n",
       "\n",
       "          contacts_not_seen cum_completed_contacts cum_confirmed cum_contacts  \\\n",
       "Month Day                                                                       \n",
       "10    23                199                  13154          3449        24855   \n",
       "      24                 53                  13260          3490        25074   \n",
       "      25                NaN                    NaN          3560          NaN   \n",
       "      26                104                  13593          3622        25374   \n",
       "      27                109                  13918          3713        26177   \n",
       "\n",
       "          cum_noncase cum_probable        ...        new_negative new_noncase  \\\n",
       "Month Day                                 ...                                   \n",
       "10    23         3049           37        ...                  76          76   \n",
       "      24         3118           47        ...                  69          69   \n",
       "      25         3241           82        ...                 123         123   \n",
       "      26         3264           79        ...                  23          23   \n",
       "      27         3357           84        ...                 NaN          93   \n",
       "\n",
       "          new_positive new_probable new_samples new_suspected percent_seen  \\\n",
       "Month Day                                                                    \n",
       "10    23            60            0         143             7          98%   \n",
       "      24            41           10         113             3          98%   \n",
       "      25            70            9         201             8          NaN   \n",
       "      26            62            0          96            11          98%   \n",
       "      27           NaN            5         NaN            10          98%   \n",
       "\n",
       "            population repeat_samples total_lab_samples  \n",
       "Month Day                                                \n",
       "10    23       6348350             36               149  \n",
       "      24       6348350             24               137  \n",
       "      25   6.34835e+06             39               240  \n",
       "      26       6348350             26               122  \n",
       "      27       6348350            NaN               NaN  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#to get rid of the hierarchical column and index\n",
    "df_model_1.columns = df_model_1.columns.droplevel(0)\n",
    "df_model_1.columns.name = ''\n",
    "df_model_1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we need to look at the column that we are interested in. Let's start with the one containing the word death."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cfr\n",
      "contacts_followed\n",
      "contacts_healthy\n",
      "contacts_ill\n",
      "contacts_not_seen\n",
      "cum_completed_contacts\n",
      "cum_confirmed\n",
      "cum_contacts\n",
      "cum_noncase\n",
      "cum_probable\n",
      "cum_suspected\n",
      "death_confirmed\n",
      "death_probable\n",
      "death_suspected\n",
      "etc_cum_admission\n",
      "etc_cum_deaths\n",
      "etc_cum_discharges\n",
      "etc_currently_admitted\n",
      "etc_new_admission\n",
      "etc_new_deaths\n",
      "etc_new_discharges\n",
      "new_completed_contacts\n",
      "new_confirmed\n",
      "new_contacts\n",
      "new_negative\n",
      "new_noncase\n",
      "new_positive\n",
      "new_probable\n",
      "new_samples\n",
      "new_suspected\n",
      "percent_seen\n",
      "population\n",
      "repeat_samples\n",
      "total_lab_samples\n"
     ]
    }
   ],
   "source": [
    "cols = df_model_1.columns.values.tolist()\n",
    "for c in cols:\n",
    "    print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>death_confirmed</th>\n",
       "      <th>death_probable</th>\n",
       "      <th>death_suspected</th>\n",
       "      <th>etc_cum_deaths</th>\n",
       "      <th>etc_new_deaths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10</th>\n",
       "      <th>23</th>\n",
       "      <td>1012</td>\n",
       "      <td>164</td>\n",
       "      <td>123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1026</td>\n",
       "      <td>180</td>\n",
       "      <td>135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1037</td>\n",
       "      <td>211</td>\n",
       "      <td>135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1044</td>\n",
       "      <td>208</td>\n",
       "      <td>135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1049</td>\n",
       "      <td>213</td>\n",
       "      <td>157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          death_confirmed death_probable death_suspected etc_cum_deaths  \\\n",
       "Month Day                                                                 \n",
       "10    23             1012            164             123            NaN   \n",
       "      24             1026            180             135            NaN   \n",
       "      25             1037            211             135            NaN   \n",
       "      26             1044            208             135            NaN   \n",
       "      27             1049            213             157            NaN   \n",
       "\n",
       "          etc_new_deaths  \n",
       "Month Day                 \n",
       "10    23             NaN  \n",
       "      24             NaN  \n",
       "      25             NaN  \n",
       "      26             NaN  \n",
       "      27             NaN  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We check for deaths\n",
    "rel_cols = [c for c in cols if ('deaths' in c) or ('death' in c)]\n",
    "df_model_1.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We decide to keep only the confirmed death because we want to keep only deaths that are indeed confirmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_model_1['New Deaths'] = df_model_1['death_confirmed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df = get_non_cumulative(df_model_1,'death_confirmed')\n",
    "df_model_1['New Deaths'] = df['death_confirmed']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we check for the new cases, it sees like they are contained in the columns containing ```new```."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>etc_new_admission</th>\n",
       "      <th>etc_new_deaths</th>\n",
       "      <th>etc_new_discharges</th>\n",
       "      <th>new_completed_contacts</th>\n",
       "      <th>new_confirmed</th>\n",
       "      <th>new_contacts</th>\n",
       "      <th>new_negative</th>\n",
       "      <th>new_noncase</th>\n",
       "      <th>new_positive</th>\n",
       "      <th>new_probable</th>\n",
       "      <th>new_samples</th>\n",
       "      <th>new_suspected</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10</th>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1186</td>\n",
       "      <td>60</td>\n",
       "      <td>397</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>143</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>231</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>41</td>\n",
       "      <td>10</td>\n",
       "      <td>113</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>123</td>\n",
       "      <td>123</td>\n",
       "      <td>70</td>\n",
       "      <td>9</td>\n",
       "      <td>201</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88</td>\n",
       "      <td>62</td>\n",
       "      <td>106</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>96</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>332</td>\n",
       "      <td>91</td>\n",
       "      <td>683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          etc_new_admission etc_new_deaths etc_new_discharges  \\\n",
       "Month Day                                                       \n",
       "10    23                NaN            NaN                NaN   \n",
       "      24                NaN            NaN                NaN   \n",
       "      25                NaN            NaN                NaN   \n",
       "      26                NaN            NaN                NaN   \n",
       "      27                NaN            NaN                NaN   \n",
       "\n",
       "          new_completed_contacts new_confirmed new_contacts new_negative  \\\n",
       "Month Day                                                                  \n",
       "10    23                    1186            60          397           76   \n",
       "      24                      41            41          231           69   \n",
       "      25                     NaN            70          NaN          123   \n",
       "      26                      88            62          106           23   \n",
       "      27                     332            91          683          NaN   \n",
       "\n",
       "          new_noncase new_positive new_probable new_samples new_suspected  \n",
       "Month Day                                                                  \n",
       "10    23           76           60            0         143             7  \n",
       "      24           69           41           10         113             3  \n",
       "      25          123           70            9         201             8  \n",
       "      26           23           62            0          96            11  \n",
       "      27           93          NaN            5         NaN            10  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We check for cases\n",
    "rel_cols = [c for c in cols if ('new' in c)]\n",
    "df_model_1.loc[:,rel_cols].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data is quite dirty here, nevertheless, it seems reasonnable to assume that the confirmed cases are kept in the new_confirmed. We decided to drop the columns starting with etc_ because of their inconsistence depending on the days."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10</th>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>14</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>11</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>7</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>5</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          New Deaths New Cases\n",
       "Month Day                     \n",
       "10    23         NaN        60\n",
       "      24          14        41\n",
       "      25          11        70\n",
       "      26           7        62\n",
       "      27           5        91"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_1['New Cases'] = df_model_1['new_confirmed']\n",
    "df_model_1 = df_model_1[['New Deaths','New Cases']]\n",
    "model_dataframes[1] = df_model_1\n",
    "model_dataframes[1].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <th>Month</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Sierra Leone</th>\n",
       "      <th>8</th>\n",
       "      <td>6.400</td>\n",
       "      <td>20.611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5.120</td>\n",
       "      <td>36.963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>16.615</td>\n",
       "      <td>56.714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>15.015</td>\n",
       "      <td>66.211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>29.875</td>\n",
       "      <td>54.333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    New Deaths  New Cases\n",
       "Country      Month                       \n",
       "Sierra Leone 8           6.400     20.611\n",
       "             9           5.120     36.963\n",
       "             10         16.615     56.714\n",
       "             11         15.015     66.211\n",
       "             12         29.875     54.333"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sl = pd.concat(model_dataframes,axis=0)\n",
    "df_sl['Country'] = 'Sierra Leone'\n",
    "df_sl['New Deaths'] = df_sl['New Deaths'].astype(float)\n",
    "df_sl['New Cases'] = df_sl['New Cases'].astype(float)\n",
    "df_sl = df_sl.reset_index().groupby(['Country','Month']).mean()[['New Deaths','New Cases']].round(3)\n",
    "df_sl.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4 - Result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>New Cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <th>Month</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Guinea</th>\n",
       "      <th>8</th>\n",
       "      <td>3.400</td>\n",
       "      <td>12.400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3.562</td>\n",
       "      <td>13.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>15.000</td>\n",
       "      <td>6.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">Liberia</th>\n",
       "      <th>6</th>\n",
       "      <td>2.000</td>\n",
       "      <td>2.143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4.273</td>\n",
       "      <td>1.818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>23.222</td>\n",
       "      <td>5.444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>37.609</td>\n",
       "      <td>6.167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>28.040</td>\n",
       "      <td>1.619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>15.538</td>\n",
       "      <td>7.800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Sierra Leone</th>\n",
       "      <th>8</th>\n",
       "      <td>6.400</td>\n",
       "      <td>20.611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5.120</td>\n",
       "      <td>36.963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>16.615</td>\n",
       "      <td>56.714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>15.015</td>\n",
       "      <td>66.211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>29.875</td>\n",
       "      <td>54.333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    New Deaths  New Cases\n",
       "Country      Month                       \n",
       "Guinea       8           3.400     12.400\n",
       "             9           3.562     13.000\n",
       "             10         15.000      6.000\n",
       "Liberia      6           2.000      2.143\n",
       "             7           4.273      1.818\n",
       "             8          23.222      5.444\n",
       "             9          37.609      6.167\n",
       "             10         28.040      1.619\n",
       "             11         15.538      7.800\n",
       "             12            NaN      5.000\n",
       "Sierra Leone 8           6.400     20.611\n",
       "             9           5.120     36.963\n",
       "             10         16.615     56.714\n",
       "             11         15.015     66.211\n",
       "             12         29.875     54.333"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = pd.concat([df_guinea,df_liberia,df_sl],axis=0)\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1- Importing the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we import the data, and we add the ```BARCODE``` column using the file name to be able to match the metadata after that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon</th>\n",
       "      <th>count</th>\n",
       "      <th>BARCODE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>7</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...</td>\n",
       "      <td>5</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               taxon  count BARCODE\n",
       "0  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      7    MID6\n",
       "1  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      1    MID6\n",
       "2  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      1    MID6\n",
       "3  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      2    MID6\n",
       "4  Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...      5    MID6"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MICROBIOME_FOLDER = DATA_FOLDER + '/microbiome'\n",
    "\n",
    "path = MICROBIOME_FOLDER\n",
    "allFiles = glob.glob(path + \"/MID[1-9].xls\")\n",
    "microbiome = pd.DataFrame()\n",
    "list_ = []\n",
    "for file_ in allFiles:\n",
    "    df = pd.read_excel(file_,header=None)\n",
    "    df['BARCODE'] = file_[18:22]\n",
    "    list_.append(df)\n",
    "microbiome = pd.concat(list_,axis=0)\n",
    "microbiome.columns =['taxon','count','BARCODE']\n",
    "microbiome.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we check wether there is any NaN in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "microbiome.isnull().values.any()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we import the metadata and change the only NA as ```unknown```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MID3</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MID4</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MID5</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MID6</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MID7</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MID8</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  BARCODE               GROUP   SAMPLE\n",
       "0    MID1  EXTRACTION CONTROL  unknown\n",
       "1    MID2               NEC 1   tissue\n",
       "2    MID3           Control 1   tissue\n",
       "3    MID4               NEC 2   tissue\n",
       "4    MID5           Control 2   tissue\n",
       "5    MID6               NEC 1    stool\n",
       "6    MID7           Control 1    stool\n",
       "7    MID8               NEC 2    stool\n",
       "8    MID9           Control 2    stool"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "METADATA_PATH = MICROBIOME_FOLDER + '/metadata.xls'\n",
    "metadata = pd.read_excel(METADATA_PATH,header=0)\n",
    "metadata.fillna(value='unknown',inplace=True)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2 - Merging and keeping a unique index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we need to merge the two tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon</th>\n",
       "      <th>count</th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>7</td>\n",
       "      <td>MID6</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID6</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID6</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID6</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...</td>\n",
       "      <td>5</td>\n",
       "      <td>MID6</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               taxon  count BARCODE  GROUP  \\\n",
       "0  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      7    MID6  NEC 1   \n",
       "1  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      1    MID6  NEC 1   \n",
       "2  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      1    MID6  NEC 1   \n",
       "3  Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      2    MID6  NEC 1   \n",
       "4  Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...      5    MID6  NEC 1   \n",
       "\n",
       "  SAMPLE  \n",
       "0  stool  \n",
       "1  stool  \n",
       "2  stool  \n",
       "3  stool  \n",
       "4  stool  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = pd.merge(microbiome,metadata)\n",
    "result.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As explained in ```Intro to Pandas II``` : \"Lets import two microbiome datasets, each consisting of counts of micro-/organisms from a particular patient\", but we can see that the metadata of each ```MID[1-9]``` are different, therefore, the count of taxon for each barcode has been sampled uing differen```(GROUP,SAMPLE)``` pair. Therefore, we can index the dataframe using ```['taxon','GROUP','SAMPLE']``` to then unstack ```GROUP and SAMPLE``` to have for each ```(GROUP,SAMPLE)``` pair (described on the y axis) the count of the taxon (on the x axis)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>BARCODE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>taxon</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <th>NEC 1</th>\n",
       "      <th>stool</th>\n",
       "      <td>7</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Sulfophobococcus</th>\n",
       "      <th>NEC 1</th>\n",
       "      <th>stool</th>\n",
       "      <td>1</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrodictium</th>\n",
       "      <th>NEC 1</th>\n",
       "      <th>stool</th>\n",
       "      <td>1</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <th>NEC 1</th>\n",
       "      <th>stool</th>\n",
       "      <td>2</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <th>NEC 1</th>\n",
       "      <th>stool</th>\n",
       "      <td>5</td>\n",
       "      <td>MID6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                 count BARCODE\n",
       "taxon                                              GROUP SAMPLE               \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... NEC 1 stool       7    MID6\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... NEC 1 stool       1    MID6\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... NEC 1 stool       1    MID6\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... NEC 1 stool       2    MID6\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal... NEC 1 stool       5    MID6"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = result.set_index(['taxon','GROUP','SAMPLE'])\n",
    "result.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>GROUP</th>\n",
       "      <th colspan=\"3\" halign=\"left\">Control 1</th>\n",
       "      <th colspan=\"3\" halign=\"left\">Control 2</th>\n",
       "      <th colspan=\"3\" halign=\"left\">EXTRACTION CONTROL</th>\n",
       "      <th colspan=\"3\" halign=\"left\">NEC 1</th>\n",
       "      <th colspan=\"3\" halign=\"left\">NEC 2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>stool</th>\n",
       "      <th>tissue</th>\n",
       "      <th>unknown</th>\n",
       "      <th>stool</th>\n",
       "      <th>tissue</th>\n",
       "      <th>unknown</th>\n",
       "      <th>stool</th>\n",
       "      <th>tissue</th>\n",
       "      <th>unknown</th>\n",
       "      <th>stool</th>\n",
       "      <th>tissue</th>\n",
       "      <th>unknown</th>\n",
       "      <th>stool</th>\n",
       "      <th>tissue</th>\n",
       "      <th>unknown</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Acidilobaceae Acidilobus</th>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Caldisphaeraceae Caldisphaera</th>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>14</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "      <td>unknown</td>\n",
       "      <td>16</td>\n",
       "      <td>28</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>23</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Stetteria</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Sulfophobococcus</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Thermodiscus</th>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Thermosphaera</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Hyperthermus</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrodictium</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>6</td>\n",
       "      <td>14</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Sulfurisphaera</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermoproteaceae Caldivirga</th>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermoproteaceae Thermocladium</th>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>6</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>18</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>14</td>\n",
       "      <td>17</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanomicrobiales Methanomicrobiaceae Methanomicrobium</th>\n",
       "      <td>unknown</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanomicrobiales Methanomicrobiaceae Methanosphaerula</th>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanomicrobiales Methanomicrobiales_incertae_sedis Methanolinea</th>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanomicrobiales Methanospirillaceae Methanospirillum</th>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methanosarcinaceae Methanimicrococcus</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methanosarcinaceae Methanosalsum</th>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methermicoccaceae Methermicoccus</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Ferroglobus</th>\n",
       "      <td>9</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Geoglobus</th>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haladaptatus</th>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Halogeometricum</th>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Halopiger</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloplanus</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloquadratum</th>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Rudaea</th>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Silanimonas</th>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Stenotrophomonas</th>\n",
       "      <td>80</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Wohlfahrtiimonas</th>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Xanthomonas</th>\n",
       "      <td>133</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Xylella</th>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Spirochaetes\" Spirochaetes Spirochaetales Spirochaetales_incertae_sedis Exilispira</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Aminiphilus</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Aminomonas</th>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Thermovirga</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Anaeroplasmatales Anaeroplasmataceae Asteroleplasma</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Haloplasmatales Haloplasmataceae Haloplasma</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>11</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Caldimicrobium</th>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Thermodesulfatator</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Geotoga</th>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>43</td>\n",
       "      <td>26</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4</td>\n",
       "      <td>13</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Kosmotoga</th>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>16</td>\n",
       "      <td>36</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>22</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Opitutales Opitutaceae Alterococcus</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Opitutales Opitutaceae Opitutus</th>\n",
       "      <td>126</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Puniceicoccales Puniceicoccaceae Cerasicoccus</th>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Puniceicoccales Puniceicoccaceae Coraliomargarita</th>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Puniceicoccales Puniceicoccaceae Puniceicoccus</th>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Subdivision3   Subdivision3_genera_incertae_sedis</th>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Verrucomicrobiae Verrucomicrobiales Verrucomicrobiaceae Persicirhabdus</th>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Verrucomicrobiae Verrucomicrobiales Verrucomicrobiaceae Verrucomicrobium</th>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Bangiophyceae</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Chlorarachniophyceae</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>85</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Streptophyta</th>\n",
       "      <td>unknown</td>\n",
       "      <td>54</td>\n",
       "      <td>unknown</td>\n",
       "      <td>10</td>\n",
       "      <td>59</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1388</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>44</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria OD1    OD1_genera_incertae_sedis</th>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria OP11    OP11_genera_incertae_sedis</th>\n",
       "      <td>unknown</td>\n",
       "      <td>26</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria TM7    TM7_genera_incertae_sedis</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>775 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GROUP                                              Control 1           \\\n",
       "SAMPLE                                                 stool   tissue   \n",
       "taxon                                                                   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...   unknown        1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...         1       15   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...         8       14   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...   unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...         2        4   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...   unknown        1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...         1        1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...   unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...         1        3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...         1  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...         5        4   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...         3        2   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...         5        5   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...   unknown        3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...   unknown        1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        18        7   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...   unknown        4   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...   unknown        1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...         1  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...         1  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...         2        2   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...   unknown        1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...         2        1   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...         9  unknown   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...         2  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...   unknown        3   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...   unknown        1   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...   unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...         3        3   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...   unknown        2   \n",
       "...                                                      ...      ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...         5  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...   unknown        1   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...        80  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...   unknown        1   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...       133  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...         1  unknown   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...   unknown  unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...         1        2   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...         8        3   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...   unknown  unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...   unknown  unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...         2        2   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...         1  unknown   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...         3        1   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...        15       12   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...        16        7   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...         2        1   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...       126        1   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...        39        1   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...         2  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...   unknown        2   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...         4  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...         9        2   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...        39        1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...         1        1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...         2        4   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...   unknown       54   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis            unknown        1   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis          unknown       26   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis            unknown  unknown   \n",
       "\n",
       "GROUP                                                       Control 2  \\\n",
       "SAMPLE                                              unknown     stool   \n",
       "taxon                                                                   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  unknown   unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  unknown         1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown        16   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown   unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown         2   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown   unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown   unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown   unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown         5   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown   unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  unknown         6   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  unknown         2   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown         5   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown         3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown   unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown        14   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown         1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown         2   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown         1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown   unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown         6   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown   unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown         3   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown         3   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown         3   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown   unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown   unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown   unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown         4   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown         1   \n",
       "...                                                     ...       ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown   unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown   unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown   unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown   unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown   unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown   unknown   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...  unknown   unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  unknown   unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  unknown         3   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  unknown   unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...  unknown         1   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...  unknown        11   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  unknown         1   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  unknown         1   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  unknown        43   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  unknown        16   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  unknown   unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  unknown         1   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown   unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown   unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown   unknown   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...  unknown   unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown   unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown   unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  unknown   unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  unknown         1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  unknown        10   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis           unknown   unknown   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis         unknown   unknown   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis           unknown   unknown   \n",
       "\n",
       "GROUP                                                                 \\\n",
       "SAMPLE                                               tissue  unknown   \n",
       "taxon                                                                  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...        5  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...       26  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...       28  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        1  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        5  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        2  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        2  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        3  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...       14  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...        5  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...       10  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...       18  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...        6  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...       17  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        7  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        1  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...       11  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        2  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        3  unknown   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...        8  unknown   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...        1  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        1  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        1  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        1  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        7  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        1  unknown   \n",
       "...                                                     ...      ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...  unknown  unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...        5  unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...        5  unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  unknown  unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...  unknown  unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...  unknown  unknown   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...        1  unknown   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...        1  unknown   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...       26  unknown   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...       36  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...        1  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...        2  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...        4  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown  unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...        2  unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...        5  unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...       59  unknown   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis                 4  unknown   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis         unknown  unknown   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis           unknown  unknown   \n",
       "\n",
       "GROUP                                              EXTRACTION CONTROL  \\\n",
       "SAMPLE                                                          stool   \n",
       "taxon                                                                   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...            unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...            unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...            unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...            unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...            unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...            unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...            unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...            unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...            unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...            unknown   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...            unknown   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...            unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...            unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...            unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...            unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...            unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...            unknown   \n",
       "...                                                               ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...            unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...            unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...            unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...            unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...            unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...            unknown   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...            unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...            unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...            unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...            unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...            unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...            unknown   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...            unknown   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...            unknown   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...            unknown   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...            unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...            unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...            unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...            unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...            unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...            unknown   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...            unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...            unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...            unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...            unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...            unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...            unknown   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis                     unknown   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis                   unknown   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis                     unknown   \n",
       "\n",
       "GROUP                                                                   NEC 1  \\\n",
       "SAMPLE                                               tissue  unknown    stool   \n",
       "taxon                                                                           \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  unknown  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  unknown  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown        7        7   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown        1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown        1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown        2        2   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  unknown        3        5   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  unknown  unknown        4   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown        3        4   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown  unknown        1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown        7       12   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown        2   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown        1        1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown        1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown        1        1   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown        1        4   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown        1  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown        1        4   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown        4  unknown   \n",
       "...                                                     ...      ...      ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...  unknown        1        1   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  unknown        2        2   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  unknown        1  unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  unknown        1  unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...  unknown  unknown  unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...  unknown        1       18   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  unknown        1        1   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  unknown  unknown        2   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  unknown        7       12   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  unknown        9       11   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  unknown        1  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown  unknown        1   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown  unknown        1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  unknown        2  unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  unknown       85  unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  unknown     1388  unknown   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis           unknown  unknown  unknown   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis         unknown  unknown  unknown   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis           unknown  unknown  unknown   \n",
       "\n",
       "GROUP                                                                   NEC 2  \\\n",
       "SAMPLE                                               tissue  unknown    stool   \n",
       "taxon                                                                           \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...        2  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...       14  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...       23  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        1  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        2  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        1  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        2  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...       10  unknown        1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...       11  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...        9  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...        5  unknown  unknown   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...        3  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        9  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        1  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...       12  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        2  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...        4  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        3  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        1  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        4  unknown  unknown   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        3  unknown  unknown   \n",
       "...                                                     ...      ...      ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  unknown   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...  unknown  unknown  unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...        2  unknown  unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...        2  unknown  unknown   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  unknown  unknown  unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...  unknown  unknown  unknown   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...       15  unknown  unknown   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...        1  unknown  unknown   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...        3  unknown        1   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...       15  unknown        4   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...       22  unknown        2   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown  unknown  unknown   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown  unknown  unknown   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  unknown  unknown        1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...        1  unknown        4   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...        2  unknown  unknown   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis           unknown  unknown  unknown   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis         unknown  unknown  unknown   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis                 2  unknown  unknown   \n",
       "\n",
       "GROUP                                                                 \n",
       "SAMPLE                                               tissue  unknown  \n",
       "taxon                                                                 \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        2  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...        1  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown  unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        1  unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...        1  unknown  \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  unknown  \n",
       "...                                                     ...      ...  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  unknown  unknown  \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...  unknown  unknown  \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...        1  unknown  \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...        1  unknown  \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  unknown  unknown  \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...  unknown  unknown  \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...  unknown  unknown  \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  unknown  unknown  \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  unknown  unknown  \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...       13  unknown  \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...        4  unknown  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  unknown  unknown  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  unknown  unknown  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  unknown  unknown  \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...  unknown  unknown  \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown  unknown  \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  unknown  unknown  \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  unknown  unknown  \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...        6  unknown  \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...       44  unknown  \n",
       "Bacteria OD1    OD1_genera_incertae_sedis           unknown  unknown  \n",
       "Bacteria OP11    OP11_genera_incertae_sedis         unknown  unknown  \n",
       "Bacteria TM7    TM7_genera_incertae_sedis           unknown  unknown  \n",
       "\n",
       "[775 rows x 15 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result['count'].unstack('GROUP').unstack('SAMPLE').fillna('unknown')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that the indices are unique :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The indices are unique : True\n"
     ]
    }
   ],
   "source": [
    "print(\"The indices are unique : {}\".format(result.index.is_unique))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<hr><h2>Data frame:titanic3</h2>1309 observations and 14 variables, maximum # NAs:1188<hr>\n",
       "<TABLE BORDER>\n",
       "<tr><td>Name</td><td>Labels</td><td>Units</td><td>Levels</td><td>Storage</td><td>NAs</td></tr>\n",
       "<tr><td>pclass</td><td></td><td></td><td><a href=\"#pclass\"><div align=right>  3</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>survived</td><td>Survived</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>name</td><td>Name</td><td></td><td><div align=right></div></td><td>character</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>sex</td><td></td><td></td><td><a href=\"#sex\"><div align=right>  2</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>age</td><td>Age</td><td>Year</td><td><div align=right></div></td><td>double</td><td><div align=right> 263</div></td></tr>\n",
       "<tr><td>sibsp</td><td>Number of Siblings/Spouses Aboard</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>parch</td><td>Number of Parents/Children Aboard</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>ticket</td><td>Ticket Number</td><td></td><td><div align=right></div></td><td>character</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>fare</td><td>Passenger Fare</td><td>British Pound (\\243)</td><td><div align=right></div></td><td>double</td><td><div align=right>   1</div></td></tr>\n",
       "<tr><td>cabin</td><td></td><td></td><td><a href=\"#cabin\"><div align=right>187</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>embarked</td><td></td><td></td><td><a href=\"#embarked\"><div align=right>  3</div></a></td><td>integer</td><td><div align=right>   2</div></td></tr>\n",
       "<tr><td>boat</td><td></td><td></td><td><a href=\"#boat\"><div align=right> 28</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>body</td><td>Body Identification Number</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>1188</div></td></tr>\n",
       "<tr><td>home.dest</td><td>Home/Destination</td><td></td><td><div align=right></div></td><td>character</td><td><div align=right>   0</div></td></tr>\n",
       "</TABLE>\n",
       "<hr>\n",
       "<TABLE BORDER>\n",
       "<tr><td>Variable</td><td>Levels</td></tr>\n",
       "<tr><td><a name=\"pclass\">pclass</a></td><td>1st</td></tr>\n",
       "<tr><td></td><td>2nd</td></tr>\n",
       "<tr><td></td><td>3rd</td></tr>\n",
       "<tr><td><a name=\"sex\">sex</a></td><td>female</td></tr>\n",
       "<tr><td></td><td>male</td></tr>\n",
       "<tr><td><a name=\"cabin\">cabin</a></td><td></td></tr>\n",
       "<tr><td></td><td>A10</td></tr>\n",
       "<tr><td></td><td>A11</td></tr>\n",
       "<tr><td></td><td>A14</td></tr>\n",
       "<tr><td></td><td>A16</td></tr>\n",
       "<tr><td></td><td>A18</td></tr>\n",
       "<tr><td></td><td>A19</td></tr>\n",
       "<tr><td></td><td>A20</td></tr>\n",
       "<tr><td></td><td>A21</td></tr>\n",
       "<tr><td></td><td>A23</td></tr>\n",
       "<tr><td></td><td>A24</td></tr>\n",
       "<tr><td></td><td>A26</td></tr>\n",
       "<tr><td></td><td>A29</td></tr>\n",
       "<tr><td></td><td>A31</td></tr>\n",
       "<tr><td></td><td>A32</td></tr>\n",
       "<tr><td></td><td>A34</td></tr>\n",
       "<tr><td></td><td>A36</td></tr>\n",
       "<tr><td></td><td>A5</td></tr>\n",
       "<tr><td></td><td>A6</td></tr>\n",
       "<tr><td></td><td>A7</td></tr>\n",
       "<tr><td></td><td>A9</td></tr>\n",
       "<tr><td></td><td>B10</td></tr>\n",
       "<tr><td></td><td>B101</td></tr>\n",
       "<tr><td></td><td>B102</td></tr>\n",
       "<tr><td></td><td>B11</td></tr>\n",
       "<tr><td></td><td>B18</td></tr>\n",
       "<tr><td></td><td>B19</td></tr>\n",
       "<tr><td></td><td>B20</td></tr>\n",
       "<tr><td></td><td>B22</td></tr>\n",
       "<tr><td></td><td>B24</td></tr>\n",
       "<tr><td></td><td>B26</td></tr>\n",
       "<tr><td></td><td>B28</td></tr>\n",
       "<tr><td></td><td>B3</td></tr>\n",
       "<tr><td></td><td>B30</td></tr>\n",
       "<tr><td></td><td>B35</td></tr>\n",
       "<tr><td></td><td>B36</td></tr>\n",
       "<tr><td></td><td>B37</td></tr>\n",
       "<tr><td></td><td>B38</td></tr>\n",
       "<tr><td></td><td>B39</td></tr>\n",
       "<tr><td></td><td>B4</td></tr>\n",
       "<tr><td></td><td>B41</td></tr>\n",
       "<tr><td></td><td>B42</td></tr>\n",
       "<tr><td></td><td>B45</td></tr>\n",
       "<tr><td></td><td>B49</td></tr>\n",
       "<tr><td></td><td>B5</td></tr>\n",
       "<tr><td></td><td>B50</td></tr>\n",
       "<tr><td></td><td>B51 B53 B55</td></tr>\n",
       "<tr><td></td><td>B52 B54 B56</td></tr>\n",
       "<tr><td></td><td>B57 B59 B63 B66</td></tr>\n",
       "<tr><td></td><td>B58 B60</td></tr>\n",
       "<tr><td></td><td>B61</td></tr>\n",
       "<tr><td></td><td>B69</td></tr>\n",
       "<tr><td></td><td>B71</td></tr>\n",
       "<tr><td></td><td>B73</td></tr>\n",
       "<tr><td></td><td>B77</td></tr>\n",
       "<tr><td></td><td>B78</td></tr>\n",
       "<tr><td></td><td>B79</td></tr>\n",
       "<tr><td></td><td>B80</td></tr>\n",
       "<tr><td></td><td>B82 B84</td></tr>\n",
       "<tr><td></td><td>B86</td></tr>\n",
       "<tr><td></td><td>B94</td></tr>\n",
       "<tr><td></td><td>B96 B98</td></tr>\n",
       "<tr><td></td><td>C101</td></tr>\n",
       "<tr><td></td><td>C103</td></tr>\n",
       "<tr><td></td><td>C104</td></tr>\n",
       "<tr><td></td><td>C105</td></tr>\n",
       "<tr><td></td><td>C106</td></tr>\n",
       "<tr><td></td><td>C110</td></tr>\n",
       "<tr><td></td><td>C111</td></tr>\n",
       "<tr><td></td><td>C116</td></tr>\n",
       "<tr><td></td><td>C118</td></tr>\n",
       "<tr><td></td><td>C123</td></tr>\n",
       "<tr><td></td><td>C124</td></tr>\n",
       "<tr><td></td><td>C125</td></tr>\n",
       "<tr><td></td><td>C126</td></tr>\n",
       "<tr><td></td><td>C128</td></tr>\n",
       "<tr><td></td><td>C130</td></tr>\n",
       "<tr><td></td><td>C132</td></tr>\n",
       "<tr><td></td><td>C148</td></tr>\n",
       "<tr><td></td><td>C2</td></tr>\n",
       "<tr><td></td><td>C22 C26</td></tr>\n",
       "<tr><td></td><td>C23 C25 C27</td></tr>\n",
       "<tr><td></td><td>C28</td></tr>\n",
       "<tr><td></td><td>C30</td></tr>\n",
       "<tr><td></td><td>C31</td></tr>\n",
       "<tr><td></td><td>C32</td></tr>\n",
       "<tr><td></td><td>C39</td></tr>\n",
       "<tr><td></td><td>C45</td></tr>\n",
       "<tr><td></td><td>C46</td></tr>\n",
       "<tr><td></td><td>C47</td></tr>\n",
       "<tr><td></td><td>C49</td></tr>\n",
       "<tr><td></td><td>C50</td></tr>\n",
       "<tr><td></td><td>C51</td></tr>\n",
       "<tr><td></td><td>C52</td></tr>\n",
       "<tr><td></td><td>C53</td></tr>\n",
       "<tr><td></td><td>C54</td></tr>\n",
       "<tr><td></td><td>C55 C57</td></tr>\n",
       "<tr><td></td><td>C6</td></tr>\n",
       "<tr><td></td><td>C62 C64</td></tr>\n",
       "<tr><td></td><td>C65</td></tr>\n",
       "<tr><td></td><td>C68</td></tr>\n",
       "<tr><td></td><td>C7</td></tr>\n",
       "<tr><td></td><td>C70</td></tr>\n",
       "<tr><td></td><td>C78</td></tr>\n",
       "<tr><td></td><td>C80</td></tr>\n",
       "<tr><td></td><td>C82</td></tr>\n",
       "<tr><td></td><td>C83</td></tr>\n",
       "<tr><td></td><td>C85</td></tr>\n",
       "<tr><td></td><td>C86</td></tr>\n",
       "<tr><td></td><td>C87</td></tr>\n",
       "<tr><td></td><td>C89</td></tr>\n",
       "<tr><td></td><td>C90</td></tr>\n",
       "<tr><td></td><td>C91</td></tr>\n",
       "<tr><td></td><td>C92</td></tr>\n",
       "<tr><td></td><td>C93</td></tr>\n",
       "<tr><td></td><td>C95</td></tr>\n",
       "<tr><td></td><td>C97</td></tr>\n",
       "<tr><td></td><td>C99</td></tr>\n",
       "<tr><td></td><td>D</td></tr>\n",
       "<tr><td></td><td>D10 D12</td></tr>\n",
       "<tr><td></td><td>D11</td></tr>\n",
       "<tr><td></td><td>D15</td></tr>\n",
       "<tr><td></td><td>D17</td></tr>\n",
       "<tr><td></td><td>D19</td></tr>\n",
       "<tr><td></td><td>D20</td></tr>\n",
       "<tr><td></td><td>D21</td></tr>\n",
       "<tr><td></td><td>D22</td></tr>\n",
       "<tr><td></td><td>D26</td></tr>\n",
       "<tr><td></td><td>D28</td></tr>\n",
       "<tr><td></td><td>D30</td></tr>\n",
       "<tr><td></td><td>D33</td></tr>\n",
       "<tr><td></td><td>D34</td></tr>\n",
       "<tr><td></td><td>D35</td></tr>\n",
       "<tr><td></td><td>D36</td></tr>\n",
       "<tr><td></td><td>D37</td></tr>\n",
       "<tr><td></td><td>D38</td></tr>\n",
       "<tr><td></td><td>D40</td></tr>\n",
       "<tr><td></td><td>D43</td></tr>\n",
       "<tr><td></td><td>D45</td></tr>\n",
       "<tr><td></td><td>D46</td></tr>\n",
       "<tr><td></td><td>D47</td></tr>\n",
       "<tr><td></td><td>D48</td></tr>\n",
       "<tr><td></td><td>D49</td></tr>\n",
       "<tr><td></td><td>D50</td></tr>\n",
       "<tr><td></td><td>D56</td></tr>\n",
       "<tr><td></td><td>D6</td></tr>\n",
       "<tr><td></td><td>D7</td></tr>\n",
       "<tr><td></td><td>D9</td></tr>\n",
       "<tr><td></td><td>E10</td></tr>\n",
       "<tr><td></td><td>E101</td></tr>\n",
       "<tr><td></td><td>E12</td></tr>\n",
       "<tr><td></td><td>E121</td></tr>\n",
       "<tr><td></td><td>E17</td></tr>\n",
       "<tr><td></td><td>E24</td></tr>\n",
       "<tr><td></td><td>E25</td></tr>\n",
       "<tr><td></td><td>E31</td></tr>\n",
       "<tr><td></td><td>E33</td></tr>\n",
       "<tr><td></td><td>E34</td></tr>\n",
       "<tr><td></td><td>E36</td></tr>\n",
       "<tr><td></td><td>E38</td></tr>\n",
       "<tr><td></td><td>E39 E41</td></tr>\n",
       "<tr><td></td><td>E40</td></tr>\n",
       "<tr><td></td><td>E44</td></tr>\n",
       "<tr><td></td><td>E45</td></tr>\n",
       "<tr><td></td><td>E46</td></tr>\n",
       "<tr><td></td><td>E49</td></tr>\n",
       "<tr><td></td><td>E50</td></tr>\n",
       "<tr><td></td><td>E52</td></tr>\n",
       "<tr><td></td><td>E58</td></tr>\n",
       "<tr><td></td><td>E60</td></tr>\n",
       "<tr><td></td><td>E63</td></tr>\n",
       "<tr><td></td><td>E67</td></tr>\n",
       "<tr><td></td><td>E68</td></tr>\n",
       "<tr><td></td><td>E77</td></tr>\n",
       "<tr><td></td><td>E8</td></tr>\n",
       "<tr><td></td><td>F</td></tr>\n",
       "<tr><td></td><td>F E46</td></tr>\n",
       "<tr><td></td><td>F E57</td></tr>\n",
       "<tr><td></td><td>F E69</td></tr>\n",
       "<tr><td></td><td>F G63</td></tr>\n",
       "<tr><td></td><td>F G73</td></tr>\n",
       "<tr><td></td><td>F2</td></tr>\n",
       "<tr><td></td><td>F33</td></tr>\n",
       "<tr><td></td><td>F38</td></tr>\n",
       "<tr><td></td><td>F4</td></tr>\n",
       "<tr><td></td><td>G6</td></tr>\n",
       "<tr><td></td><td>T</td></tr>\n",
       "<tr><td><a name=\"embarked\">embarked</a></td><td>Cherbourg</td></tr>\n",
       "<tr><td></td><td>Queenstown</td></tr>\n",
       "<tr><td></td><td>Southampton</td></tr>\n",
       "<tr><td><a name=\"boat\">boat</a></td><td></td></tr>\n",
       "<tr><td></td><td>1</td></tr>\n",
       "<tr><td></td><td>10</td></tr>\n",
       "<tr><td></td><td>11</td></tr>\n",
       "<tr><td></td><td>12</td></tr>\n",
       "<tr><td></td><td>13</td></tr>\n",
       "<tr><td></td><td>13 15</td></tr>\n",
       "<tr><td></td><td>13 15 B</td></tr>\n",
       "<tr><td></td><td>14</td></tr>\n",
       "<tr><td></td><td>15</td></tr>\n",
       "<tr><td></td><td>15 16</td></tr>\n",
       "<tr><td></td><td>16</td></tr>\n",
       "<tr><td></td><td>2</td></tr>\n",
       "<tr><td></td><td>3</td></tr>\n",
       "<tr><td></td><td>4</td></tr>\n",
       "<tr><td></td><td>5</td></tr>\n",
       "<tr><td></td><td>5 7</td></tr>\n",
       "<tr><td></td><td>5 9</td></tr>\n",
       "<tr><td></td><td>6</td></tr>\n",
       "<tr><td></td><td>7</td></tr>\n",
       "<tr><td></td><td>8</td></tr>\n",
       "<tr><td></td><td>8 10</td></tr>\n",
       "<tr><td></td><td>9</td></tr>\n",
       "<tr><td></td><td>A</td></tr>\n",
       "<tr><td></td><td>B</td></tr>\n",
       "<tr><td></td><td>C</td></tr>\n",
       "<tr><td></td><td>C D</td></tr>\n",
       "<tr><td></td><td>D</td></tr>\n",
       "</TABLE>\n",
       "<hr>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "titanic = pd.read_excel(DATA_FOLDER+'/titanic.xls',header=0,converters={'ticket':str,'boat':str,'cabin':str,'home.dest':str})\n",
    "titanic.head()\n",
    "\n",
    "#We replace the embarcation port by their full name\n",
    "treatment_map = {'S': 'Southampton', 'Q': 'Queenstown', 'C': 'Cherbourg'}\n",
    "titanic['embarked'] = titanic['embarked'].map(treatment_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pclass</th>\n",
       "      <th>survived</th>\n",
       "      <th>name</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>sibsp</th>\n",
       "      <th>parch</th>\n",
       "      <th>ticket</th>\n",
       "      <th>fare</th>\n",
       "      <th>cabin</th>\n",
       "      <th>embarked</th>\n",
       "      <th>boat</th>\n",
       "      <th>body</th>\n",
       "      <th>home.dest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Allen, Miss. Elisabeth Walton</td>\n",
       "      <td>female</td>\n",
       "      <td>29.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24160</td>\n",
       "      <td>211.3375</td>\n",
       "      <td>B5</td>\n",
       "      <td>Southampton</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>St Louis, MO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Allison, Master. Hudson Trevor</td>\n",
       "      <td>male</td>\n",
       "      <td>0.9167</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>113781</td>\n",
       "      <td>151.5500</td>\n",
       "      <td>C22 C26</td>\n",
       "      <td>Southampton</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Montreal, PQ / Chesterville, ON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Allison, Miss. Helen Loraine</td>\n",
       "      <td>female</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>113781</td>\n",
       "      <td>151.5500</td>\n",
       "      <td>C22 C26</td>\n",
       "      <td>Southampton</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Montreal, PQ / Chesterville, ON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Allison, Mr. Hudson Joshua Creighton</td>\n",
       "      <td>male</td>\n",
       "      <td>30.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>113781</td>\n",
       "      <td>151.5500</td>\n",
       "      <td>C22 C26</td>\n",
       "      <td>Southampton</td>\n",
       "      <td>NaN</td>\n",
       "      <td>135.0</td>\n",
       "      <td>Montreal, PQ / Chesterville, ON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Allison, Mrs. Hudson J C (Bessie Waldo Daniels)</td>\n",
       "      <td>female</td>\n",
       "      <td>25.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>113781</td>\n",
       "      <td>151.5500</td>\n",
       "      <td>C22 C26</td>\n",
       "      <td>Southampton</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Montreal, PQ / Chesterville, ON</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pclass  survived                                             name     sex  \\\n",
       "0       1         1                    Allen, Miss. Elisabeth Walton  female   \n",
       "1       1         1                   Allison, Master. Hudson Trevor    male   \n",
       "2       1         0                     Allison, Miss. Helen Loraine  female   \n",
       "3       1         0             Allison, Mr. Hudson Joshua Creighton    male   \n",
       "4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   \n",
       "\n",
       "       age  sibsp  parch  ticket      fare    cabin     embarked boat   body  \\\n",
       "0  29.0000      0      0   24160  211.3375       B5  Southampton    2    NaN   \n",
       "1   0.9167      1      2  113781  151.5500  C22 C26  Southampton   11    NaN   \n",
       "2   2.0000      1      2  113781  151.5500  C22 C26  Southampton  NaN    NaN   \n",
       "3  30.0000      1      2  113781  151.5500  C22 C26  Southampton  NaN  135.0   \n",
       "4  25.0000      1      2  113781  151.5500  C22 C26  Southampton  NaN    NaN   \n",
       "\n",
       "                         home.dest  \n",
       "0                     St Louis, MO  \n",
       "1  Montreal, PQ / Chesterville, ON  \n",
       "2  Montreal, PQ / Chesterville, ON  \n",
       "3  Montreal, PQ / Chesterville, ON  \n",
       "4  Montreal, PQ / Chesterville, ON  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "titanic.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Question 1: We decided to represent the answer with a dataframe to make the visualization easier**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pclass</th>\n",
       "      <td>int</td>\n",
       "      <td>'1' to '3'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>survived</th>\n",
       "      <td>int</td>\n",
       "      <td>'0' to '1'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <td>str</td>\n",
       "      <td>'Abbing, Mr. Anthony' to 'van Melkebeke, Mr. P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sex</th>\n",
       "      <td>str</td>\n",
       "      <td>'female' to 'male'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>age</th>\n",
       "      <td>float</td>\n",
       "      <td>'0.1667' to '80.0'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sibsp</th>\n",
       "      <td>int</td>\n",
       "      <td>'0' to '8'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>parch</th>\n",
       "      <td>int</td>\n",
       "      <td>'0' to '9'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ticket</th>\n",
       "      <td>str</td>\n",
       "      <td>'110152' to 'WE/P 5735'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fare</th>\n",
       "      <td>float</td>\n",
       "      <td>'0.0' to '512.3292'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cabin</th>\n",
       "      <td>str</td>\n",
       "      <td>'A10' to 'T'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>embarked</th>\n",
       "      <td>str</td>\n",
       "      <td>'Cherbourg' to 'Southampton'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>boat</th>\n",
       "      <td>str</td>\n",
       "      <td>'1' to 'D'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>body</th>\n",
       "      <td>float</td>\n",
       "      <td>'1.0' to '328.0'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>home.dest</th>\n",
       "      <td>str</td>\n",
       "      <td>'?Havana, Cuba' to 'Zurich, Switzerland'</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            type                                              range\n",
       "pclass       int                                         '1' to '3'\n",
       "survived     int                                         '0' to '1'\n",
       "name         str  'Abbing, Mr. Anthony' to 'van Melkebeke, Mr. P...\n",
       "sex          str                                 'female' to 'male'\n",
       "age        float                                 '0.1667' to '80.0'\n",
       "sibsp        int                                         '0' to '8'\n",
       "parch        int                                         '0' to '9'\n",
       "ticket       str                            '110152' to 'WE/P 5735'\n",
       "fare       float                                '0.0' to '512.3292'\n",
       "cabin        str                                       'A10' to 'T'\n",
       "embarked     str                       'Cherbourg' to 'Southampton'\n",
       "boat         str                                         '1' to 'D'\n",
       "body       float                                   '1.0' to '328.0'\n",
       "home.dest    str           '?Havana, Cuba' to 'Zurich, Switzerland'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranges = []\n",
    "for i,var in enumerate(titanic.columns):\n",
    "    ranges.append(\"'{}' to '{}'\".format(titanic.iloc[:, i].dropna().min(),titanic.iloc[:, i].dropna().max()))\n",
    "\n",
    "df = titanic.transpose()\n",
    "df['type'] = df[0].map(lambda x: type(x).__name__)\n",
    "df['range'] = ranges\n",
    "df[['type','range']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Following the documentation we see that the values that can be transformes as categories are : ``` pclass, sex, cabin, embarked, boat```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "categorical_col= ['pclass','sex', 'cabin', 'embarked', 'boat']\n",
    "for col in categorical_col:\n",
    "    titanic[col] = titanic[col].astype('category')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Question 2 : Now we plot the histograms**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x10da5df98>"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEICAYAAABYoZ8gAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFSRJREFUeJzt3WuwnVd93/HvzxIICja2cSJ0LIOcVokrBzCECggZoLhg\nmZtMk7oioREZM01n1MZpSFK50xTRiVLzhgQ6MYFw06Rgo0KIVWghRsElpMQ3cDCycSxsy5aOLmBi\nbAyxkfzvi73U7Aidc/a56Rx7fT8zZ/Z61rOe/fz33tJvP3vty5OqQpLUj5MWugBJ0oll8EtSZwx+\nSeqMwS9JnTH4JakzBr8kdcbg12NCkt9O8uFZbL83ycvnrqJJ9/WWJNdOsv5Pk/zCiahFOh6DXxNK\n8t2hv0eTfH9oeVEFV5JTk7wryT2tvt1J3pnk6Qtd27Gq6lVV9ZHJxiRZmqSSrDoxVaknBr8mVFVP\nPfoH3AO8bqjvh4IrydITXyUkeRLwZ8A5wKuAU4CfBr4DvGAhalrsFuqx0uJg8GvG2vTLx5JcmeRB\n4E1JXpzkL5Pcn2R/kncneUIb/4dJLj/mOj6d5Fdae2WSTyb5ZpK7kmwasZQ3A88A3lBVX6+qR6vq\nUFW9vao+e5y6J6vxpLZ8KMl3knw1yZq27rVJbkvyYJs6+veT3z353baPO5O8amjFF5O8ubV/PMkX\n2r6+leSjbdgX2uWu9grmZ9v4f9NezdyX5E+SrBi63guT/HW7rv+W5C+G9vOWtp93J/k28J+SrE7y\n+STfbvv+oyRPG7q+vUl+PcnXWg3vS7I8yWeTPNCmrE4d8THSImLwa7beAHwUeBrwMeAwcClwBvAS\nYB3wy23slcCGJAFo0zCvAD6W5CTgU8ANwJnAK4HfSHL+CDX8M+B/V9X3Rqx5shovBF4ErAZOAzYA\n327rPgRcUlUnA88B/s8k+/hp4Bbg6cDvAh+YYNxW4NNtXyuB32/9L22X57ZXWJ9oTx7/Bfg5BvfR\nOPARgCQ/CmwHfqPdrruAtcep6TbgR4B3AAF+m8GT5hrgx4DfOmabNzB4jM4BfrbV+pvAjwLLgFGf\nnLWIGPyarS9W1f9sR9nfr6obquq6qjpcVXcC7wNe1sZeCzwBeHFbvhj486o62PpOqarfqapHqmo3\ng7DcMEINTwf2j1rwFDX+gMFU0Tlt7K1VdWBo3ZokJ1fVt6vqy5Ps5htV9cGqOgJsA1YmOeM4434A\nrAJWVNXfVtVfTHKdvwC8v6purqq/BTYDL0uyEngtcHNVXV1VP2DwZPOtY7a/p6reU1VH2mP111W1\ns93fh9o2Lztmm3e3V097gS8CX6qqv2r7/xPgeZPUq0XK4Nds3Tu8kOScNn1zIMkDDI5QzwCoqkcZ\nvCp4Yxv+87QjVuBZwDPb1Mj9Se5ncGT5jBFquA9YMeWo0Wr8U+APgPcAB5P8QZKT26ZvAF4P3JPk\n2iQvnGQ3B4baR1+JPPU4497K4MnwxiS3JNk4yXWOAXuOLlTVA8DfMDj6H2PosajBry/uPWb7Yx+r\nZyTZnmRfux8+TLsfhhwcan//OMvHu01a5Ax+zdaxP+/6XuBrwD+qqlOA/8xgSuGoK4F/keRs4PnA\nH7f+e4E7qurUob+Tq+p1I9TwOeDCJE8eseZJa6yq36uq5wM/yWAK5Nda/3VV9XoG0xyfAq4acX8T\nqqr9VfWWqlrBYNrkfe2+Od7P5o4zeIIEoD0hnQbsY/CKZ+XQujB4Qvh7uztm+R3Aw8Cz2/3wZv7+\nY6XHKYNfc+1kBp+meSjJP+bv5s6BwTQL8ACD6ZX/VVUPtlVfAh5J8tYkT0qyJMmzk/zUCPv8MIMj\n7E8k+YkMnJHkt5JcMJ0ak6xtf0uBh4BHgEeTPDnJzyc5pU2lPAg8OvK9MoEkFyc5GtD3MwjnI22K\n6D4G8+5HXQlckuQ5SZYB/5XBVNleBk9Ez0/yulb7pQzm8idzcruN30lyFvDrs709emww+DXX3gps\nZBCM72UwtXOsKxm8IXv0EyxU1WHg1QzekLybwfz0exnMt0+qzTe/AtjN4Oj/QeAvGbzhfMM0azyV\nwXsL97c69gPvbOs2AnvatMglwJumqm0ELwRuSPIQg1c/m6rqnrbubcBH29TXP6+qzzCYlvpkq+uZ\nDOb9ae+T/MtW633APwS+wuCIfiJvY3B/fwfYAXxiDm6PHgPiiVikx58kSxhMDf1cVf35QtejxcUj\nfulxIsm6DL7BvIzBxzJ/AFy/wGVpETL4pcePnwHuBL4JXMDgC22TTfWoU071SFJnPOKXpM4syh9q\nOuOMM2rVqlULXYYkPabcdNNN36qqqT7GuziDf9WqVdx4440LXYYkPaYk2TP1KKd6JKk7Br8kdcbg\nl6TOGPyS1BmDX5I6Y/BLUmcMfknqjMEvSZ2ZMvjbiS1uHvp7IMmvJjk9yTVJ7miXpw1tc1mS3Ulu\nn+BEGJKkBTLlN3er6nbgPPj/v/G9j8GJIDYDO6vq8iSb2/J/SLKGwQmyz2VwHtDPJfnxdkahRWnV\n5k8vdAnz6u7LX7PQJUhaRKY71XM+8I2q2gOsB7a1/m3ARa29Hriqqh6uqrsYnBVp7VwUK0mavekG\n/wYGp80DWF5V+1v7ALC8tc9kcOLso/bywyd9/iFJtiSpJDU+Pj7NsiRJoxo5+JM8EXg98D+OXVeD\nH/Wf1Q/7V9WWqkpVZWxsbDZXJUmaxHSO+C8EvtxO6gxwMMkKgHZ5qPXvA84a2m5l65MkLQLTCf43\n8nfTPAA7gI2tvRG4eqh/Q5JlSc4GVuN5PyVp0Rjp9/iTPAV4JfDLQ92XA9uTXALsAS4GqKpdSbYD\ntwKHgU2L+RM9ktSbkYK/qh4Cnn5M330MPuVzvPFbga2zrk6SNOf85q4kdcbgl6TOGPyS1BmDX5I6\nY/BLUmcMfknqjMEvSZ0x+CWpMwa/JHXG4Jekzhj8ktQZg1+SOmPwS1JnDH5J6ozBL0mdMfglqTMG\nvyR1xuCXpM4Y/JLUGYNfkjozUvAnOTXJx5N8PcltSV6c5PQk1yS5o12eNjT+siS7k9ye5IL5K1+S\nNF2jHvG/C/hMVZ0DPBe4DdgM7Kyq1cDOtkySNcAG4FxgHXBFkiVzXbgkaWamDP4kTwNeCnwAoKoe\nqar7gfXAtjZsG3BRa68Hrqqqh6vqLmA3sHauC5ckzcwoR/xnA98EPpTkK0nen+QpwPKq2t/GHACW\nt/aZwL1D2+9tfZKkRWCU4F8KPB94T1U9D3iINq1zVFUVULMpJMmWJJWkxsfHZ3NVkqRJjBL8e4G9\nVXVdW/44gyeCg0lWALTLQ239PuCsoe1Xtr5JVdWWqkpVZWxsbNT6JUnTNGXwV9UB4N4kP9G6zgdu\nBXYAG1vfRuDq1t4BbEiyLMnZwGrg+jmtWpI0Y0tHHPfvgI8keSJwJ/BLDJ40tie5BNgDXAxQVbuS\nbGfw5HAY2FRVR+a8cknSjIwU/FV1M/CC46w6f4LxW4Gts6hLkjRP/OauJHXG4Jekzhj8ktQZg1+S\nOmPwS1JnDH5J6ozBL0mdMfglqTMGvyR1xuCXpM4Y/JLUGYNfkjpj8EtSZwx+SeqMwS9JnTH4Jakz\nBr8kdcbgl6TOGPyS1BmDX5I6Y/BLUmdGCv4kdye5JcnNSW5sfacnuSbJHe3ytKHxlyXZneT2JBfM\nV/GSpOmbzhH/P62q86rqBW15M7CzqlYDO9sySdYAG4BzgXXAFUmWzGHNkqRZmM1Uz3pgW2tvAy4a\n6r+qqh6uqruA3cDaWexHkjSHRg3+Aj6X5KYk/7r1La+q/a19AFje2mcC9w5tu7f1TSrJliSVpMbH\nx0csS5I0XaMG/89U1XnAhcCmJC8dXllVxeDJYcaqaktVpaoyNjY2m6uSJE1ipOCvqn3t8hDwSQZT\nNweTrABol4fa8H3AWUObr2x9kqRFYMrgT/KUJCcfbQOvAr4G7AA2tmEbgatbewewIcmyJGcDq4Hr\n57pwSdLMLB1hzHLgk0mOjv9oVX0myQ3A9iSXAHuAiwGqaleS7cCtwGFgU1UdmZfqJUnTNmXwV9Wd\nwHOP038fcP4E22wFts66OknSnPObu5LUGYNfkjpj8EtSZwx+SeqMwS9JnTH4JakzBr8kdcbgl6TO\nGPyS1BmDX5I6Y/BLUmcMfknqjMEvSZ0x+CWpMwa/JHXG4Jekzhj8ktQZg1+SOmPwS1JnDH5J6szI\nwZ9kSZKvJPlUWz49yTVJ7miXpw2NvSzJ7iS3J7lgPgqXJM3MdI74LwVuG1reDOysqtXAzrZMkjXA\nBuBcYB1wRZIlc1OuJGm2Rgr+JCuB1wDvH+peD2xr7W3ARUP9V1XVw1V1F7AbWDs35UqSZmvUI/7f\nA34TeHSob3lV7W/tA8Dy1j4TuHdo3N7WN6kkW5JUkhofHx+xLEnSdE0Z/EleCxyqqpsmGlNVBdRs\nCqmqLVWVqsrY2NhsrkqSNImlI4x5CfD6JK8GngSckuS/AweTrKiq/UlWAIfa+H3AWUPbr2x9kqRF\nYMoj/qq6rKpWVtUqBm/a/llVvQnYAWxswzYCV7f2DmBDkmVJzgZWA9fPeeWSpBkZ5Yh/IpcD25Nc\nAuwBLgaoql1JtgO3AoeBTVV1ZNaVSpLmxLSCv6quBa5t7fuA8ycYtxXYOsvaJEnzwG/uSlJnDH5J\n6ozBL0mdMfglqTMGvyR1xuCXpM4Y/JLUGYNfkjpj8EtSZwx+SeqMwS9JnTH4JakzBr8kdcbgl6TO\nGPyS1BmDX5I6Y/BLUmdmc+pFacGt2vzphS5hXt19+WsWugQ9DnnEL0mdMfglqTNTBn+SJyW5Pslf\nJdmV5O2t//Qk1yS5o12eNrTNZUl2J7k9yQXzeQMkSdMzyhH/w8Arquq5wHnAuiQvAjYDO6tqNbCz\nLZNkDbABOBdYB1yRZMl8FC9Jmr4pg78GvtsWn9D+ClgPbGv924CLWns9cFVVPVxVdwG7gbVzWrUk\nacZGmuNPsiTJzcAh4Jqqug5YXlX725ADwPLWPhO4d2jzva1vqn1sSVJJanx8fOQbIEmanpGCv6qO\nVNV5wEpgbZKfPGZ9MXgVMGNVtaWqUlUZGxubzVVJkiYxrU/1VNX9wOcZzN0fTLICoF0easP2AWcN\nbbay9UmSFoFRPtXzI0lObe0nA68Evg7sADa2YRuBq1t7B7AhybIkZwOrgevnunBJ0syM8s3dFcC2\n9smck4DtVfWpJF8Ctie5BNgDXAxQVbuSbAduBQ4Dm6rqyPyUL0marimDv6q+CjzvOP33AedPsM1W\nYOusq5MkzTm/uStJnfFH2iQtGH9kb2F4xC9JnTH4JakzBr8kdcbgl6TOGPyS1BmDX5I6Y/BLUmcM\nfknqjMEvSZ0x+CWpMwa/JHXG4Jekzhj8ktQZg1+SOmPwS1JnDH5J6ozBL0mdMfglqTNTBn+Ss5J8\nPsmtSXYlubT1n57kmiR3tMvThra5LMnuJLcnuWA+b4AkaXpGOeI/DLy1qtYALwI2JVkDbAZ2VtVq\nYGdbpq3bAJwLrAOuSLJkPoqXJE3flMFfVfur6sut/SBwG3AmsB7Y1oZtAy5q7fXAVVX1cFXdBewG\n1s514ZKkmZnWHH+SVcDzgOuA5VW1v606ACxv7TOBe4c229v6JEmLwMjBn+SpwCeAX62qB4bXVVUB\nNZtCkmxJUklqfHx8NlclSZrESMGf5AkMQv8jVfXHrftgkhVt/QrgUOvfB5w1tPnK1jepqtpSVamq\njI2NjVq/JGmaRvlUT4APALdV1TuHVu0ANrb2RuDqof4NSZYlORtYDVw/dyVLkmZj6QhjXgL8K+CW\nJDe3vv8IXA5sT3IJsAe4GKCqdiXZDtzK4BNBm6rqyJxXLkmakSmDv6q+CGSC1edPsM1WYOss6pIk\nzRO/uStJnTH4JakzBr8kdcbgl6TOGPyS1BmDX5I6Y/BLUmcMfknqjMEvSZ0x+CWpMwa/JHXG4Jek\nzhj8ktQZg1+SOmPwS1JnDH5J6ozBL0mdMfglqTMGvyR1xuCXpM5MGfxJPpjkUJKvDfWdnuSaJHe0\ny9OG1l2WZHeS25NcMF+FS5JmZpQj/g8D647p2wzsrKrVwM62TJI1wAbg3LbNFUmWzFm1kqRZmzL4\nq+oLwLeP6V4PbGvtbcBFQ/1XVdXDVXUXsBtYO0e1SpLmwEzn+JdX1f7WPgAsb+0zgXuHxu1tfZKk\nRWLWb+5WVQE12+tJsiVJJanx8fHZXp0kaQIzDf6DSVYAtMtDrX8fcNbQuJWtb0pVtaWqUlUZGxub\nYVmSpKnMNPh3ABtbeyNw9VD/hiTLkpwNrAaun12JkqS5tHSqAUmuBF4OnJFkL/A24HJge5JLgD3A\nxQBVtSvJduBW4DCwqaqOzFPtkqQZmDL4q+qNE6w6f4LxW4GtsylKkjR//OauJHXG4Jekzhj8ktQZ\ng1+SOmPwS1JnDH5J6ozBL0mdMfglqTMGvyR1xuCXpM4Y/JLUGYNfkjpj8EtSZwx+SeqMwS9JnTH4\nJakzBr8kdcbgl6TOGPyS1BmDX5I6Y/BLUmfmLfiTrEtye5LdSTbP134kSdMzL8GfZAnw+8CFwBrg\njUnWzMe+JEnTM19H/GuB3VV1Z1U9AlwFrJ+nfUmSpmHpPF3vmcC9Q8t7gRdOtkGSLcDb2uL3ktw2\nP6UtCmPA+InaWd5xovbUBR+7x7bH++P3rFEGzVfwT1tVbQG2LHAZJ0SSqqqxha5D0+dj99jm4zcw\nX1M9+4CzhpZXtj5J0gKbr+C/AVid5OwkTwQ2ADvmaV+SpGmYl6meqjqc5N8CnwWWAB+sql3zsa/H\nqLcvdAGaMR+7xzYfPyBVtdA1SJJOIL+5K0mdMfglqTMGvyR1xuCXpM4Y/JLUGYNfkjpj8EtSZwz+\neZZkbZJ/0tprkvxaklcvdF1SD5Kck+T8JE89pn/dQtW0GPgFrnmU5G0MzkmwFLiGwS+Ufh54JfDZ\nqtq6gOVpFpL8UlV9aKHr0MSS/AqwCbgNOA+4tKqubuu+XFXPX8j6FpLBP4+S3MLgH9wy4ACwsqoe\nSPJk4Lqqes6CFqgZS3JPVT1zoevQxNr/vxdX1XeTrAI+DvxRVb0ryVeq6nkLWuACWjQ/y/w4dbiq\njjA4v8A3quoBgKr6fpJHF7g2TSHJVydaBSw/kbVoRk6qqu8CVNXdSV4OfDzJsxg8ht0y+OfXI0n+\nQVV9D/ipo51JngYY/IvfcuAC4G+O6Q/wf098OZqmg0nOq6qbAdqR/2uBDwLPXtjSFpbBP79eWlUP\nA1TVcNA/Adi4MCVpGj4FPPVocAxLcu2JL0fT9IvA4eGOqjoM/GKS9y5MSYuDc/yS1Bk/zilJnTH4\nJakzBr8kdcbgl6TO/D9oADlEOaUtgQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10da0a390>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "titanic['pclass'].value_counts().plot(kind='bar').set_title('Travel Class histogram')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x10dda9198>"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEkCAYAAAAoz1ptAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAF81JREFUeJzt3W2wXVd93/Hvz5ItbMNgOSaKriQsQQWuRME81BjoUIgB\ni9AgzzTjiodUULeFIAIkTImc0iLoqHFSmjQvaigkgAgJGpVArAB1cASmIQQUjF1AMoqFhZB09WCb\nGIOhAol/X5wlOFYl3XN1H8713d/PzJmz9tpr7/0/V1e/s+86DztVhSSpO84ZdgGSpOll8EtSxxj8\nktQxBr8kdYzBL0kdY/BLUscY/Jo1krwqyecmcX+3JvnXp1n32CTfSzJnso4nTReDX1MqyZokX0zy\nYJIjrf26JBl2bRNRVd+qqkdW1fEzjZvsJyNpMhj8mjJJ3gz8PvBfgJ8DFgCvBZ4DnDfE0v4/D8cz\n9/T4f1jj5i+NpkSSRwPvAF5XVR+pqu9Wz+1V9YqqOtrGzUvyziTfSnI4ybuTnN/WPS/J/iRvbn8t\nHEzy6r5j/EySrUkeSLIdePxJNVyW5JYk306yK8m1fes+kORdST6Z5EHg+ad5KJcm+esk303yqSSX\ntO2XJqkkc9vyq5Lc3cbtSfKKJP8QeDfwrDYtdP+Jn02SDya5J8neJG89EeBJ5iT5r0nubft5/UnH\nuTXJxiR/DXwfeFySVye5sx377iSv6XucJ36Gb+n7GV6T5BeS/F372fzmBP6p9XBUVd68TfoNWAUc\nA+aOMe73gK3AxcCjgD8Hfqute17bxzuAc4FfoBd289v6zcAW4ELgScAB4HNt3YXAPuDVwFzgqcC9\nwIq2/gPAd+j99XEO8IhT1HYr8A3gCcD5bfmGtm4pUG3fFwIPAE9s6xYCK1v7VSdq6tvvB4Gb2uNd\nCvwdcF1b91pgJ7AYmA/85Ynj9NX0LWBlO/a5wEvoPekF+KftZ/S0k36G/7GN/TfAPcCftOOvBH4A\nLBv274y36bsNvQBvs/MGvBI4dFLf54H7W9A8twXVg8Dj+8Y8C9jT2s9rY+f2rT8CXAnMAX4EXNa3\n7j/3Bf+/AP7qpOP/D+Btrf0B4INjPIZbgbf2Lb8OuLm1Tw7++4F/Dpx/0j4eEvyt7h/SnoBa32uA\nW1v708Br+ta94BTB/44x6v4z4I0n/QzntOVHtf09s2/8bcA1w/6d8TZ9N6d6NFXuAy45MUUBUFXP\nrqqL2rpzgMcAFwC3Jbm/TYXc3Pp/sp+qOta3/H3gkW3MXHpn9Sfs7WtfCjzzxH7bvl9B77WGE/q3\nPZ1Dpzj2Q1TVg/SeaF4LHEzyiSSXnWZ/l9A78+6vdS+wqLVHTqrrVDU+pC/Ji5N8oU3b3E/vL6NL\n+obcVz99EfoH7f5w3/ofnOpxafYy+DVV/gY4Cqw+w5h76YXOyqq6qN0eXVWDhNA99KYwlvT1Pbav\nvQ/4bN9+L6reu3B+pW/MpH01bVX9RVW9kN40z9eB957mGPfS+0vl0pPqPtDaB+lN85zQ//h+crgT\njSTzgD8F3gksaE+sn6T315R0Sga/pkRV3Q+8HbgxyS8leVSSc5JcTm9qhKr6Mb2A/L0kPwuQZFGS\nqwfY/3Hgo8CGJBckWQGs7RvyceAJSX45ybnt9o/bC66TKsmCJKuTXEjvye57wI/b6sPA4iTn9dW9\nBdjYfiaXAr8OfKiN3wK8sf0cLgJ+Y4zDnwfMoz0RJnkx8KJJfHiahQx+TZmq+h16ofYWegF4mN48\n+2/Qm++ntXcDX0jyAL0XM5844CFeT2+K4hC9Ofv39x37u/QCcA0w2sb8Nr2QnGzn0Huco8C36b3A\neuIvi08DO4BDSe5tfb9K77WNu4HP0Xuh9X1t3XuBTwFfAW6nd/Z+DDjl5wXa43wDvSeMvwdeTu/F\ncum0UuWFWKSZqp3Bv7uqLh1zsDQgz/ilGSTJ+e099nOTLALeBnxs2HVpdvGMX5pBklwAfBa4jN4L\n35+g99bMB4ZamGYVg1+SOsapHknqmLljD5l+l1xySS1dunTYZUjSw8ptt912b1U9ZqxxMzL4ly5d\nype+9KVhlyFJDytJ9o49yqkeSeocg1+SOsbgl6SOMfglqWMMfknqGINfkjrG4JekjjH4JaljDH5J\n6pgZ+cndh4ul6z8x7BJmjW/e8JJhlyB1hmf8ktQxBr8kdYzBL0kdY/BLUscY/JLUMQa/JHXMQMGf\n5NeS7EjytSQfTvKIJBcnuSXJXe1+ft/465PsTrIrydVTV74kabzGDP4ki4A3AM+oqicBc4A1wHpg\nW1UtB7a1ZZKsaOtXAquAG5PMmZryJUnjNehUz1zg/CRzgQuAUWA1sKmt3wRc09qrgc1VdbSq9gC7\ngSsmr2RJ0kSMGfxVdQB4J/At4CDwnar6FLCgqg62YYeABa29CNjXt4v9re+MkmxIUklqdHR0HA9B\nkjQeg0z1zKd3Fr8MGAEuTPLK/jFVVUBNpJCq2lBVqaqMjIxMZFeSpDMYZKrnBcCeqrqnqn4EfBR4\nNnA4yUKAdn+kjT8ALOnbfnHrkyTNAIME/7eAK5NckCTAVcCdwFZgbRuzFriptbcCa5LMS7IMWA5s\nn9yyJUlna8xv56yqLyb5CPBl4BhwO/Ae4JHAliTXAXuBa9v4HUm2ADvb+HVVdXyK6pckjdNAX8tc\nVW8D3nZS91F6Z/+nGr8R2Dix0iRJU8FP7kpSxxj8ktQxBr8kdYzBL0kdY/BLUscY/JLUMQa/JHWM\nwS9JHWPwS1LHGPyS1DEGvyR1jMEvSR1j8EtSxxj8ktQxBr8kdYzBL0kdM8jF1p+Y5I6+2wNJ3pTk\n4iS3JLmr3c/v2+b6JLuT7Epy9dQ+BEnSeIwZ/FW1q6our6rLgacD3wc+BqwHtlXVcmBbWybJCmAN\nsBJYBdyYZM4U1S9JGqfxTvVcBXyjqvYCq4FNrX8TcE1rrwY2V9XRqtoD7AaumIxiJUkTN97gXwN8\nuLUXVNXB1j4ELGjtRcC+vm32t74zSrIhSSWp0dHRcZYlSRrUwMGf5DzgpcD/PHldVRVQEymkqjZU\nVaoqIyMjE9mVJOkMxnPG/2Lgy1V1uC0fTrIQoN0faf0HgCV92y1ufZKkGWA8wf8yfjrNA7AVWNva\na4Gb+vrXJJmXZBmwHNg+0UIlSZNj7iCDklwIvBB4TV/3DcCWJNcBe4FrAapqR5ItwE7gGLCuqo5P\natWSpLM2UPBX1YPAz5zUdx+9d/mcavxGYOOEq5MkTTo/uStJHWPwS1LHGPyS1DEGvyR1jMEvSR1j\n8EtSxxj8ktQxBr8kdYzBL0kdY/BLUscY/JLUMQa/JHWMwS9JHWPwS1LHGPyS1DEDBX+Si5J8JMnX\nk9yZ5FlJLk5yS5K72v38vvHXJ9mdZFeSq6eufEnSeA16xv/7wM1VdRnwFOBOYD2wraqWA9vaMklW\nAGuAlcAq4MYkcya7cEnS2Rkz+JM8Gngu8IcAVfXDqrofWA1sasM2Ade09mpgc1Udrao9wG7gisku\nXJJ0dgY5418G3AO8P8ntSf6gXYN3QVUdbGMOAQtaexGwr2/7/a3vjJJsSFJJanR0dPBHIEkal0GC\nfy7wNOBdVfVU4EHatM4JVVVATaSQqtpQVamqjIyMTGRXkqQzGCT49wP7q+qLbfkj9J4IDidZCNDu\nj7T1B4Alfdsvbn2SpBlgzOCvqkPAviRPbF1XATuBrcDa1rcWuKm1twJrksxLsgxYDmyf1KolSWdt\n7oDjfhX44yTnAXcDr6b3pLElyXXAXuBagKrakWQLvSeHY8C6qjo+6ZVLks7KQMFfVXcAzzjFqqtO\nM34jsHECdUmSpoif3JWkjjH4JaljDH5J6hiDX5I6xuCXpI4x+CWpYwx+SeoYg1+SOsbgl6SOMfgl\nqWMMfknqGINfkjrG4JekjjH4JaljDH5J6hiDX5I6ZqDgT/LNJF9NckeSL7W+i5PckuSudj+/b/z1\nSXYn2ZXk6qkqXpI0fuM5439+VV1eVSeuxLUe2FZVy4FtbZkkK4A1wEpgFXBjkjmTWLMkaQImMtWz\nGtjU2puAa/r6N1fV0araA+wGrpjAcSRJk2jQ4C/gL5PcluTftr4FVXWwtQ8BC1p7EbCvb9v9re+M\nkmxIUklqdHR0wLIkSeM1aPD/k6q6HHgxsC7Jc/tXVlXRe3I4a1W1oapSVRkZGZnIriRJZzBQ8FfV\ngXZ/BPgYvambw0kWArT7I234AWBJ3+aLW58kaQYYM/iTXJjkUSfawIuArwFbgbVt2FrgptbeCqxJ\nMi/JMmA5sH2yC5cknZ25A4xZAHwsyYnxf1JVNyf5W2BLkuuAvcC1AFW1I8kWYCdwDFhXVcenpHpJ\n0riNGfxVdTfwlFP03wdcdZptNgIbJ1ydJGnS+cldSeoYg1+SOmaQOX5JDzNL139i2CXMKt+84SXD\nLmFSecYvSR1j8EtSxxj8ktQxBr8kdYzBL0kdY/BLUscY/JLUMQa/JHWMwS9JHWPwS1LHGPyS1DEG\nvyR1zMDBn2ROktuTfLwtX5zkliR3tfv5fWOvT7I7ya4kV09F4ZKkszOeM/43Anf2La8HtlXVcmBb\nWybJCmANsBJYBdyYZM7klCtJmqiBgj/JYuAlwB/0da8GNrX2JuCavv7NVXW0qvYAu+ldnF2SNAMM\nesb/34C3AD/u61tQVQdb+xC9a/MCLAL29Y3b3/okSTPAmMGf5J8BR6rqttONqaoCaiKFJNmQpJLU\n6OjoRHYlSTqDQc74nwO8NMk3gc3Azyf5EHA4yUKAdn+kjT8ALOnbfnHrO6Oq2lBVqaqMjIyM4yFI\nksZjzOCvquuranFVLaX3ou2nq+qVwFZgbRu2FriptbcCa5LMS7IMWA5sn/TKJUlnZSLX3L0B2JLk\nOmAvcC1AVe1IsgXYCRwD1lXV8QlXKkmaFOMK/qq6Fbi1te8DrjrNuI3AxgnWJkmaAn5yV5I6xuCX\npI4x+CWpYwx+SeoYg1+SOsbgl6SOMfglqWMMfknqGINfkjrG4JekjjH4JaljDH5J6hiDX5I6xuCX\npI4x+CWpYwx+SeqYQS62/ogk25P8nyQ7kry99V+c5JYkd7X7+X3bXJ9kd5JdSa6eygcgSRqfQc74\njwI/X1VPAS4HViW5ElgPbKuq5cC2tkySFfSuzbsSWAXcmGTOVBQvSRq/QS62XlX1vbZ4brsVsBrY\n1Po3Ade09mpgc1Udrao9wG7gikmtWpJ01gaa408yJ8kdwBHglqr6IrCgqg62IYeABa29CNjXt/n+\n1jfWMTYkqSQ1Ojo68AOQJI3PQMFfVcer6nJgMXBFkiedtL7o/RVw1qpqQ1WlqjIyMjKRXUmSzmBc\n7+qpqvuBz9Cbuz+cZCFAuz/Shh0AlvRttrj1SZJmgEHe1fOYJBe19vnAC4GvA1uBtW3YWuCm1t4K\nrEkyL8kyYDmwfbILlySdnbkDjFkIbGrvzDkH2FJVH0/yN8CWJNcBe4FrAapqR5ItwE7gGLCuqo5P\nTfmSpPEaM/ir6ivAU0/Rfx9w1Wm22QhsnHB1kqRJ5yd3JaljDH5J6hiDX5I6xuCXpI4x+CWpYwx+\nSeoYg1+SOsbgl6SOMfglqWMMfknqGINfkjrG4JekjjH4JaljDH5J6hiDX5I6xuCXpI4Z5NKLS5J8\nJsnOJDuSvLH1X5zkliR3tfv5fdtcn2R3kl1Jrp7KByBJGp9BzviPAW+uqhXAlcC6JCuA9cC2qloO\nbGvLtHVrgJX0Lsp+Y7tsoyRpBhgz+KvqYFV9ubW/C9wJLAJWA5vasE3ANa29GthcVUerag+wG7hi\nsguXJJ2dcc3xJ1lK7/q7XwQWVNXBtuoQsKC1FwH7+jbb3/rG2veGJJWkRkdHx1OWJGkcBg7+JI8E\n/hR4U1U90L+uqgqoiRRSVRuqKlWVkZGRiexKknQGAwV/knPphf4fV9VHW/fhJAvb+oXAkdZ/AFjS\nt/ni1idJmgEGeVdPgD8E7qyq3+1btRVY29prgZv6+tckmZdkGbAc2D55JUuSJmLuAGOeA/wy8NUk\nd7S+3wRuALYkuQ7YC1wLUFU7kmwBdtJ7R9C6qjo+6ZVLks7KmMFfVZ8DcprVV51mm43AxgnUJUma\nIn5yV5I6xuCXpI4x+CWpYwx+SeoYg1+SOsbgl6SOMfglqWMMfknqGINfkjrG4JekjjH4JaljDH5J\n6hiDX5I6xuCXpI4x+CWpYwa5Atf7khxJ8rW+vouT3JLkrnY/v2/d9Ul2J9mV5OqpKlySdHYGOeP/\nALDqpL71wLaqWg5sa8skWQGsAVa2bW5MMmfSqpUkTdiYwV9V/xv49kndq4FNrb0JuKavf3NVHa2q\nPcBu4IpJqlWSNAnOdo5/QVUdbO1DwILWXgTs6xu3v/WNKcmGJJWkRkdHz7IsSdJYJvziblUVUJOw\nnw1VlarKyMjIRHcnSTqNsw3+w0kWArT7I63/ALCkb9zi1idJmiHONvi3Amtbey1wU1//miTzkiwD\nlgPbJ1aiJGkyzR1rQJIPA88DLkmyH3gbcAOwJcl1wF7gWoCq2pFkC7ATOAasq6rjU1S7JOksjBn8\nVfWy06y66jTjNwIbJ1KUJGnq+MldSeoYg1+SOsbgl6SOMfglqWMMfknqGINfkjrG4JekjjH4Jalj\nDH5J6hiDX5I6xuCXpI4x+CWpYwx+SeoYg1+SOsbgl6SOMfglqWOmLPiTrEqyK8nuJOun6jiSpPGZ\nkuBPMgf478CLgRXAy5KsmIpjSZLGZ6rO+K8AdlfV3VX1Q2AzsHqKjiVJGocxr7l7lhYB+/qW9wPP\nPNMGSTbQu5A7wPeT3Dk1pXXSCDA67CLOJL897Ao0JDP+dxMeVr+flw4yaKqCf9yqagOwYchlzEpJ\nqqpGhl2HdDJ/N4djqqZ6DgBL+pYXtz5J0pBNVfD/LbA8ybIk5wFrgK1TdCxJ0jhMyVRPVR1L8nrg\nL4A5wPuqasdUHEsDefuwC5BOw9/NIUhVDbsGSdI08pO7ktQxBr8kdYzBL0kdY/BLUscY/JLUMQa/\nJHWMwS9JHWPwz1JJzk/yxGHXIWnmMfhnoSS/CNwB3NyWL0/iV2Zo6JI8Icm2JF9ry09O8tZh19U1\nBv/stIHeNRHuB6iqO4BlwyxIat4LXA/8CKCqvkLvu7w0jQz+2elHVfWdk/r8bg7NBBdU1faT+o4N\npZIOmzHfx69JtSPJy4E5SZYDbwA+P+SaJIB7kzyediKS5JeAg8MtqXv8krZZKMkFwL8HXgSE3rek\n/qeq+r9DLUydl+RxwHuAZwN/D+wBXllV3xxmXV1j8EuadkkuBM6pqu8Ou5YuMvhnkSR/zhnm8qvq\npdNYjvQTSX79TOur6nenqxY5xz/bvHPYBUin8ahhF6Cf8oxfkjrGM/5ZqL2T57eAFcAjTvRX1eOG\nVpQEJHkEcB2wkof+bv6roRXVQb6Pf3Z6P/Aueu+Pfj7wQeBDQ61I6vkj4OeAq4HPAosBX+CdZk71\nzEJJbquqpyf5alX9o/6+Ydembktye1U9NclXqurJSc4F/qqqrhx2bV3iVM/sdDTJOcBdSV4PHAAe\nOeSaJGhf1QDcn+RJwCHgZ4dYTyc51TM7vRG4gN4ndp8OvBL4l0OtSOp5T5L5wH8AtgI7gd8Zbknd\n41TPLJTkGfQ+uXspcG7rrqp68vCqkjRTGPyzUJJdwL8Dvgr8+ER/Ve0dWlESkOQien99LqVvqrmq\n3jCsmrrIOf7Z6Z6q8vv3NRN9EvgCJ52UaHp5xj8LJbkKeBmwDTh6or+qPjq0oiQgyZer6mnDrqPr\nDP5ZKMmHgMuAHfz0rKr8kIyGLcmvAd8DPs5DT0q+PbSiOsjgn4WS7Koqr7erGSfJOmAjvavDnQif\n8lPl08s5/tnp80lWVNXOYRcineTNwD+oqnuHXUiXGfyz05XAHUn20PtzOvh2Ts0Mu4HvD7uIrjP4\nZ6dVwy5AOo0H6Z2UfIaHzvH7ds5pZPDPQr5fXzPYn7WbhsgXdyVNqyTnA4+tql3DrqWr/K4eSdMm\nyS8CdwA3t+XLk/hhw2lm8EuaThuAK+i9nZOqugPwrZzTzOCXNJ1+VFXfOanPr26YZr64K2k67Ujy\ncmBOu0ToG4DPD7mmzvGMX9KUS/JHrfkNetfbPQp8GHgAeNOw6uoq39Ujacol2Qm8APhf9K4D/RB+\nV8/0cqpH0nR4N71vi30c8KW+/tD7zh5f4J1GnvFLmjZJ3lVVvzLsOrrO4JekjvHFXUnqGINfkjrG\n4JekjjH4Jalj/h8unaNJGAEOBgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10db20b38>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "titanic['sex'].value_counts().plot(kind='bar').set_title('Gender histogram')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x10dee9c50>"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAFFCAYAAADrZ4aSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGbtJREFUeJzt3Xu4XXV95/H3hwRBjSgIjQRSEjEioAIKiNeqdApew4xT\nxHoBij4dxctoqwX1kUwtHW2tl6E606laUCxIGR2w1ltRiuIod0VASuQiIdwZFEXB4Hf+2CvDJp7k\n7JPk7LUPv/frefZz1vqty/7us5PPXue31/qtVBWSpHZs0XcBkqTxMvglqTEGvyQ1xuCXpMYY/JLU\nGINfkhpj8GtOSXJikj/vu46pJPlSksP7rkOajsGvWZfk2iS/SPKzocff9F3XpkiyIsnJw21V9YKq\nOmkWnuvEJPd2v7c7knwtyRM2cl/PTbJqc9eoucXg17i8pKoWDD3e2HdBSeb3XcMM/GVVLQB2Bm4B\nTpzpDubY69UsMvjVqyRHJDk3yYeS3Jnk6iTP6NqvT3LLFN0n23dHvXcl+dckuwzt7yPddj9NcmGS\nZw8tW5Hk9CQnJ/kpcESSeUnemeRH3f4uTLJ4Q/tKcjDwTuDl3VH497r2s5O8tpveIsm7k1zXvYZP\nJXlkt2xJkkpyeJIfJ7ktybtG+X1V1d3APwBP7Pa1VZIPJ1ndPT6cZKtu2XOTrEryp0luAk4BvgQs\nGvrLa9GM3zTNeQa/JsHTgO8Dj2YQaqcC+wGPA14F/E2SBUPrvxJ4L7A9cAnwmaFl5wN7A9t1+/rH\nJFsPLV8OnA48qtvubcArgBcC2wB/CNy9oX1V1ZeBvwA+2/31stcUr+mI7vE84LHAAmDd7q1nAbsB\nBwLvSbL7Bn5HAHS/h1cCF3dN7wIO6OrcC9gfePfQJo/p6t8FeA3wAmD10F9eq6d7Tj0IVZUPH7P6\nAK4FfgbcOfR4XbfsCOCqoXWfBBSwcKjtdmDvbvpE4NShZQuA+4DF63nu/wvs1U2vAM5ZZ/mVwPIR\nX8e6+zp5neVnA6/tps8C3jC0bDfgV8B8YEn3GnceWn4ecNh6nvdE4Jfd7+0m4Exg127Zj4AXDq17\nEHBtN/1c4F5g66HlzwVW9f1vwke/D/v8NC6HVNW/rGfZzUPTvwCoqnXbho/4r187UVU/S3IHsAi4\nPsmfAEd188XgKH77qbbtLGYQnr9hhH1tyCLguqH56xiE/sKhtpuGpu/mga9xXR+oqndP0T7V8wx3\n39xaVb8cqWI1w64ezUWL1050XR/bAau7Pvh3AIcC21bVo4CfABnadt3haK8Hdl33CUbY13TD2q5m\n0L2y1m8Da3jgh9zmMNXzDHffrFunw/HK4Nec9MIkz0ryEAZ9/d+pquuBRzAI11uB+Unew+AofUM+\nDrw3ybIMPDnJo0fY183AkiTr+z90CvDWJEu7D6e13wms2biXvF6nAO9OskOS7YH3ACdvYP2bgUev\n/aJZbTL4NS5fWOc8/s9vwr7+ATgOuAN4KoMvgAG+AnwZ+DcGXR6/5De7dtb1QeA04KvAT4FPAA8d\nYV//2P28PclFU+z3k8CngXOAa7rt3zTqC5yBPwcuYPDl+KXARV3blKrqhww+LK7uzqLyrJ4Gpcq/\n/CSpJR7xS1JjDH5JaozBL0mNMfglqTEGvyQ1ZiKv3N1+++1ryZIlfZchSXPKhRdeeFtV7TDdehMZ\n/EuWLOGCCy7ouwxJmlOSXDf9Wnb1SFJzDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhoz\nkRdwjduSY77Ydwmz6tr3vajvEiRNEI/4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEv\nSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLU\nGINfkhozUvAneWuSy5L8IMkpSbZOsl2SryW5qvu57dD6xyZZmeTKJAfNXvmSpJmaNviT7AS8Gdi3\nqp4IzAMOA44BzqqqZcBZ3TxJ9uiW7wkcDHwsybzZKV+SNFOjdvXMBx6aZD7wMGA1sBw4qVt+EnBI\nN70cOLWq7qmqa4CVwP6br2RJ0qaYNvir6gbgA8CPgRuBn1TVV4GFVXVjt9pNwMJueifg+qFdrOra\nJEkTYJSunm0ZHMUvBRYBD0/yquF1qqqA2pRCkqxIUklq9erVm7IrSdIGjNLV87vANVV1a1X9Cvgc\n8Azg5iQ7AnQ/b+nWvwFYPLT9zl3bBlXViqpKVWXRokUzeQ2SpBkYJfh/DByQ5GFJAhwIXAGcCRze\nrXM4cEY3fSZwWJKtkiwFlgHnbd6yJUkba/50K1TVd5OcDlwErAEuBv4nsAA4LclRwHXAod36lyU5\nDbi8W//oqrpvluqXJM3QtMEPUFXHAcet03wPg6P/qdY/Hjh+00qTJM0Gr9yVpMYY/JLUGINfkhpj\n8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/\nJDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtS\nYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqzEjBn+RRSU5P8sMkVyR5\nepLtknwtyVXdz22H1j82ycokVyY5aPbKlyTN1KhH/B8BvlxVTwD2Aq4AjgHOqqplwFndPEn2AA4D\n9gQOBj6WZN7mLlyStHGmDf4kjwSeA3wCoKrurao7geXASd1qJwGHdNPLgVOr6p6qugZYCey/uQuX\nJG2cUY74lwK3An+f5OIkH0/ycGBhVd3YrXMTsLCb3gm4fmj7VV3bBiVZkaSS1OrVq0d/BZKkGRkl\n+OcDTwH+e1XtA/ycrltnraoqoDalkKpaUVWpqixatGhTdiVJ2oBRgn8VsKqqvtvNn87gg+DmJDsC\ndD9v6ZbfACwe2n7nrk2SNAGmDf6qugm4PsluXdOBwOXAmcDhXdvhwBnd9JnAYUm2SrIUWAact1mr\nliRttPkjrvcm4DNJHgJcDRzJ4EPjtCRHAdcBhwJU1WVJTmPw4bAGOLqq7tvslUuSNspIwV9VlwD7\nTrHowPWsfzxw/CbUJUmaJV65K0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9J\njTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQY\ng1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4\nJakxBr8kNcbgl6TGjBz8SeYluTjJP3Xz2yX5WpKrup/bDq17bJKVSa5MctBsFC5J2jgzOeJ/C3DF\n0PwxwFlVtQw4q5snyR7AYcCewMHAx5LM2zzlSpI21UjBn2Rn4EXAx4ealwMnddMnAYcMtZ9aVfdU\n1TXASmD/zVOuJGlTjXrE/2HgHcCvh9oWVtWN3fRNwMJueifg+qH1VnVtkqQJMG3wJ3kxcEtVXbi+\ndaqqgNqUQpKsSFJJavXq1ZuyK0nSBoxyxP9M4KVJrgVOBZ6f5GTg5iQ7AnQ/b+nWvwFYPLT9zl3b\nBlXViqpKVWXRokUzeAmSpJmYNvir6tiq2rmqljD40vbrVfUq4Ezg8G61w4EzuukzgcOSbJVkKbAM\nOG+zVy5J2ijzN2Hb9wGnJTkKuA44FKCqLktyGnA5sAY4uqru2+RKJUmbxYyCv6rOBs7upm8HDlzP\nescDx29ibZKkWeCVu5LUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5Ia\nY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEG\nvyQ1xuCXpMbM77sAaVMsOeaLfZcwq65934v6LkEPQh7xS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCX\npMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjZk2+JMsTvKNJJcnuSzJW7r27ZJ8LclV3c9th7Y5NsnK\nJFcmOWg2X4AkaWZGOeJfA/xxVe0BHAAcnWQP4BjgrKpaBpzVzdMtOwzYEzgY+FiSebNRvCRp5qYN\n/qq6saou6qbvAq4AdgKWAyd1q50EHNJNLwdOrap7quoaYCWw/+YuXJK0cWbUx59kCbAP8F1gYVXd\n2C26CVjYTe8EXD+02aquTZI0AUYO/iQLgP8F/Oeq+unwsqoqoDalkCQrklSSWr169absSpK0ASMF\nf5ItGYT+Z6rqc13zzUl27JbvCNzStd8ALB7afOeubYOqakVVpaqyaNGiUeuXJM3QKGf1BPgEcEVV\nfXBo0ZnA4d304cAZQ+2HJdkqyVJgGXDe5itZkrQpRrn14jOBVwOXJrmka3sn8D7gtCRHAdcBhwJU\n1WVJTgMuZ3BG0NFVdd9mr1yStFGmDf6q+haQ9Sw+cD3bHA8cvwl1SZJmiVfuSlJjDH5JaozBL0mN\nMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4Jakxo4zOKUmzYskxX+y7hFl1\n7fte1HcJU/KIX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5J\naozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TG\nGPyS1JhZC/4kBye5MsnKJMfM1vNIkmZmVoI/yTzgo8ALgD2AVyTZYzaeS5I0M7N1xL8/sLKqrq6q\ne4FTgeWz9FySpBmYP0v73Qm4fmh+FfC0DW2QZAVwXDd7d5IrZqe0ibAIWD2uJ8v7x/VMTfC9m9se\n7O/fLqOsNFvBP2NVtQJY0XMZY5GkqmpR33Vo5nzv5jbfv4HZ6uq5AVg8NL9z1yZJ6tlsBf/5wLIk\nS5M8BDgMOHOWnkuSNAOz0tVTVWuSvBH4CjAP+GRVXTYbzzVH/Ze+C9BG872b23z/gFRV3zVIksbI\nK3clqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWrMxIzV82CXZCcGAyj9/995VZ3TX0WSWmXw\nj0GS9wMvBy4H7uuaCzD454AkX2Dwfg37CXAB8LdV9cvxV6VRJNkKeBmwhAcedP1ZXzVNAoN/PA4B\ndquqe/ouRBvlamAH4JRu/uXAXcDjgb8DXt1TXZreGQw+pC8E/P/XMfjH42pgS/yHN1c9o6r2G5r/\nQpLzq2q/JI5BNdl2rqqD+y5i0hj843E3cEmSsxgK/6p6c38laQYWJPntqvoxQJLfBhZ0y+7tryyN\n4NtJnlRVl/ZdyCQx+MfjTByWei77Y+BbSX4EBFgKvCHJw4GTeq1M03kWcESSaxgcdAWoqnpyv2X1\ny9E5x6S7L8Hju9krq+pXfdaj0STZAjiAQR/xE7rmK/1Cd25IMuWtCKvqunHXMkkM/jFI8lwGR4bX\nMjjiWAwc7umcc0OSi6tqn77r0MwleS+Ds+e+XVU/77ueSWHwj0GSC4E/qKoru/nHA6dU1VP7rUyj\nSPIB4P8Anyv/w8wpSY4Eng08ncGZWN8EzqmqM3otrGcG/xgk+f66fYpTtWkyJbkLeDiwBvgl9/cT\nb9NrYRpZkscAhwJ/AmxbVY/ouaReGfxjkOSTwK+Bk7umVwLzquoP+6tKevBL8nFgD+BmBkf73wIu\nqqo1vRbWM8/qGY/XA0cDa0/f/Cbw0f7K0Uwkec5U7X5HMyc8msF9v+8E7gBuaz30wSP+sUjylqr6\nyHRtmkzdkA1rbQ3sD1xYVc/vqSTNUJLdgYOAtzL4a3vnnkvqlcE/BkkuqqqnrNPmmSJzVJLFwIer\n6mV916INS/JiBl/uPgd4FPAd4JtV9cleC+uZXT2zKMkrgD8AliYZvoBrGwZ/dmpuWgXs3ncRGsnB\nDLpWP1JVq/suZlIY/LPr28CNwPbAXw+13wV8v5eKNGNJTuD+0Tm3APYGLuqvIo2qqt6YZCGwX5Kn\nAOdV1S1919U3u3rGpDudbH8GAXJ+Vd3Uc0kaUZLDh2bXANdW1bl91aPRJfl94APA2QxOw3028Paq\nOr3Puvpm8I9BkqOA44CvM/jH9zvAn7XezziXOOTG3JTke8C/W3uUn2QH4F+qaq9+K+uXXT3j8Q5g\nn6q6HSDJoxl0Axn8c8BUQ24kcciNuWGLdbp2bsdbzhr8Y3I7g379te7q2jQ3/DXwe+sOuQE45Mbk\n+3KSr/DAm+h8qcd6JoJdPWOQ5FPAkxjcDaiA5Qy+3P0+QFV9sL/qNB2H3JjbkvwHBsMzw+BUzs/3\nWc8k8Ih/PH7UPdZaO0BU0+OFzCEXdJf+Dw+5cUGP9WhESd5fVX8KfG6KtmZ5xC9No7th99EMHTUC\nH/MeypNvPRdPNv/XmsE/Bkn2Bd4F7MLQX1mt/+ObS7qzenZnMNjelVXlLRcnWJLXA28AHssD/9p+\nBHBuVb2ql8ImhME/BkmuBN4OXMogOADvAjRXJHkR8D8YBMjaWy/+UVU1/yXhpErySGBb4L8Cxwwt\nuquqmr9q3uAfgyTfqqpnTb+mJlGSHwIvrqqV3fyuwBer6gkb3lJ9696rVVV1T3da7pOBT1XVnf1W\n1i+DfwySHAi8AjiLwQ2fAaiqz613I02MJOdX1X5D82Fw6f9+G9hMEyDJJcC+wBLgnxmcWLFnVb2w\nz7r65lk943Ekgxt1b8n9XT3F0JkGmjzdaYAwOKvnn4HTGLxvvw+c31thmolfV9Wa7r08oapOSHJx\n30X1zeAfj/2qare+i9CMvWRo+mYGQ20A3MpgXH5Nvl91o+S+hvvfzy17rGciGPzj8e0ke1TV5X0X\notFV1ZFJ5gFvrqoP9V2PNsqRwH8Cjq+qa5IsBT7dc029s49/DJJcAewKXMOgj3/tzbo9nXMOSHJe\nVe3fdx3S5mLwj0GSXaZq93TOuSHJhxh0D3wW+Pna9qpyTP4Jl+SZwAruv4Zm7UHXY/usq28G/xgl\n+S2G+oar6sc9lqMRJfnGFM3lPXcnX3cq7luBC4H71ravHSm3VQb/GCR5KYMRHhcBtzA4+riiqvbs\ntTDpQS7Jd6vqaX3XMWmaH5d6TN4LHAD8W1UtBQ5kcNNnzQFJFib5RJIvdfN7dDfX0eT7RpK/SvL0\nJE9Z++i7qL55xD8GSS6oqn27uwHtU1W/TvK91u8CNFd0gf/3wLuqaq8k84GLq+pJPZemadhNNzVP\n5xyPO5MsAM4BPpPkFoa+JNTE276qTktyLEB3QdB9022k/lXV8/quYRLZ1TMey4FfMPiS6csMBvt6\nyQa30CT5eXe7zAJIcgDwk35L0ijsppuaXT3SNLo+4ROAJwI/AHYA/mNVfb/XwjQtu+mmZvCPQTdO\nyPuB32JwHvHac4m36bUwjawLjN0YvHdXVtWvei5JI1g7wF6Si6tqn67tkqrau+/a+mQf/3j8JfCS\nqrqi70K00fZnMMLjfOApSaiqT/VbkkZgN90UDP7xuNnQn7uSfJrBkBuXcP9FQAUY/JPvbcCZwK5J\nzqXrpuu3pP7Z1TOLhob1/R3gMcD/xvH455xurKU9yv8sc5LddL/JI/7ZNXzmzt3A7w3NOx7/3PED\nBh/cN/ZdiGYmycMYHPXvUlWvS7IsyW5V9U9919Ynj/jHIMkzq+rc6do0WZJ8gcEH9COAvYHzeOBf\nbC/tqTSNKMlnGYzT85qqemL3QfBtv9zVOJwArHuZ+FRtmixnAguBb67T/mw8+p8rdq2ql3c3Y6Gq\n7u5undk0g38WJXk68AxghyRvG1q0DTCvn6o0A8uBY6vq0uHGJHcAfwF8opeqNBP3Jnko95/VsytD\nf7W1yuCfXQ8BFjD4PT9iqP2neGbBXLBw3dAHqKpLkywZfznaCMcxuFp+cZLPAM8Ejui1oglgH/8Y\nJNnFm67MPUmuqqpl61m2sqoeN+6aNHPdefwHMDir5ztVdVvPJfXOI/7xODHJb3zCtj5C4BxwQZLX\nVdXfDTcmeS2DLww14ZI8p5u8q/u5R3fx3Tl91TQJPOIfgyRPHZrdGngZsKaq3tFTSRpBkoXA54F7\nuT/o92XQhffvq+qmvmrTaLozs9bamsEV2Be2ftBl8PfEG3jPHUmex2CANoDLqurrfdajjZdkMfDh\nqnpZ37X0ya6eMUiy3dDsFsBTgUf2VI5mqKq+AUx1Qw/NPauA3fsuom8G/3hcyOB0sgBrgGuA5scE\nl2ZbkhPoTuVkcNC1D3BRfxVNBrt6JD1oJTma+6+ZuR241ivmPeIfiyRbAq8H1p5hcDbwtw4WJc2O\n7v/cXwGvAa7tmhcyuGL+3CR7V9UlPZXXO4/4xyDJx4EtgZO6plcD91XVa/urSnrwSvLfgIcBb62q\nu7q2bYAPMBha++CqWtpjib0y+Mcgyfeqaq/p2iRtHklWAsvWHUo7yTzgNuAFVfWdXoqbAN5sfTzu\n68YIASDJY7n/hh6SNr9fT3X/hKq6D7i15dAH+/jH5e3AN5Jc3c0vAY7srxzpQe/yJK9Z9/aYSV4F\nNH83PLt6ZlGS/YDrq+qmJFsBfwQcAqwEjqmqO3otUHqQSrITgxsd/YIHXnX9UAZXXd/QV22TwOCf\nRUkuAn63qu7oxgw5FXgTg5t67F5VjtApzaIkzwf27GYvr6qz+qxnUhj8s2j4C9wkH2XQt7iim7+k\n9bsASeqHX+7OrnndjZ4BDgSGx3jx+xVJvTB8ZtcpwL8muY1BX+M3AZI8DvhJn4VJapddPbMsyQHA\njsBXq+rnXdvjgQVV1fyYIZLGz+CXpMbYxy9JjTH4JakxBr8kNcbgl6TGGPyS1Jj/Bz5WPy70Durj\nAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10de929b0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "titanic['embarked'].value_counts().plot(kind='bar').set_title('Embarcation Port')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x10e81e668>"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEpCAYAAACUUUmzAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAF7JJREFUeJzt3X+Q3Hd93/HnCyFsl1+2Y1foJIEMFQWZBNEGAYVQMCU2\npoNMShRRYsTETGlraJiGTmWmU45plTEdfmRoA8xQE0Tq2FULxC6GgHHMr0mwjY0xloiDiGVsnWwP\nGMcGUxfJ7/6xX9mL0N3t3e3dru/zfMzs3O7n+93vvnZPet33vvfd/aSqkCS14zGjDiBJWloWvyQ1\nxuKXpMZY/JLUGItfkhpj8UtSYyx+SWqMxa/mJflSkh8lOW7UWaSlYPGraUnWA78GFPCakYaRlojF\nr9a9Efg68HFg+5HBJL+U5P8kuS/JdUn+c5Kv9S1/VpIrk9yT5JYkW/uWnZ1kb5L7kxxI8o6lfELS\nbB476gDSiL0ReD9wDfD1JKuq6i7gD4GfAE8B1gOfB24DSPJ44ErgPwKvAn4ZuDLJzVW1F7gI2FpV\nX01yEnDa0j4laWbu8atZSV4CPA3YXVXXA98D/nmSFcA/A95VVQ90Zb6r767/FNhfVX9UVYeq6pvA\nJ4Hf7Jb/DNiY5ElV9aOqumHJnpQ0AItfLdsOfKGqftDd/pNu7FR6vw3f3rdu//WnAS9Icu+RC/AG\ner8dQO+HxtnAbUm+nORFi/kkpLnyUI+alOQEYCuwIsmd3fBxwInAKuAQsBb4627Zur673w58uape\neaxtV9V1wJYkK4G3AruPur80Uu7xq1XnAIeBjcCm7vJs4Kv0jvt/CphM8neSPKsbO+IzwDOTnJtk\nZXd5fpJnJ3lckjckeXJV/Qy4D3hoKZ+YNBuLX63aDvxRVX2/qu48cgH+G73DNm8FngzcCfwxcAnw\nIEBV3Q/8OrANmOrWeQ+93xgAzgX2J7kP+Jfd9qSxESdikWaX5D3AU6pq+6wrS2POPX7pGLrz9H8l\nPZuB84BPjzqXNAz+cVc6tifSO7wzAdwFvA+4bKSJpCHxUI8kNcZDPZLUmLE81HPKKafU+vXrRx1D\nkh5Vrr/++h9U1amzrTeWxb9+/Xq+8Y1vjDqGJD2qJLltkPU81CNJjbH4JakxFr8kNcbil6TGWPyS\n1BiLX5IaY/FLUmMsfklqjMUvSY0Zy3fuarjW77hiKNvZf+Grh7IdSaPlHr8kNcbil6TGzFr8SY5P\ncm2SbyXZk+Td3fhkkgNJbuwuZ/fd54Ik+5LckuTMxXwCkqS5GeQY/4PAGVX14yQrga8l+Vy37ANV\n9d7+lZNspDcJ9en0Zi/6YpJnVtXhYQaXJM3PrHv81fPj7ubK7jLTtF1bgEur6sGquhXYB2xecFJJ\n0lAMdIw/yYokNwJ3A1dW1TXdorcluSnJx5Kc1I2tAW7vu/sd3dhsjzGZpJLU1NTUHJ6CJGkuBir+\nqjpcVZuAtcDmJM8BPgw8HdgEHKQ3GfW8VdVkVaWqMjExsZBNSZJmMKezeqrqXuBq4Kyquqv7gfAQ\n8FEeOZxzAFjXd7e13ZgkaQwMclbPqUlO7K6fALwS+Kskq/tWey1wc3f9cmBbkuOSnAZsAK4dbmxJ\n0nwNclbPamBXkhX0flDsrqrPJPnjJJvo/aF3P/AWgKrak2Q3sBc4BJzvGT2SND5mLf6qugl43jHG\nz53hPjuBnQuLJklaDL5zV5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pekxlj8ktQYi1+SGmPx\nS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjZm1+JMcn+Ta\nJN9KsifJu7vxk5NcmeS73deT+u5zQZJ9SW5JcuZiPgFJ0twMssf/IHBGVT0X2AScleSFwA7gqqra\nAFzV3SbJRmAbcDpwFvChJCsWI7wkae5mLf7q+XF3c2V3KWALsKsb3wWc013fAlxaVQ9W1a3APmDz\nUFNLkuZtoGP8SVYkuRG4G7iyqq4BVlXVwW6VO4FV3fU1wO19d7+jG5vtMSaTVJKampoa+AlIkuZm\noOKvqsNVtQlYC2xO8pyjlhe93wLmraomqypVlYmJiYVsSpI0gzmd1VNV9wJX0zt2f1eS1QDd17u7\n1Q4A6/rutrYbkySNgUHO6jk1yYnd9ROAVwJ/BVwObO9W2w5c1l2/HNiW5LgkpwEbgGuHHVySND+P\nHWCd1cCu7sycxwC7q+ozSf4S2J3kPOA2YCtAVe1JshvYCxwCzq+qw4sTX5I0V7MWf1XdBDzvGOM/\nBF4xzX12AjsXnE6SNHS+c1eSGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUGItfkhpj\n8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1BiLX5IaY/FLUmMsfklqzKzFn2RdkquT7E2y\nJ8nvduOTSQ4kubG7nN13nwuS7EtyS5IzF/MJSJLmZtbJ1oFDwO9V1Q1Jnghcn+TKbtkHquq9/Ssn\n2QhsA04HJoAvJnlmVR0eZnBJ0vzMusdfVQer6obu+v3Ad4A1M9xlC3BpVT1YVbcC+4DNwwgrSVq4\nOR3jT7IeeB5wTTf0tiQ3JflYkpO6sTXA7X13u4OZf1Ac2fZkkkpSU1NTc4klSZqDgYs/yROATwJv\nr6r7gA8DTwc2AQeB9y0kSFVNVlWqKhMTEwvZlCRpBgMVf5KV9Er/4qr6FEBV3VVVh6vqIeCjPHI4\n5wCwru/ua7sxSdIYGOSsngAXAd+pqvf3ja/uW+21wM3d9cuBbUmOS3IasAG4dniRJUkLMchZPS8G\nzgW+neTGbuydwOuTbAIK2A+8BaCq9iTZDeyld0bQ+Z7RI0njY9bir6qvATnGos/OcJ+dwM4F5JIk\nLRLfuStJjbH4JakxFr8kNcbil6TGWPyS1BiLX5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pek\nxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNGWTqRWno1u+4Yijb2X/hq4eyHakl7vFLUmMsfklqzKzF\nn2RdkquT7E2yJ8nvduMnJ7kyyXe7ryf13eeCJPuS3JLkzMV8ApKkuRlkj/8Q8HtVtRF4IXB+ko3A\nDuCqqtoAXNXdplu2DTgdOAv4UJIVixFekjR3sxZ/VR2sqhu66/cD3wHWAFuAXd1qu4BzuutbgEur\n6sGquhXYB2wednBJ0vzM6Rh/kvXA84BrgFVVdbBbdCewqru+Bri97253dGOzbXsySSWpqampucSS\nJM3BwMWf5AnAJ4G3V9V9/cuqqoBaSJCqmqyqVFUmJiYWsilJ0gwGKv4kK+mV/sVV9alu+K4kq7vl\nq4G7u/EDwLq+u6/txiRJY2CQs3oCXAR8p6re37focmB7d307cFnf+LYkxyU5DdgAXDu8yJKkhRjk\nnbsvBs4Fvp3kxm7sncCFwO4k5wG3AVsBqmpPkt3AXnpnBJ1fVYeHnlySNC+zFn9VfQ3INItfMc19\ndgI7F5BLkrRIfOeuJDXG4pekxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY\n/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4Jakxg8zApQGt33HF0La1/8JXD21bktTPPX5JaozFL0mN\nsfglqTGzFn+SjyW5O8nNfWOTSQ4kubG7nN237IIk+5LckuTMxQouSZqfQfb4Pw6cdYzxD1TVpu7y\nWYAkG4FtwOndfT6UZMWwwkqSFm7W4q+qrwD3DLi9LcClVfVgVd0K7AM2LyCfJGnIFnKM/21JbuoO\nBZ3Uja0Bbu9b545ubFbd4aNKUlNTUwuIJUmayXyL/8PA04FNwEHgfQsNUlWTVZWqysTExEI3J0ma\nxryKv6ruqqrDVfUQ8FEeOZxzAFjXt+rabkySNCbmVfxJVvfdfC1w5Iyfy4FtSY5LchqwAbh2YREl\nScM060c2JLkEeBlwSpI7gHcBL0uyCShgP/AWgKrak2Q3sBc4BJxfVYcXJ7okaT5mLf6qev0xhi+a\nYf2dwM6FhJIkLR7fuStJjbH4JakxFr8kNcbil6TGOBGL1HEiHbXCPX5JaozFL0mNsfglqTEWvyQ1\nxuKXpMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1JhZiz/Jx5LcneTm\nvrGTk1yZ5Lvd15P6ll2QZF+SW5KcuVjBJUnzM8ge/8eBs44a2wFcVVUbgKu62yTZCGwDTu/u86Ek\nK4aWVpK0YLMWf1V9BbjnqOEtwK7u+i7gnL7xS6vqwaq6FdgHbB5SVknSEMz3GP+qqjrYXb8TWNVd\nXwPc3rfeHd3YrJJMJqkkNTU1Nc9YkqTZLPiPu1VVQA1hO5NVlarKxMTEQjcnSZrGfIv/riSrAbqv\nd3fjB4B1feut7cYkSWNivsV/ObC9u74duKxvfFuS45KcBmwArl1YREnSMD12thWSXAK8DDglyR3A\nu4ALgd1JzgNuA7YCVNWeJLuBvcAh4PyqOrxI2SVJ8zBr8VfV66dZ9Ipp1t8J7FxIKEnS4vGdu5LU\nGItfkhoz66GecbV+xxVD29b+C189tG1J0rhzj1+SGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKX\npMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4Jakxj9pP55Ra4KfQajG4xy9JjbH4JakxFr8kNWZB\nx/iT7AfuBw4Dh6rqV5OcDPxPYD2wH9haVT9aWExJ0rAMY4//5VW1qap+tbu9A7iqqjYAV3W3JUlj\nYjEO9WwBdnXXdwHnLMJjSJLmaaHFX8AXk1yf5F90Y6uq6mB3/U5g1SAbSjKZpJLU1NTUAmNJkqaz\n0OJ/SVVtAl4FnJ/kpf0Lq6ro/XCYVVVNVlWqKhMTEwuMJUmazoKKv6oOdF/vBj4NbAbuSrIaoPt6\n90JDSpKGZ97Fn+TxSZ545Drw68DNwOXA9m617cBlCw0pSRqehZzOuQr4dJIj2/mTqvqzJNcBu5Oc\nB9wGbF14TEnSsMy7+Kvqb4DnHmP8h8ArFhJKkrR4fOeuJDXG4pekxlj8ktQYi1+SGmPxS1JjLH5J\naozFL0mNcc5dSXPiPMCPfu7xS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUGItfkhpj8UtS\nYyx+SWqMxS9JjfGzeiQtC8P6DKEWPj9o0fb4k5yV5JYk+5LsWKzHkSTNzaIUf5IVwB8CrwI2Aq9P\nsnExHkuSNDeLtce/GdhXVX9TVf8PuBTYskiPJUmag1TV8DeavA44q6re3N0+F3hBVb11hvtMAu/q\nbj4AfGcIUSaAqSFsZ9jGMZeZBmOmwY1jruWe6WlVdepsK43NH3erahKYHOY2k1RVTQxzm8MwjrnM\nNBgzDW4cc5mpZ7EO9RwA1vXdXtuNSZJGbLGK/zpgQ5LTkjwO2AZcvkiPJUmag0U51FNVh5K8Ffg8\nsAL4WFXtWYzHmsW7R/CYgxjHXGYajJkGN465zMQi/XFXkjS+/MgGSWqMxS9JjbH4JakxFr8kNcbi\nl6TGWPyS1Jix+ciGhUryGwOs9n+r6rOLHkaSxtiyOY8/yQ+By4DMsNpLq+oZSxQJgCSDvGP5nqp6\n02JnOcJMgzHT4MYxl5mmt2z2+IHPVdXvzLRCkv+xVGH6PBt48wzLQ2/ugqVkpsGYaXDjmMtM06kq\nL4t4AbYOYx0zmWlcM41rLjNNf1k2h3oAkjwZOAtY0w0dAD5fVfeOLpUkjZdlU/xJ3khvIpcv8MhH\nQK8FXgm8u6o+MaJcjwXOA15Lb8IF6OW7DLioqn5mJjM9mjONay4zzZBjGRX/LfRm+br3qPGTgGuq\n6pkjynUJcC+wC7ijG14LbAdOrqrfMpOZHs2ZxjWXmWbIsYyK/6+B51fV3x41/mTgG1W1YVS5pvuh\nM9MyM5np0ZJptsf2tRq/TMvpDVw7gRuSfDjJO7vLR4AbumWjck+S30zy8Gud5DFJfgv4kZnMtAwy\njWsuM01j2ezxw8OHdc7kF/+4O7L/EEnWA+8BzuCRb+yJwNXAjqq6dQwyBXjymGUCOAn48zHKNI6v\nU+j9exrZ6zRNLhjPf+cjfa3G5d/Usir+cZfklwCq6oejznKEmQZjpsGNYy4z/bzldKhnWkm+PeLH\nf1KSZ1TVD/u/yUl+ZYSZnpLkKV2exyT5jSSnjypPvyOvU5LfH3WWI7rX6Und6/SsUeVI8tQkx3c3\n7wFek+S/JvlX3Rkjo8r1miTHwSPfv1Fl6ZfkpUn+fpfnWUnekeTVI870hCSvA94IvCHJWf2HfpYk\nw3LZ45/hs3oCfKSqTl3KPA8/eLIV+APgbmAl8Kaquq5bdkNV/YMRZHoLsIPea/Me4E3AzcBLgP9S\nVReNINMHjx4CzgU+AVBV/2YEmf60qs7prm+h9338EvBi4Per6uMjyHQzsLmqHkjyHuAZwJ/SO3RA\nzfLu9UXM9VPgJ8DngEvoHWI9PIosfZn+ANhM7xMKPg+8gl6+fwx8s6r+3QgybQXeAdwEvBz4C3o7\n4L8MvKGqlmQndTkV/8+Ai4FjPaHXVdUTlzgSAEluBF5VVQeTbKZXZBdU1aeTfLOqnjeCTN8GXgCc\nANwG/L2qurP7G8nVVbVpBJluB75M730YRz5v6b30/pNQVbtGkOnh70+Sv6D3H/PWJKcAV1XVc0eQ\naW9VbeyuX0/vTLaHutvfGkWm7rG/Se+Hz+uAbcBzgE8Dl1TVl0eUaU+X4wR6f+9b0/3AXEmv+J8z\ngkw3AS/scpwCXFxVZ3a//X+kqv7RUuRYTp/VcxPw3qq6+egFSf7JCPIcsaKqDgJU1bVJXg58Jsk6\njv1Dain8rKoeAB5I8r2qurPL96Mko8q0EfhP9N55/Y6qmkryrlEUfp/+1+JxR/7wVlU/SPLQiDLd\nnuSMqvpzYD+wDrjtyPHiEaruJIqPAh9N8hRgK3BhkrVVtW5Emarve3Xk+/kQozvMHeCn3fWfAH8X\noKpuSvKkpQqxnIr/7cB90yx77VIGOcr93fH97wF0e/4vo/fr+aiOqVeSld27BB8+3tkdOx7Jf4iq\nuh94e5J/CFyc5IpRZenz3CT30fvPelyS1d3373HAihFlejPwiSSTwN8CN3a/VZ4I/NsRZQJ+/lNx\nu52JDwIfTPK00UTiiiRfBY4H/juwO8nX6R3q+cqIMn0W+LMkX6G3k/O/AJKcDDN+svBQLZtDPeMq\nyXOBn1TVvqPGV9L7MKaLR5DpqcBUVR06anwN8Oyq+uJSZzoqR4B/Dbyoqn57lFmOJcmJ9F6nvxxh\nhmcDz6S383YHcN2RQz4jyvOyqvrSqB5/OkleRG/P/+tJnkFvJ/D7wP8e1euV5Gx6v+F+q6qu7MYe\nA6ysqgeXJIPFL0ltGfWv0pKkJWbxS1Jjln3xJ9mS5AWjziFJ42LZFz+989X/Q5LPjTpIvyS70vtA\nuSU/l3g6ZhqMmQY3jrnM5B93RybJ84Gn0nsX5r8fdR4w06DMNLhxzGWmZVb8cepFSZrVsin+OPWi\nmczk1ItmGizHMip+p140k5mcetFMg+RYRsXv1ItmMpNTL5ppAMvprB6nXjSTmUZnHHOZaRrLZo8f\nHj6s49SLyyPTOE5zCE4HOZdc4zbNIYznv/Mlf52WTfEnSc3yZAZZZzHF6d8GYqbBjGMmGM9cZvp5\ny+lQz9VJ3pbeJ08+LMnjkpyRZBe9P6CMTP3i1IuvHFWWOB3knJTTQc4oYzglZJwOcvoMy2iP/3jg\nd4A3AKfR+8v5CfR+uH0B+FBVfXN0CX9Rku9X1VNnX3Poj+t0kINlcjrIwXON3ZSQcTrI6XMsl+Lv\nl95n3Z8C/HTUb95Kcvl0i4AzqurxS5kHIE4HOWgmp4McPNfYTQkZp4Oc1nKageth3ZsgDo46R+fX\ngN8GfnzUeOj95B8Fp4McjNNBDm4cp4SscjrIY1qWxT9mvg48cKw9jPTedDYKTgc5SCCng5yLcZwS\n0ukgp7EsD/VoZulNB/lAVX33qHGng5xG4nSQA+YYmykh43SQ02ew+BfXOJ5maiYzDds45jLT9Eb9\na2sLxvE0UzOZqYVcZpqGe/yLLMc+zfR4esdjR3KaqZkWlGmkpwiP4+s0rrkeRZmW/N+Uxb+EMkan\nmR5hpsGYaXDjmMtMRz22xS9JbfEYvyQ1xuKXpMZY/JLUGItfkhrz/wHN4o8zANxNgwAAAABJRU5E\nrkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10e1804e0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pd.cut(titanic.age, [0,10,20,30,40,50,60,70,80,90]).value_counts().sort_index().plot(kind='bar').set_title('Ages')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***Question 3 : Now we plot the number of passenger depending on the cabine floor (represented by the letter at the beginning of the cabin***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "titanic['cabin floor'] = titanic['cabin'].map(lambda x : x[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Floor repartion is :\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "T    77.540%\n",
       "C    07.181%\n",
       "B    04.966%\n",
       "D    03.514%\n",
       "E    03.132%\n",
       "A    01.681%\n",
       "F    01.604%\n",
       "G    00.382%\n",
       "Name: cabin floor, dtype: object"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_passenger = titanic['cabin floor'].value_counts().sum()\n",
    "floor_repartition = titanic['cabin floor'].value_counts()\n",
    "print(\"Floor repartion is :\")\n",
    "floor_repartition.map(lambda x: \"{:06.3f}%\".format(x/n_passenger *100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x10e180668>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAW4AAAD7CAYAAABKfn7LAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xl8XHW5+PHPMzPJZE/3Jd0GqEDLvq9tseJVCQouqAhY\nkFUrqBcvjt4fWL0ocbniglp3rwICLsgyirK1QIEuFErowlYCXZIutE2XtElmzvP745zANGSZTGZy\n5iTP+/Wa18yc9Tkz5zxz5nu+5/sVVcUYY0xwhPwOwBhjTN9Y4jbGmICxxG2MMQFjidsYYwLGErcx\nxgSMJW5jjAkYS9x9JCIXiMi/exg/Q0ReHMiYvPUeIiLPicguEblmoNdv8kNEfi8iN3qvfdm38klE\nGkTkzG7G9Wt7ReRGEdkqIk0iEhMRFZFI9tEWjoJL3N4XuVdEdnsf+O9FpMKnWN7xZavqbar6H2nT\nqIhMTRv/uKoeMtCxAtcBj6pqpar+2If1mzzL574lIgtE5LJ8LDtb/dleEZkMXAtMV9VxuY3MfwWX\nuD0fVNUK4GjgGOCrAx1AAH+ZpwAr/Q6ikATpOwxSrAExGXhTVTfncyW+fW+qWlAPoAE4M+39d4FE\n2vso8H3gDWATMB8o9cadAawHvgZs9ZZ1Qdq8tcCzwE5gHTAvbVwMUOBSb9mPec8K7PYepwAXA094\n8zzmjd/jjf9ERwxpy50GLAB24CbWD6WN+z3wUyAB7AIWAwf18Nl8yFvGDm+Z07zhjwApYJ8Xx8Fd\nzLsAuAlY4m3/PcCItPF/BpqAZm+7DksbdxawyotxA/Blb/go4H4vnm3A40DIG1cD/BXYArwGXJO2\nvHnAXcAfvGWuBI5PG3+s9z3t8uK6E7gxbfzZwHPeep8Ejuy0/3wFeB5oBSLe+w3e8l4E3tPN59vt\n/uGNP91b3w5v/MXe8FLgf4HXvc/vCd7eJ7v8znqI9RhguRfrncAdHdvOO/etBuDL3vzN3vQlaeOv\nAxqBjcBluPvq1C62+1vsv//c4g0/FVjqLXspcGoP++Yk4G/e9/1m2jIOwt0/38Q9Jm8DhnXahq/i\n7l/bgd91bENftzdtujOBvYDjbc/vefv4jqTtn/fi7revAJd3yjE/9D63jd7raKcc8xXc4+WP9HAc\n5C1P5jsR9zmgtMQNTATqgR+ljb/Z+8BHAJXAfcBNaR9qEviB9+HPwk2qh6SNPwL3n8aRuIn/XG9c\nxxf7B6Ac92Dc78v2prsYL3F77/c7GNJ3NqDI2ym+BhQDs3EPyI54fu/t0CfiHrS3AXd087kc7G3L\ne73lXuctu9gbvwC4rIfPdQFu8jrc276/Aremjf+M93l27LTPpY1rBGZ4r4cDx3qvb8L94SzyHjMA\n8T7fZ4AbvO0+EFgLvM+bbx5ukjgLCHvLedobV4ybAL/gLfMjQBtvJ69jgM3ASd68c3D3mY4DqwE3\nqU/yvsNDcJNsTdr33OWPIz3vH1O87+58L66RwNHeuJ96n+8EL6ZTvc+xt++sc6wd2/4lb/qPAe30\nnLiX4CahEcBq4Cpv3PtxE8thQBlwK90k7q72H29524GLcPfN8733I7uYNwyswD02y4ES4HRv3FRv\n+6PAaNyTgh922oYXvM9gBLAom+3t5rtMnzfG/on7MeBnXqxH4/7gzPbGfRN4Ghjjxfwk8D+dcsx3\nvG0qpZvjIK950o/k3GNA7pezG/cgUeBhvF9o3KSwh7QDD/cs+LVOH2p52vi7gOu7WdcPgZs7fbEH\ndvdle8MuJvPEPQP34Amljf8T3pkcbuL+ddq4s4A13cR6PXBX2vsQbiI+o6sDr5sDsy7t/XTchBju\nYtph3nZVe+/fAK4EqjpN903cM/epnYafBLzRadhXgd95r+cBD3WKZa/3eqa3XZI2/gnePph/3nEQ\npY1/EZiVtv98Jm3cVNxEfyZQ1Md9MX3/+CpwdxfThHDP7o7K4jvrHOtM3DO89G1/kp4T2YVp778L\nzPde/xbvhCbtc+hL4r4IWNJpmqfw/mV0Gn4KbuKLdLXsTtOeCzzbaRuuSnt/FvBqX7e3i/V0njfm\nbX8E90ciBVSmjb8J+L33+lXgrLRx7wMa0pbbxv7/bLo8DvL5KNQy7nNVtRL3QzoU968IuL9+ZcAz\nIrJDRHYAD3jDO2xX1T1p71/H/YVGRE4SkUdFZIuINANXpS27w7ocbkcNsE5VnU7xTEh735T2ugXo\n7kJsjTcvAN4y13VaVm/St+113LODUSISFpE6EXlVRHbiHiDw9mfzUdwD6nURWSgip3jDv4d7Bvlv\nEVkrInFv+BSgpuM78r6nrwFj09bfebtLvPLCGmCDekdEF3FPAa7ttOxJ3nzvmF5VXwG+iPtjsVlE\n7hCR9Gnf0sv+MQn3gO5sFO5ZW1fjMvnO1nWavvO2v07Putt/ajotu6/79X6xp8XS1f42CXhdVZOd\nR4jIWO8z3+DtW7fS8zH31vHajUyPl57UANtUdVen9U5IG/96p3HpMW1R1X1p77s7DvKmUBM3AKq6\nEPes9PveoK24ZzeHqeow71Gt7oXMDsNFpDzt/WTcsxiA23GLWSapajXu3xvpvNpuXmdjIzBJRNI/\n58m4Z13ZLGtKxxsREdwDpi/LmtQpjnbcz/RTwDm4Z6XVuGcn4H02qrpUVc/B/ev4d9x/MajqLlW9\nVlUPxC3L/U8ReQ/ugfha2nc0TN3aLmdlEGMjMMHbvq7iXgd8q9Oyy1T1T2nT7Pe9qertqno67uen\nuH9zu9LT/rEOt7y2s624xT5djcvkO0uPtattn9xNrL1pxC1q7DCpuwm7iAM6xZ4WS1f72zpgcjcX\n6r7tLfsIVa0CLuSdx1zn/XIj+bURGCEilZ3WuyFt/JRO49Jj6rx/dXcc5E1BJ27PD4H3ishR3hnL\nr4CbRWQMgIhMEJH3dZrnGyJSLCIzcC9k/dkbXon7S7tPRE7ETVg92YJ7gePAHqbZ1MP4xbhnBdeJ\nSJGInAF8EPeCU1/dBdSKyHtEpAi3qlMr7l/pTF0oItNFpAz3791fVDWF+7m04pa3l+EebAB4n+MF\nIlKtqu24F+4cb9zZIjLVSzTNuH8/HdxyyF0i8hURKfXO6A8XkRMyiPEpbzmfF5GIiJyDew2gw6+A\nq7yzYxGRchGp7XQQvsWr3z5bRKK4CbbjolVXeto/bgPOFJGPe3GNFJGjvX3yt8APRKTG29ZTvPX1\n9Tt7Creo7xpvf/lIp23vi7uAS0Rkmvd9X9/L9J33438AB4vIp7zt/QRukdb9Xcy7BPeHos77PkpE\n5DRvXCVu0WeziEwA/quL+eeKyEQRGQH8N+5Fx7xR1XW438FNXqxH4lZKuNWb5E/A/xOR0SIyCvda\nza1dL63H4yBvCj5xq+oW3AuGN3iDvoL7t+Rp76/XQ7gXoDo04V5E2Yh7sF2lqmu8cZ8Dvikiu7zl\n3dXLultwr7gv8v6Wn9zFZPOA//PGf7zT/G24ifoDuGdmPwM+nRZPxlT1RdyzlZ94y/ogbrXJtj4s\n5o+4/2CacP/ed9yo8wfcv4MbcK/uP91pvouABu/zvgq4wBv+LtzPfzdu0vmZqj7q/RicjXvR5zUv\n3l/jns33tp1tuBckL8W9Sn8hbrJo9cYvAy4HbsH9nl/Bve7QnShQ58XQhPuvobvqpd3uH6r6Bm5x\n0bW4NQeeA47yRn8Z9yL6Um/cd3Cva/TpO0vb9ou95XwCt6ZGn6nqP4EfA4/iHS/eqNZuZvkR8DER\n2S4iP1bVN3G/w2txf9CvA85W1a1drCvlbdtU3Osh673YAb6BW0uoGbf2VFfbczvwb9wL2K8CN/Zp\nY7NzPu4/y43A3cDXVfUhb9yNwDLc2iv1uLV8eoqpy+MgP2G7ZP/itGDzzmhvVdWJvU071IjIAtzP\n5td+x9JXIrIY9yLU7/yOJahEZBpu7Y1oV2XRJlgK/ozbDD0iMktExnl/0efgVs17wO+4gkZEPiwi\nUREZjvsv4D5L2oODJW5TiA7BrRe8A/ev+sdUtdHfkALpStyqkK/ilrt+1t9wTK4MqqISY4wZCuyM\n2xhjAsYStzHGBIwlbmOMCRhL3MYYEzCWuI0xJmAscRtjTMBY4jbGmICxxG2MMQFjidsYYwLGErcx\nxgSMJW5jjAkYS9zGGBMwlriNMSZgLHEbY0zAWOI2xpiAscRtjDEBY4nbGGMCxhK3McYEjCVuY4wJ\nGEvcxhgTMJa4jTEmYCxxG2NMwFjiNsaYgIn4HYAx/RGLJ8LAMGC49yjF3a+LvEf66/T3CrR1euz1\nHi1pj53Ajoa6Wh2wjTKmF6Jq+6MpLLF4ohKYAkz2nicBo3g7OY9Ie10FSJ5DSgJbgS3eI/11x6MJ\neA3YYEne5JslbjPgYvFEFDgYOBQ4gLcT9GTvMcy/6PqtFWjATeJr0x6vAa821NXu8i80M1hY4jZ5\nE4snIrjJ+UjgCOAwYBpusg77GJqfNgEr0h7PAS821NUmfY3KBIolbpMTsXgihJuYTwFOBo7FTdLF\nfsYVEK3ASvZP5svt7Nx0xxK3yUosnqjGTdCnAKcCJ+GWN5vccIB64ElgEbCooa62wdeITMGwxG0y\nEosnRgL/AbwbN1FPJ/8XBc3+1gELgQXAwoa62lf8Dcf4xRK36ZJXze5E4P3e43is3n+hWQ/8E7gf\neKihrrbF53jMALHEbd4SiydqcJP0+4AzcavdmWDYBzyKm8Tvb6irfcPneEweWeIe4mLxxBTgE97j\nWJ/DMblTj5vE7wOetrrlg4sl7iHIO7M+D/gk7kVFK6se3NYBtwN/bKirXel3MKb/LHEPEbF4YhTw\nMdxkPQMrrx6qVgC3Arc31NVu9DsYkx1L3IOYdwPM2cDluDVCrG0a08HBLRO/Ffir1RkPFkvcg1As\nnjgAN1lfDIz3NxoTAHtxi1J+0lBXu8LvYEzvLHEPEt6di+8HPu89W7m1ycbjwE+Au+02/MJliTvg\nYvFEFe7Z9eeAA30Oxwwe64GfA79sqKvd6ncwZn+WuAPKu9j4Rdwz7GqfwzGDVytwB/DDhrra5/wO\nxrgscQeMV5Xvy8AVQLnP4Zih5X7gfxrqapf4HchQZ4k7ILwLjl/BveAY9TcaM8T9G/hmQ13tIr8D\nGaoscRe4WDwxFbgBOB+rzmcKy7+A6xvqapf6HchQY4m7QMXiieG4CXsubh+JxhSqe4H/11BXW+93\nIEOFJe5uiMhI4GHv7Tgghdu3IMCJqtqWj/XG4oki3BoiN2CNPJngSAG/xE3g2/wOZrCzxJ0BEZkH\n7FbV7+dzPbF44lzgu8C78rkeY/JoG+5Jx/yGutqU38EMVtZeRQGIxRPHxOKJR4G7saRtgm0EcAuw\nPBZPzPI7mMHKzrgzkK8z7lg8UQZ8G7ga+xE1g9NdwJcb6mrX+R3IYGLJwiexeGI2bpvJX8C+BzN4\nfRxYE4snvhSLJ6wZhhyxhDHAYvFEVSye+CXwEHaLuhkayoAfAI/G4omYz7EMCpa4B1AsnqgFVuK2\nLWJnH2aomQXUx+KJy/0OJOisjDsD/S3jjsUTw3BbXLswl3EZE2D/BC6zzhyyY4k7z2LxxEm4jfTE\nfA7FmEKzHfh8Q13t7X4HEjSWuPPEuxBzLW6tEbvz0Zju/QaY21BX2+p3IEFhiTsPYvFENfBH4IN+\nx2JMQCwBPtpQV7ve70CCwBJ3jsXiiSOBvwJT/Y7FmIDZDHy8oa52od+BFDqrVZJDsXjiU8BTWNI2\nJhtjgIdi8cQX/Q6k0NkZd47E4on/Bm70Ow5jBonbgMsb6mr3+h1IIbLE3U9eJ70/wW3RzxiTO8uA\nsxrqarf0OuUQY4m7H2LxRAnwJ+Bcv2MxZpB6GfiPhrraBr8DKSSWuLMUiydG4DYgf5rfsRgzyDUC\nH2ioq13hdyCFwhJ3FmLxxGTgAWCa37EYM0Rsx03ei/0OpBBY4u6jWDwxBXgcmOR3LMYMMbuAsxvq\nah/zOxC/WXXAPojFE+NxuzOzpG3MwKsEHojFE+/xOxC/2Rl3hmLxxChgAXCYz6EYM9TtBt7dUFe7\nzO9A/GKJOwPeLeyPAMf6HYsxBnA77j6toa72Zb8D8YMVlfTC614sgSVtYwrJaODfXvHlkGOJuwex\neKIIuAer8mdMIYrhlnlX+x3IQLPE3bMfA2f6HYQxpltHAvd6N8MNGZa4uxGLJy4FrvI7DmNMr2YC\nv/M7iIFkFye7EIsnTgQeA6J+x2KMydjnG+pqf+p3EAPBEncnsXhiDPAMMNHvWIwxfdIGnNpQV/uM\n34HkmxWVpInFExHgz1jSNiaIioE/e51zD2qWuPf3PdzyMmNMMB3AECjvtqISj3cb7YOA+B2LMabf\n/rOhrvZmv4PIF0vcQCyeqALqgcl+x2KMyYl24LiGutp6vwPJBysqcf0vlrSNGUyKgPmxeGJQ/oMe\n8ok7Fk+cAVzmdxzGmJw7FbjU7yDyYUgXlcTiiSjwPHCw37EYY/JiG3BIQ13tVr8DyaWhfsYdx5K2\nMYPZCNzaYoPKkD3jjsUTY4FXgXK/YzHG5JUCZwymnnOG8hn3V7GkbcxQIMBPY/HEoMl3PW6IiIRF\n5FcDFcxAicUTk7AGpIwZSg4HPu53ELnSY+JW1RRus4mDzfVYA1LGDDU3DJaz7kw24hERuUVEThSR\n6R2PvEeWJ7F44iDgEr/jMMYMuGnAJ/wOIhciGUzzSe+5Nm2YAgfmPpwBMY/MttsYM/h8BfiT30H0\n15CqVRKLJ2qAN4Cw37EYY3zzgYa62gf8DqI/Mjrz9IpG3u29fVhV1+QvpLy6BEvaxgx1/wUEOnH3\nWsYtIhfhtpp3tPd4SEQuyHdguea1WfAZv+Mwxvju3bF4YorfQfRHJmfcXwaOU9UmABEZB/wLuC2f\ngeXBbIJbLm+MyR0BLgC+7Xcg2cqoakxH0u78OmAu9zsAY0zBCFypQbpMEverIvINEanxHl8H1uY7\nsFyKxRMjgXP9jsMYUzCmx+KJY/wOIluZJO6rgENwW9F7HjgUuDKfQeXBx7Ebbowx+7vQ7wCy1Wvi\nVtXNqvpJVR3lPc5X1c0DEVwOvd/vAIwxBeeTQb2TMpNaJSIiV4rIn73H5SISmF4lvJ7bz/A7DmNM\nwakBTvQ7iGxk8mvzXeA84O/e4zzgO/kMKsdOBKr8DsIYU5BO9zuAbGSSuN8HvF9Vb1PV23BvfQ9S\n0cN7/Q7AGFOwZvgdQDYySdyC2zZJB/WGBYUlbmNMd04LYofCmdyA8y/gnyLye+/9HAJyu2gsnqgE\nTvI7DmNMwRqJ22rgKr8D6YtMEvd1uNX/PuK9vxv4Zd4iyq1jsJYAjTE9m8FgS9yq6gA/9x5Bc5jf\nARhjCt5pwC/8DqIvuk3cIvI99i/b3o+qXpeXiHLLErcxpjfv8juAvurpjHv3gEWRP4f6HYAxpuBN\n9juAvuopcVep6rUicp6q/nnAIsotaw3QGNObcbF4oqihrrbd70Ay1VN1wPd4z18diEByLRZPhIFJ\nfsdhjCl4IWCi30H0RU9n3BtFpB44QESWdB6pqoV+q2gNVqPEGJOZycBrfgeRqZ4S2znAscCtuF39\nBE2F3wEYYwIjUOXc3SZuVW0HFotIraq+NIAx5UqJ3wEYYwJjbLYzishY4GbgZGA70AZ8V1XvzlFs\n75BJs65BTNpgidsYk7nibGbyWkr9O/CYqh6oqscBnyTPZeaDuQzYErcxJlNFWc43G2hT1fkdA1T1\ndeAnOYmqG5a4zZAyMrV57WEtLS8VR6pDUUlKNFIm0XCJiIg4oqRwSKHqoJLC0RQOjigOHQ8HRXGk\n45WKgwPueFFUVBzwXvu9vSYzrRrJ9r6Vw4DluYwlExknbhEpTp9eVVvyElHuWOI27zA+sts5s+Z7\n0c0vHdo6ZfvxoR0Vo2gp1tIi3b23VNpSJSENlYTC0dJwWagsUpUsDVdKabg8XBwujRaFisvDEqkS\nQsNFpDyT9Tmoo2jSwUk6aFLRlOM+ko44Ha9TDo7jvXYccVIp1PGGvfWcEke99+r+uDjqiPucwlHv\nx0UdHFLi/sh0/Ni89cMjKu7Pzts/NN5zyEFFRUOKikLIQUOKhhRCuK/DCiFFw0BYIcx+r4nw9rP7\nWgLTkuhjuViIiPwUt43vNlU9IRfL7EqviVtEPox72j++YxDurfDhfAWVI4GpTG8Gzgt6wEH3tu0c\nfutRq17839INB1/+L115xvPFsc1jTly/cfxp7dsqJhyChEars2ers2/zRif5RrMmm5KO82YxTksp\npMYBkbBE9kbD5dvLwhXNZZGqPWWRqtbySFWyNFKlpeHycDRUWlwUKikNhyKVIcLDQ4Sr39FzVHcN\nSnTb0ETwBOWHq5TiLVlu4krgox1vVHWuiIwCluXkA+xGJmfc38PtbPdpr8GpoMj2izCDmsg6Hf3y\np3duPnVEKrXsq+eOPLG0Ndn+mQef2DfjmcePEBi+o3rqqvUTztiybcS0MamS2OmkJVxVJ6nOttec\nZNOW1mTj3r3JzaGtba9VofumAKO7XSuSjIbLt5WGK5rLIpW7yyJV+8ojVe1lkSqnNFwRiobLiopD\nJaVhKSoPS3gYyAgRKfSTo16FkBBIcZjQOy/+FdYP111ZzvcI8JCIrAPe9IZ9LjchdU9Ue/6URGRJ\nAG62eYdYPHEAsNbvOEzhuSHyh4WfiTwwC+CJ0pLnPzt29BREqkvadPfFDzrPnFGv00PqJuF9xcM2\nbaw5/eWmcScW74uOOIweikjU2dfspDav12TjdifVmNTUmyXq7BkJyclAtI9hanGoZEdpuGJHaaRq\nV3mkam9ZpLKtLFKVKg1XSkm4PFIcKikpChWXhyRSLchwEbHiwexdNrFuxm+ymVFE9gD34bb9vwXY\nA8xX1TtzGN/+68wgcX8V2AHcCezrGF7oZdyxeKKcwdFQlsmxo+SVl+6J3nBwx/vno8UvXjh+7AgV\nGQ1Q3K4tn37YWXrmczotpIzpmM6RSOuW0Ue9sL5m5p6dVQccoKFwRk0qqKqjzo4Nmmza5KQa9zjJ\nTaLOjkp03zjQ8b0vITMRKdpdEq7YXhap3FkWqWp5q/jGTfThaLi0uCgULQ9LpNIrp6/M1boHgXMm\n1s24N5sZRWS3qg7oDX+ZJO704pGObstUVQv+b1wsntgDlPkdhyk0qmujF+wICcM7hqwtirz+kQnj\nwymRt+rfFiV13wWPOovf94weHFbekWB3VUx8df2EWeu3jDpqeDJSdhhZFG2otu3W1JZ1TrJxu5Ns\nbNPU1mJ1do+A9klARhdAsxUi3FoSLt9eGqloLotUdRTfJEvDlVoaqQhHQ2VFRaFoWSRUVBEiPAwY\n/o5y+sHjhIl1M7IqlxaRFFDvvX1NVT+cu7C6WWdviTvIYvHEa0DM7zhM4Xm8+JrFk0Jb9+vWrikc\nbjp74vjdraHQ1PThkaS2fuIxZ8nZS/TAsDKhq+W1R8p2NI47edXG8afSUjZuGiLDu5ouU6qqOLua\nnFRTo5Ns3O2kNqmmtpeje8eBU0Nm/cXmlCCpaLhsW2m4ork0Urm7PFK1ryxc1VYWqXJKIxUSDZcX\nFYdKSiNSVBGWcDVu8U229aMH2riJdTM2ZTNjQZ5xB1ksnngK9zZUY/bzjcjvHpsTeXBm5+E7QqHt\n759Us3FPKPSOTjjCKW0/7wnn6Q89rbGI033Lk4qkto2Ytmr9hFnbtg87uMYJF+e0oX7V5D5NbV3n\nJBu3OqnGVk1uKVJn1zBomwhU53Jd/VUUijaXhiu2l0Yqd5WHq1rKItVtZZFKpzRSSUm4PBINlUQj\nbvFNlbgXZEt9CLNlYt2MrP/dFFTiFpGHVfU9IrKF/a/zdhSVjOlyxgISiyd+DVzqdxym8BwjL794\nd/Trh3Q1bo/I7rMm1by0LRw+tqvxIUeTH3lSn/7IImdSxGFKb+vaWzJq/foJM9duGnNcWVtx9eHk\n8SKiOru3OMnNG53Uxp2abHKc1LZStGUMOJMo/Cq8hKWopSRcvq0sUrGzzE30+8ojVanSSKW+VZ9e\noqXhUFFViNAwEcnFD9XzE+tmHJXtzIWWuMeraqOIdLljerd1FrRYPHElML/XCc2QIzjOq9ELd4aE\nYV2Nb4O2cybWLF9fFOn2H1vI0dQHF+vijz/ujCtKZdZpRypU3LJpzHEvbKg5vXVX5eSpSChnFyd7\noppq19S2dU6qaYsmN+51UlvC6jRXo60TcHs6D6QQobZouHxbaaSi2U30lXvLI9XtZZFKLQlXhErC\nZcVFXvFNSMLVwAgR6VzMdOvEuhkXZRtDQSXu/SZyy6k6zk7WqGoyr1HlSCyeOI48V4Q3wbUoevWS\nCfJmt1VdU5A6v2bck6ujxTN6Wo6oOmct1cXnL3RGFyeZ2tO0nTVXHfDi+gmzmraOPHxUKlwyjXcm\nlbxTZ+92txrjxmYn1dRRjXE0pCaRZeNLBcyJhsq2l0YqdpSGK3eXR6r2FodLf/eBX3ztl34H1heZ\n1Co5HfgT0IJbTFICfFJVn8x/eP0TiyeKgZ30vQ6tGQJujPxm4YWRh2f1Nt2VY0cvfLKstNfpUNX3\nLdfFFz7iDI8m6bIYpietRZVbNtac9lLTuJNDe0tGHYZIVV+XkUuqTkqdHes12bTZSTW2eNUYq9DW\n8aBZN4NagN597Z33L/A7iL7IJHGvAK5R1YXe+xnALaqadZnQQIrFE0uAvLUZYILreHlx9V+i35iW\nybT/PWrEwnsrK3pP3gCqOnuFLpnzkFNV2k5Gy+/MkVD71pFHvrBhwsydO6oPmqKhSCyb5eSLautO\nTW5Z76QatzvJxnZNbY261RiTkwE/LjBmywGGXXvn/buymVlEzgXuBqap6pqcRtbTejNJ3J2TdFfD\nClUsnvgZ8Fm/4zCFR3CctdELd4lkVhPj5uHVj/22uur0vhRnzKx3ln7m305ZWRvvqKXSF7vLxjds\nmDDz9c2jj6lqL6o4nAKtZqeqqs7OjZpqavKqMaKp7RVuNUatgYJrdGrptXfen/Wd4SJyJ243iY+o\n6tdzF1Yv680gcX8feFZVb/Pefwo4WlWvG4D4+i0WT1yA2/2aMe/wVPTzS8fLtoz/kd1aVfnUd0YM\nO76vifNudi5JAAAWPUlEQVTUVc4zlz/gFJe3ckTfo9xfMlyys3HcSasax5+a3F1ecygSGtXfZQ4E\n1fYWrxrjNifZ2KqpLUXq7Bru3Wzk112cddfeeX9WHaKLSAXwIvBu4D5V7XPxWLZ6qlXSUQ1QcK86\nt3qjosDWIFQHBIjFEyOAzQSgKpQZeDdFfrXw/MijmRWBeP5RXrbsK6NHTkekz3flnrTGefaKfzqh\nyn3k5B+rIs72YQevXj9h5pZtI6aNd8LRAUseuaTOrk1OclOjk2zcqakmdVLby7xqjBPJ77Gbdfm2\niFwAzFbVS0XkSeBqVX0mp9F1t+4eEneP9VODUB2wQyyeWAi842YLY06SVavujN44va/zLSotqf/s\n2NGTVKTL6oS9Oe5lZ8VV/3Cc6haOyWb+7uyNjmjcWHP6K01jT4i2Rocfns2PSyFRTbZqatt6J9W4\nxUk27tPUlohbjbFtItCvu1NxKy6MvPbO+7OqJSci9wM/UtUHReQaYLKqfrmfMWW27sF852SHWDzx\nJeAHfsdhCk8IJ/Vq9MI9IvS5BscLxcUvX1AzttoRyfrf55Frnfq59zttw/dwXLbL6E4qFNm3efSx\nL2ysOb1lZ2XsQA2F89oP4kBTp+VNJ7Vpg5NsbNZkU8pxtpXydjXGTIqy7rj2zvvPz2bdIjICWI/b\nGmBH/wQKTNEBSKqZlHFPAr4LHEVarzKqmtENB4UgFk9MBN6g8C6MmALwdHTusnGy/fhs5m2IRN74\n8MTxJEUm9yeGw153Vn7+Pqdl5K781YDaWTn55fUTZm3YOvLIkclI6fRsGsUKArfN9O3rnWTTZvdm\no80hdZqr0NYa0PQ208+59s77s20R8ArgOFW9Mm3YQuB6Vc1Jbzo9rj+DxP0gcAfwZeAzuDU0XlXV\nb+Q7uFyKxROPAT3eSGGGpu9EfrHgE5GFZ2Q7/6ZweFPtxPE7W0OhfrdJcsg6XX31faldo5s5QfJ4\notFWVL6tcdwpqxvHnSItZWMPIze3jhc8t830Leud5Lr1qX1Pf+jaO+9vy2Y5IvIo8B1VfSBt2DW4\n1QLzXostk8T9rKoeIyL1qnqEd7voU6p6Uo8zFphYPHEV8HO/4zCF55TQypV/Kv5Wv6rrNYdCO94/\nsWbd7nCo37VGAKZu0BevuTe1bewOTs5nAgdwJJTcNuKwlesnzNy+Y9i7JjmhooPyub4C8eO582d/\nwe8gspVJ4l6sqieJyFPAJ4BNwOogFZUAxOKJSmAdBdZ6mvFfmFTylehF+0ToV3sTLSJ7aifWrNka\nCeesvPqAJn3lmntSm2u2cbIMUFOuLaVj3lg/YeZrm0cfW9FWXHU4IoPxzuOj5s6f/bzfQWQrk8T9\nPeAmoBb3Al8r8BdV/WL+w8utWDzxXeC//I7DFJ4l0c8tGyM7sirnTtcGbR+eOP6ZN4qKTslFXB0m\nb9a119yTapy0lZNlAKu2JsPR3ZvGHL9yY83p7bsqJr4LCQ2GW92XzZ0/O9B3U/epVom4F2CqVPWF\n/IWUP95FyrVkdsXZDCH/W/TzhR8NP96n+tzdccD5VM3YRSuj0ZxfU5mwVV//wj2pdVM2c7Jk1tl3\nzihoc/VBa9ZPmLX5zRGHjU6Fo9PSO1IOkCvmzp/9K7+D6I9MzrjPBJaqarP3fhhwrKo+MgDx5Vws\nnrgVuMDvOExhOT1UX39r8U05KZ/u8Lmxoxc+nknjVFkYt03XfeGeVMOBTZwsPp2ItBZXb95Qc9pL\nTWNPKtpXMvIw3DsJC90G4KC582e39jplAcvo4iRuolbvfQhYpqpdNjJf6GLxxDHAcr/jMIUlQrL9\n5ein20Ry28/j10eNWPC3yoozcrnMdGN26Iar7029evAGThIfW8F0JNy2ZdRR9RsmzNjTXHVgTEOR\nflWPzKOr586ffYvfQfRXJon7OVU9utOwwDQy1ZVYPPEIbvsCxrxlafSzz4yW5pzfCHPLsOrHfzGs\n6tR81pse2ayNV9+XennaOk6UtPst/LK7vGbt+gmz1m0ZffSw9kj5YYgMaLFON7I62+7UGTDAHapa\nl9PI+iiTxP048GVVXey9Pwm4WVVPHYD48iIWT8wEFvodhyksNxf9dOGHw4vyUrRxR2XF098aOfxY\nRPLaMcHwXbr58/c5qw9/XU8QKIjb3dsjpc1NbkfKzp6y8dNw7zr0Q1Zn2370cNObTBL3KcDfgJXe\noOnAR1T16TzHllexeOIu4Dy/4zCFY2ZoxfN/KP7Okfla/r/LSpdfO2bUwQNRFly9W7d8LuGsPHqt\nHi/0r5pjLinibBt+6Mr1E2Zu2z780Jx3pNyDV4Dpc+fPbu/rjIFM3AAiMhzoqN70lKpuz2tUAyAW\nT0wBVhOsRt9NHhWRbHsp+umkSP7OVBeXRFdePm5MjbrHVN5Vtui2q/7h1B//sh4j9L09lnzbWzJy\nw4aaGWs3jT2+pLV42BF57Ej5vLnzZ/8lmxm7KCq5SVXvzE1Y2RkSjUx1JxZPfAO4we84TOFYFr1q\n+SjZmdcL76uKi145v2ZchSMyLp/rSVexV3dc8U9nxUkv6tFSoDehpULFLZvHHPvChprTW3dWTp6K\nhHPVkfKTc+fPPi3bmQN7xj1YxeKJMmANMMnvWExh+HHRTxZ8KPzUGflezxuRyPpzJo5PJXtpPjnX\nyvZp82X/cp47bZUeKf1vFjWvmitjbkfKo44YmQqXTM+yI2UFTps7f/ZT2cZhibsAxeKJ84Hb/Y7D\nFIbZoeXP/7b4+3kr5063JRzactbEmm37QqEB7/ygtFV3XfKgs3xmvR4ecjtKKWhtRZVvbhx/6prG\ncSeH9paOnt6HRrFumTt/9tX9Wbcl7gIViyceBM70Ow7jv2LaW1+MzlGRgalS1xyS5g9MnPD6rnBo\nQH4sOou26Z45DzvLZq/Q6SFldO9z+M+RUPLNkYe/sL5mVvOOYVMnayhyQDeTvgocOXf+7Jb+rK+L\nMu4HVDXen2X2lyVuIBZP1AArgED03Wfya3n0yudGyK6je58yN/aKtNROHL9qSyTS77ZSslXcrnsv\nfMRZ+h/P6iEhJVDtkewpG/v6+gmzXt88+pjK9qLKjo6UHWDW3Pmzn/A7vnywxO2JxRMfAu7xOw7j\nv1uKfrTg7PDiMwZyne3Q/pEJ45c2FBf5en9EUVL3nb/AWfKBZfqusJKri4MDJhmO7moae+LKxnGn\nJD59+yU3+h1PvljiThOLJ36G21GEGcLODC1b8eviHwz4ncEOOBeNH/vE8yVR3/tHjaS07ROPOYvP\nXqIHhB2C1uXZcuCUaWtWZ9VJQhBY4k4TiydKgWW4NxmZISpK27410YtFxJ+2P64eM2rhgvKyvNzB\n2VfhlLZ/dJGz+NyndHLEoVDbH0m3Gzh22prVL/sdSD5Z4u4kFk8cCSzBxwZ7jP+ejV6xYrjs9q09\nnm+OHL7wz5UVMwul2dSQo8lzn9LFH33CmVDkEPM7nh5cNG3N6lv9DiLfBqRHjSBpqKt9Hghsl0Ym\nN55ypvt6d/ANb26f9bkdzYtQTfkZRwcnJJG/nRY67aL/Ck+69d2hJ9vCvOp3TF34v6GQtMESd5ca\n6mp/AfzE7ziMf+5Jner73YWf3bHz9Ovf3L4U1YJpO9oJSfjek0OnXvRf4QN+d2boqdYIhVIk8RRD\n6PqUFZV0IxZPhIH7gff7HYsZeCW07l0dvSQsQl5b88vEQ2Wlz35pzKh3FWRHBar63md1yUWPOMNK\n2hnwG4k8LwGnTluz+k2f1j/gLHH3IBZPVAGPAYFte9xk77no5c8Pkz2+3BjT2dKS6KpLx40Zp/41\nidqrd69wllz8kFNR2jagF/c349YgWTuA6/SdFZX0oKGudidwFvC637GYgbfYObRgWsE8YV/r9Ls2\nNm0PqTb6HUt3Hj0qdOKcayPTf3J2aNmeKAPRL20LcPZQS9pgZ9wZicUThwKLgII92zG5d1Zo8fKf\nFf+ooLroW+c2TpVsF4n5HUtvTlntLL/8ASdSsY98/GtpA86Ztmb1A32ZKe329SIgCfwBt2MYJ/ch\n5o8l7gzF4okTgH9R4C2qmdwppbVlVfSSIhF/OuPtzla3cao394ZCh/odSyZOeMl57sp/OFTtJVfN\nCCSBj09bs/ruvs6Y3mCUiIzBbWBukap+PUexDQhL3H3gdTT8IAFoTc3kxoroZfXV0pLT3t9zYWdI\nmj8wsaZhZzgcmOsvx7zirPhswkkNa6E//2Ic3LraWbXo2bmlPxE5EFgKjNIAJUMr4+6DhrraZ4Ez\ncC+ImCFgqXPoNr9j6EqVo9UPr9t48NhkcqnfsWTq2amho674QuTY//lkqH5bBcuyWIQCV2abtLtc\noOpaIAyMydUyB4Il7j5qqKt9ATd5F+xFIpM7f0+dVnhV8DwlqqUPrNt4zIFt7U/6HUtf1B8QOuKq\nqyPHz/tUeNWWKpZkOFsKuGLamtW/zmdsQWFFJVmKxRPvAh6BwDXAY/qgnL27X4heWiJCxO9YuqOg\nF48f89jykpKCaN+krw5er2uuvjfVPKaZEwW6usW/BfjktDWr7+vvugZLUYkl7n6IxRMHAP8Gpvod\ni8mf+uilKytl72F+x9GbL40ZteCh8rIz/I4jWwc16svX3JPaOm47J8nbpQFbgQ9OW7P66Vyso9PF\nydHAbbgdoNvFyaEkFk+MAP4MzPY7FpMfvy367oLZ4efO8DuOTHx75PCFfyqgxqmyEWvSV6+5N9U0\n4U3GCtROW7P6pVwtu4vqgH8EfmDVAYegWDwRAX7MEGorYSg5N/TEsh8W/8y33mn66lfVVYt+PLz6\nJEQKtngnA09X79YPPTH3hS1+B1KILHHnUCye+BzwIyjc8lDTd+Xs3fVC9NIyEcJ+x5Kpv1WUL/n6\nqBFHIjIgfWfm2K3AZfVz6gumca1CY4k7x2LxxHtwi07sRp1B5IXoZ1ZVyL5AdbCxoKz0uavHjDoQ\nkSq/Y8mQA1xfP6f+234HUuisOmCONdTVPgycCKz2OxaTO884Bweu7v4ZLXuP/r/GTRtEdavfsWRg\nI3CmJe3MWOLOg4a62leA44Ff+B2LyY17UqeW+x1DNo5tbZv2lw1Nu8KqG/2OpQcJ4Kj6OfWP+h1I\nUFhRSZ55vcf/BhjldywmexW07KyPXlYepHLudBsi4Y0fnFjT2i5ygN+xpGkDrqufU/8jvwMJGjvj\nzrOGutp7gSNx63ubgNpNWVULJTmrljbQJiRTNQ++saGy1HEKpQjvJeBkS9rZscQ9ABrqahtxe9L5\nEmBXygNquTM1cOXc6UY6zqiH39gwoTqVes7nUP4AHFc/p/5Zn+MILCsqGWCxeOII3OpOBdGzisnc\neeEFS75X9MsT/Y6jv1qFfWdPrHm+KRIZ6G1ZD3y2fk79/QO83kHHzrgHWENdbT1wHBDHbYPBBMS/\nUscfokqg7rDrSlQp+ee6jcdObWtbNECrVOBnwHRL2rlhZ9w+8to6+RnWIXFgrIpesqZMWgPRgUFv\nFPQz48Y8tqw0r41TPY97lh2oFgwLnZ1x+6ihrva1hrraDwAfA97wOx7Tu+ecgzb5HUOuCMjvmjbP\nev/uPQvysPhdwH8Cx1rSzj1L3AWgoa72r8A04Ebs4mVBu9c5NYi3kPfoe1vePOPC5p0Lyc3fb8Vt\nce/Q+jn1N9fPqU/1ZWYRSYnIcyKyQkSWi8ipOYhp0LGikgITiycOBL4JnI/9sBacYeza8Wz0ymqR\nLtuNDrTfVlcuunn4sP40TvVP4Kv1c+pXZBtDp2ZX3wd8TVUD2c54PlniLlCxeOJw3DPwc/yOxexv\nVfTiF8uk7RC/48iHv1eUL7l+1IgjECntw2yLga/Uz6lf2N/1d0rc5wEXqOq5/V3uYGOJu8DF4okT\ngW8BZ/odi3HdUfw/C08OrR60Z4GPlZasmDt2dAyR6l4mXQN8rX5OfZ97W+9OWnvZJcB4YLaqPpOr\n5Q8W9le8wDXU1S5pqKt9L25HDU/5HY+B+1KnDLpy7nQz9+476o+NmxpFtbu2sNcDlwGH5zJpe/aq\n6tGqeihubas/SIA7hcgXO+MOmFg8UQt8BZjhdyxD1XB2blsevWr4YCznTvdqUaThoxPGR1IiHf2q\nrgG+C9xWP6e+LR/r7KJPyE3AEaoa6LtWc80Sd0DF4onjcatbnYd13DDgVkfnvFwq7e/yO4582xgJ\nN354wvjlLaHQb4C/18+pz2vC6FTGfSjwBDBWVftUO2Wws8QdcLF4YhJwDXA50FuZpMmRu4q/sfDE\n0IuDtpwbtz/GvwI/Yl7zgBXRpZVxg9vj+9dUNTFQ6w8KS9yDRCyeqAAuBb4AFFLTnYPSp8P/fvqb\nRb8/2e848mA78EvgFuY1r/c7GNM1S9yDTCyeCONe1LkE+CBQ7G9Eg9NImrc+U/LZwdTG+lLgt8Af\nmNdsbegUOEvcg1gsnhgBfAo3iR/rcziDzpronFdLpP0gv+Poh5eA24Hbmdf8st/BmMxZ4h4ivOZk\nLwEuAMb4HM6g8JfieY8dH3pppt9x9FEjcAdusl7mdzAmO5a4h5hYPFEEfAD4JHA2UOlvRMF1Sfif\nT3296I+n+B1HBppxLzTeDjzKvObAN0071FniHsJi8UQUeC/wUeBDwAh/IwqWMWzfsqRk7mi/4+hG\nK24nvLcBCeY1W+Nlg4glbgNALJ6IAKfhJvAPAoO+jnIuvBj99NqoJA/0Ow5PE27fpv/CTdbNPsdj\n8sQSt+lSLJ44FHgf7q32M4Fh/kZUmO4uvuHxY0Kv+HUXaxuwCHgAN1k/z7xmO6CHAEvcplexeCIE\nHAO823vMwMrGAbg0/I8nry+6daDajE4Cy4EF3uMx5jXvGaB1mwJiidv0mVeschxuEp8FnACM9DUo\nn4xj26anSz4/Nk+Lb2f/RP0E85p352ldJkAscZuciMUTU3CTefpjMN2g0q2Xop9+rViS/b1bdTNu\n/4wrvOfngVXMa85LY04m2Cxxm7yJxROTcRP4scDRwMG4t+MX+RlXrt1T/P8ePyq0NtNy7nZgNZ2T\n9LzmpnzFZwYfS9xmQHnFLDHcWisHd3qeTADbiL8ifP+irxXdflraIMU9g37De6zFbThpBbCaec3t\n+YxHRMYBP8QtwtoBbAK+qKov5XO9ZuBY4jYFw6tXPgWYANSkPY8HxuLe8TkGt765X21h7wG24N6B\n2Ag0TpX1rz0Uve5N3k7U6/yqN+11OvAk8H+qOt8bdhRQpaqP+xGTyT1L3CZwvLP2EUAFUO49Kjo9\nd7wu82ZT7+Gkve48bA+ws9NjV/rrhrragm4XWkRmA/NUNWi34ps+sMRtzCAiItcAB6jql/yOxeRP\n4MoTjTFmqLPEbczgshK3Jo8ZxCxxGzO4PAJEReSKjgEicqSIWOfSg4glbmMGEXUvWn0YOFNEXhWR\nlcBNuA1QmUHCLk4aY0zA2Bm3McYEjCVuY4wJGEvcxhgTMJa4jTEmYCxxG2NMwFjiNsaYgLHEbYwx\nAWOJ2xhjAsYStzHGBIwlbmOMCRhL3MYYEzCWuI0xJmAscRtjTMBY4jbGmICxxG2MMQFjidsYYwLG\nErcxxgSMJW5jjAkYS9zGGBMwlriNMSZgLHEbY0zAWOI2xpiAscRtjDEBY4nbGGMCxhK3McYEjCVu\nY4wJmP8PPW4cHsxSFRUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10e286588>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "floor_repartition.plot(title='Repartition of passengers according to cabin floors',kind='pie')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Question 4 : Now for each travel class, we calculate the proportion of the passengers that survived.**\n",
    "The code is quite straightforward, we compute the number of survivor per class, then take the percentage per class. All the reste is simply cosmetic for it to look better"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAW8AAAD7CAYAAAClvBX1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XmYU9X5wPHvm0ySWRlWQUANbqi4A3VXXOqWWrVuFXet\n1Vqr1toarUv6c4t1ryitba1VUBF3jUtd2MQVFBBUFDEICLJnFmbP+f1x78DMkBlmSXJzh/fzPHkY\nkpt7TnJz33vuWcUYg1JKKXfxOJ0BpZRSHafBWymlXEiDt1JKuZAGb6WUciEN3kop5UIavJVSyoW6\ndfAWkYiIjHM6Hyq99Lh2T3pcO8b1wVtERovIDBGpEJFlIvK6iBzsUF6CIjJJRNaLyFcicpQT+egO\ncuy43iIin4tIvYhEnMhDd5Erx1VEthKRp0TkBxFJiMh0Edkv2/noClcHbxG5GrgfuB3oD2wLPAT8\n3KEsPQV8BvQB/gw8KyL9HMqLa+XgcV0A/AmIOZR+t5Bjx7UY+AQYDvQG/gvERKTYgbx0jjHGlQ+g\nFKgATmtjmwgwrsn/JwLLgQQwFRjW5LXjgS+AcmApcI39fF/gVWAdsAaYBnhSpLUzUAOUNHluKnCp\n09+Vmx65dlxbpDsOiDj9HbnxkcvHtck+y4DhTn9X7X24ueR9AJAPvNCB97wO7ARsBXwKjG/y2r+B\nS4wxJcDuwLv2838AlgD9sEoL1wOp5hQYBiw0xpQ3eW62/bxqv1w7rio9cvq4isjegB/rLssV8pzO\nQBf0AVYZY+rb+wZjzKONf9t1l2tFpNQYkwDqgN1EZLYxZi2w1t60Dtga2M4YswDrSp5KMVYJoaky\nYFB786eA3DuuKj1y9riKSA/gCeAv9r5dwc0l79VAXxFp1wVIRLwiEhWRb0WkDIjbL/W1/z0F61Zs\nkYhMEZED7Ofvwroa/09EFopIuJUkKoAeLZ4rxbqtU+2Xa8dVpUdOHlcRKQBeAT40xtzRsY/kLDcH\n7w+w6phPauf2o4ETgaOwgmrQfl4AjDGfGGNOxLpFexF4xn6+3BjzB2PM9lgNK1eLyJEp9j8P2F5E\nSpo8t5f9vGq/XDuuKj1y7riKSMB+7xLgkk58Jke5Nnjbtzc3AQ+JyEkiUigiPhE5TkT+muItJVg/\nntVAIVaLNwAi4heRs+xbsjqs6o6k/drPRGRHERGsapGGxtda5OdrYBZws4jki8gvgD2A59L5ubu7\nXDuu9rY+EcnHOl/y7OPrTd+n7v5y7biKiA94FqgCzjPGpDz2Oc3pFtOuPoCzgBlAJVbLdAw40LRo\nvcaqk34JqxpjEXAuVkPGjlgNFW9g1ZuVYXUhOth+3++xbtkqsa7QN7aRlyAwGesHMR84yunvx62P\nHDuuj9n7bPo43+nvyI2PXDmuwGH2/tZjVXk2Pg5x+jtq70PsD6KUUspFXFttopRSWzIN3kop5UIa\nvJVSyoU0eCullAtp8FZKKRfS4K2UUi6kwVsppVxIg7dSSrmQBm+llHIhDd5KKeVCGryVUsqFNHgr\npZQLafBWSikX0uCtlFIupMFbKaVcSIO3Ukq5kAZvpZRyIQ3eSinlQhq8lVLKhTR4K6WUC2nwVkop\nF9LgrZRSLqTBWymlXEiDt1JKuZAGb6WUciEN3kop5UIavJVSyoU0eCullAtp8FZKKRfS4K2UUi6k\nwVsppVxIg7dSSrmQBm+llHIhDd5KKeVCGryVUsqFNHgrpZQLafBWSikX0uCtlFIupMFbKaVcSIO3\nUkq5kAZvpZRyIQ3eSinlQhq8lVLKhTR4K6WUC2nwVkopF9LgrZRSLqTBWymlXEiDt1JKuZAGb6WU\nciEN3kop5UIavJVSyoU0eCullAtp8FZKKRfS4K2UUi6kwVsppVxIg7dSSrmQBm+llHIhDd5KKeVC\nGryVUsqFNHgrpZQLafBWSikX0uCtskJEfi4i4TTtqyId+1HKzcQY43QeVDchInnGmPospFNhjCnO\ndDpK5TJHSt5aCsttIlIkIjERmS0ic0XkDBGJi0hf+/URIjLZ/jsiIk+IyHTgCRH5UESGNdnXZHv7\n80VkjIiUisgiEfE0SWuxiPhEZAcReUNEZorINBHZxd5miIh8ICKfi8it2f9GlMo9GQveIpLX2mvG\nmJeNMdFMpa267FjgB2PMXsaY3YE3NrP9bsBRxpgzgQnA6QAisjWwtTFmRuOGxpgEMAs4zH7qZ8Cb\nxpg64BHgd8aY4cA1wMP2Ng8AY40xewDL0vEBlXK7zQZvLYVtkT4Hfioid4rIIXbAbcvLxpgq++9n\ngFPtv08Hnk2x/QTgDPvvXwITRKQYOBCYKCKzgH8AW9vbHAQ8Zf/9RIc/jVLdUHtK3loK28IYY74G\n9sUK4reKyE1APRt/L/kt3lLZ5L1LgdUisidWgJ6QIomXgWNFpDcwHHjX3vc6Y8zeTR67Ns1WGj6a\nUt1Ge4K3lsK2MCIyEFhvjBkH3IUVyONYgRbglM3sYgLwJ6DUGDOn5YvGmArgE6wL8avGmAZjTBnw\nnYicZudBRGQv+y3TsX4bAGd1+oMp1Y20Wi/dyBjztYjsCxyPVQp7hw6UwkSkaSns0hRJvAzc3qIU\nVoRdCmstW5vLd0cEwzE/4MX6TI3/1sWjoco239h97QHcJSJJoA74DVAA/FtEbgEmb+b9z2IF5lva\n2GYCMBEY1eS5s4CxInID4AOeBmYDVwJPisi1wEut7TAYjpUCg5s8BrDxN27sRz3WHdti4HtgcTwa\nqt7M51Eq52y2q6BdCltjjKkWkZ8BvwKKgXuMMa+LyH3APsaYUSISASqMMXc3ef9vgQPsbYbZz50P\njDDGXG7/fyJQDZQbYy6zn3sfuM8YM1FEBNjTGDNbRF4GnjHGjBOR3wB3tafbWDAc64lVah8KDAG2\ntx9BNr0ANVoPLAeWAAuBb4CPgA/i0dD6zaWpMsc+nodgHdORwD5Ar07u7kdgJtaxfR+YHo+Gqtp+\ni8qmYDiWD+wA9AQKsQoTTf+tBpbajyXxaGitQ1nNmvYE72Owbp03KYUBZVilsBFtBO/+WF/oLcaY\nv9jPnU/z4H0qdinMGDPFfm4IMBarusQHPG2M+T/7+SexLiAvAVelCt7BcKwI6+Q+wn7sQ/p619QD\nnwLT7MdbGswzLxiODcaqPjkd6y4tU72larGqaiYAE+PR0JoMpaNSCIZjfYCf2I/hWO1oQ+jY8a7E\nijsLsc7Rd4AZ8WioIb25dU63GqQTDMd8QAg41/7Xn6WkK7CqCv4LTIlHQ93nS3WYXcIejRW0DwYk\ny1moBd7EKjC8qFUsmREMxwZgXZTPwLpTz8RxTgBTsQL52/FoaF4G0siabhG8g+HYNsAlWFU6/R3O\nThyrIXVsPBrS3jCdFAzH+gNXY93plTicnUY/AvcCD8ejIR0c1kV2CfsUrAvzYWR/0OBs4J/AuHg0\ntLmOGDnH1cE7GI5tDUSAC2lH42uWVWF1b4zGo6FVTmfGLYLh2LZYPVUuovW2CKetwWqQ/Vs8Glrn\ndGbcJhiObQfcDJyNVSXqtPXA48A98WhogdOZaS9XBu9gONYD6wT/PVZjRS6rwDrR79YTvXV2j58w\ncD0QcDg77bUWuBb4Vy5XlYnIz4Hd0jGquSvzygTDsYHADVgX5mxVaXZEEngBCLshiLsqeAfDMcG6\njf4L0Nfh7HTUWuCaeDT0qNMZyTXBcOww4O/ALk7npZOmAb+OR0NfOZWBXJ4ULBiO9QOuwzp3c/Vu\nqqkaIIp115yzbRyuCd52FcljwNEOZ6Wr/gf8Kh4NLXY6I06z76AeAM53OCvpUIt1wt8aj4bqOrsT\nESnCGtw2GGvMwS3AnVi9s1aJyAjg7ia9u3bA6vL6PVaPjIuMMfPsfU3GGp28OzAC+DMwBxhijEna\naX1lv39b4CGgH1Y1wsXGmK/a27urNcFw7Cys6sMenf1OHLQA+F08GtrcqHJHuGI+72A4diLWj87t\ngRuszzAnGI6d7nRGnBQMx/YCZtA9AjdY1QA3AZPs6oHO6hbTUQTDsR7BcGwcMA53Bm6AHYHXg+HY\ns108phmR08E7GI4FguHY34EXcV81SVt6AhOC4dhDwXDM63Rmsi0Yjp0JfADs5HReMuAg4LNgOHZQ\nJ9/v+ukoguHY/lgXie4ylcEpwMxgOPYTpzPSVM4Gb/uW+g2sLoDd1WXAC8FwLNcbXdMmGI7dinUb\nXuB0XjJoK+DdYDh2fkff6OZJwYLhmATDsRuw2gCGtOc9LjIAmBIMx0Y7nZFGORm87Q77U2g+70V3\ndQIwORiObeV0RjItGI49iFXvuiXwA/8JhmOXd+RNbp0UzB4gNx6rjj7Xuu2mSz4wPhiO3WZ3nnBU\nzgXvYDi2A9YPprVJqbqjkcCHwXCsO1YjNJbIHgI6FMi6iQc7GMD3AD62qy9uBm7F6l31gIjMADY3\nvPtZrGD7TBvbTMDqY920ZH4WcJGIzAbmASfaz18J/FZEPgcGpdqZfef4MnDmZvLWXVwPPGdPweGY\nnOptEgzHdsYavur0KEmnLAIO6E4jM+0SyljaUf21/MkwNT/MRzxWM4C3pA+DLv4HFfMmsebNhzZu\naAymvoYB591PYMCOm+ynoaqc1a8/QHX8MzwFPeh12HkU7TYKgPqylax8MUr92qUU7XEUvY/41Yb3\n/fjMzfQ85GwCW2fkGnp5PBp6aPObuYsdwF4DDnU6Lw54GwjFo6FaJxLPmeBtdwV8H2uWv01UfjGF\nddOfoqF8Jd6iXvQ5/ioCA4ey6uW7qFm+gIayFfQ/83byt92z1TTqVi1m9VtjqV2+AG9hKb0Ov4DC\nnQ8EHD2pW5oNHBqPhsqykVimBcOxu4E/tGfb5U+GKRp2OCV7HdPmdhWfv03i/acZ+Ot/Yk042dzK\nl/8KxtDnuCuoXbGQFRP/woCz78LfbztW/+9h/P2CFO12GMseu5K+P7+WwNY7UfnlVKoXz6XP0Zd1\n6nO204XxaOg/mUwgm+wSd4wto3qzNU8BZzkxSCsnqk3sq3eMVgJ31XefsXbKY/Q9/iq2+f1E+o+O\nktdzAACBwcPoe8If8Ba1PRuoSTaw4vlbKNxhJNtc+RS9j72cVa/eQ92apQAkPpxI8R5HMujSf1P1\nzYfULPsGgMovp5LXs3+2AjfAXsCL9ohDVwuGY2fTzsDdERVz36Fo2BEpA3eytpr189+n5yFn4/EX\nkD94GIU77UflvEkA1K/7kfxt98QTKMI/YCfq1y0nWbOexIfP0uvQc9Od1Zb+bvfEcL1gOOYBnmPL\nDtxgVRXd40TCjgdv+7b6cawpW1NKvDee0gPPJDBoF0Q85JX0Ja+kL+L10WPkieQPHgbS9kepW72Y\nhoo1lIw8CfF4KdhuLwKDdqNy7ruA4yd1S4djzVDoWsFwbF+svsMdsm7Kf1n8t9EsH/dHqr/fpL2N\n+sQKahbPo2j3I1K+v37tUsTjxdd7Y/Wsf6sh1K1aZP3dbzuq4p+RrK6g9scF+Ppuy7ppT9BjxIl4\n8js16rsj/MCz9qRbbhfB6pOu4PfBcCzthZTNcTx4Y/U++EVrL5pkAzXLF5CsSrD0Hxez5KHzWPPW\nWJJ1NV1P2Rhqc+OkTuWXwXAso/fwmWIPh36BDnYH7DXqAgZd8i8GX/Zfivc6lhXP3ULd2ubV/xVz\n3yEweDd89p1XS8naKiTQPFnxF5KstbpC99j/NGqWfMHyJ8OU7BOCZD21K+MU7PgTVr58F8vHX0vZ\nzFc6ku2OGgRMtHtnuFIwHDsWa44StdFd9viFrHE0eAfDsX2wWtRb1VC5DpL1rJ8/nf5n3cnWF/yN\n2h8XkvggVRfW1vl6D8ZbWErZx89hGuqp+u5TqhfPxdgXgRw4qVO5KxiODc12omnwH6zh1h0SGDgU\nT6AQyfNRvMeRBAbtStXCGc22qZz3LsW7H9nqPjz+AkxN80VwkjXr8fitgO4tKKHfidcy8MIxlIz4\nOWve+ge9j7qEsg8n4u+3Hf1/eSsVs16nblVGZy84BLgtkwlkij3r4ziyP696rhPgn9nsMeZY8LZL\nHo+xmT6h4rMmmCvZ9wTyinvjLSylZORJVH07o623bbofbx79fnEDVd/OYMmYcyj7+AWKdjkYb4k1\ncDNHTuqWCoEnguGYa/rNBsOxM7AWwkgDgSYN6tVLvqChYg2FQ1sfvJjXaxAm2bChLQOgbuV3+Ppu\nt8m2FbPeIDBwKP5+QWpXLsI/YEfE68PXL0jtynh6PkLrrrYLL65ht8NMBPo4nZccVUQWz1cnS943\nAq13DbF584utANvkOi+dvOj7txrCgNFRtrnyKfqfcQv165YTGLjzJts5fFK3NBLru8p5wXCsF9bg\njw5LVldQtXAmpr4Wk2ygYt4kapbMpWD74Ru2qZz7DoU7H4gn0PqAVI8/n8KdD2Dde+NJ1lZTvWQe\n67/5iKJhhzfbrqFyHeWfxig92Bowl9ezP9XfzyFZW0Xt8m82NIhnkBd4xGXTI9yBtTSZat1+WFMb\nZ5wjJTq7xHFde7cv3uMoyme+SsGQ4eDNo2zGixTuMBIAU19H4+hd01CPqa8Fry9lT4TaFd/h6z0I\nY5KUf/oa9ZVrKd79qGbbNJ7UA86xluFsPKkDg3aldvk39Bh5cic/dZf8ORiOPRuPhj53IvEOuItO\n9tE3yQbWTRtH3ZolIB58vQfT7+QbNjQ8mvpaKr96j34nbfqzSXzwDNWL59H/9L8A0Pvoy1j9+gMs\nGXMWnvwe9DnmMvz9mpe81076N6UHnbmhOqV0/9NY+eIdVHz2OkV7HJWt3kUjsAYudeqCl03BcGwX\n4Irv7z212fOmvpaSfY6n908vbfb8uulPkXhvPFudcSsFwdTj7Vrr1w851XW3M24IhmPPxKOhrzOZ\niCP9vIPh2FSser92MQ31rHnnESq/mILk+Sja5RB6jboAyfOzZOyFNJStaLb9oEv/TV5p/01O6rWT\nHqVi9puYZAOBwcPo/dNL8PVqPlnYqlfvoWDH/Sja5WCg8Ud0B/Vr7B/RkRd38dN32pvxaChnW/eD\n4dgBWP30VcdUAEPj0dAPTmekLcFw7BWsWQg3SNZWsWTMOWx1WoT8bXbf8Hzd2mWsfP5WktXl9Ald\n3Wbwbq1fv8P98dNhUjwaSt0lKk2yHryD4diRWCOTVMf9NB4N5eR3FwzH3gEy+mPtxsbEo6HfOZ2J\n1gTDsSOwFu1tpuLzd0hMf5KBl/yr2Z3uj8/cRI/hJ7D6f2Ppc9wVnQrePz5zM72PvBhfn8GsfOlO\nCnc+kILth7P8yTADzrzdqR5gHXVKPBp6PlM7d6LO+y8OpNld3OJ0BlKxV8LRwN15FwfDsZTzhjjN\nHoyTchBKxdx3KNq9+WCpyq/eQ7w+Cuxqzc1prV9/Dnbd7YyM1n1nNXgHw7GfYs0PrDpnf7uPba5p\nd/uFSikAXO10JlpxLikmibMGS82lqEm3zWTNetZN/S+9j/p1u3bcVr/+HO2621Ej7ZqGjMh2yTuS\n5fS6o6yP5GqLvSJO25ORqPa4OBiOlTqdiabs0c8pezpVzHt3k8FS66Y/SdGwI8grbV+bdVv9+nO0\n625nZKz0nbXgHQzH9sRarUN1zRHBcGzTTsvOcawFt5spAc5xOhMtHIa1vuUmKue+S3GLKQqqF82m\nfOYrLB5zNovHnE1D+SpWvRQl8WGqBX1Sad6vv1GOdd3tqKOC4djwzW/WcdkseZ+dxbS6Mw9wntOZ\nALAHI2zRa3GmWa7Nh31hqierl3xJQ8VqCoce3Oz5/r+8jYEXPsTA8x9k4PkP4i3uTe9jLqdk303H\nbLWnXz/kRH/8dMhI6Tsrwdtu9MiZ5YO6gfNzYSUPrMWU+zmdiW7kgGA4FnQ6E7BhGcKUq/a0NljK\nW9ADb3GvDQ/Egye/eENf+sQHz/DjM9ZsGI39+hc/eBaL/zaa8pmvNuvX3yhVf/zqRXNY+vD5FOzw\nk1zu693UycFwLO2jUrM1SOdwWlmFQ3XKEKypOCc5nI/ussBsrhCsVXCiTmcEaw3MlENZ+xzbvoWB\nBv/m0Wb/Lz3gdEoPsP72Fpay9Xn3bXYffX/WvIknr0c/tj733naln0O8wPG0cwHn9spWtYme5Ol3\n6uY3yZxgOFbAxqWyVPrkStXJBU5noJs5Id07zFbwTtNERaqJwxxOfz+siXhUeu3p9GLUwXBsMHCA\nk3noho5J9zTAGQ/ewXBsV6Dbr4zugN3sebOdov31M8fp73aUw+l3Rz1Ic4ErGyXvLXFh0mwQnP1u\nD978JqqTnO5SO8rh9LurtFadZCN475eFNLZUo5xI1O49pLfVmeN08NYLc2akdbRlNoJ3+yY5UJ3h\n1GK2uwA5NRqwmxnu1ALUwXCsBNh0knuVDjvaBZ+0yGjwtiea3yWTaWzhhjiUbspRdyptAsA2DqW9\nN7rEWaYE6MTygK3JdMm7Pw4t+LCF6BMMx5zo8ZG2H6BqlVPBe7OrW6kuSdtdTaaD98DNb6K6KOhA\nmrk0t0p3NdihdF0x3tzFNHirDZwIpFryzjynSt66uHBmafBWGzgRSPW4Zp5TJe/eDqW7pdgxXTvK\nRp23yiwn6rwLHEhzS5NyXpEs0OCdWSXp2pETy6Cp9ErrkNt2cqQb2xbGieMKGrwzLW3nTqZ7gtRl\neP9pNULmf3m9b/wqp/PREctMn0oHpo5xTVey7eWHRff6xn6/oyzt63ReOqKMwrUOTQnkquBdSHVl\nIdVVTuejA2rStSMN3rYjPTNn/ct3z/Yi7Op0XjpmQcyBRBscSLND9pYF8+/xjV29vSzbT8R9vWOK\nqf7Q6Ty4wVjf/Z8c5p0zyul8dECPdE2yqsEbOM07+eO/5j2ypwj5TuelE9J2Je+AagfSbJdRnllz\nor5/1gyQtW4f2evEcQVYjXODvzqsQGrcVvWbtpi4xQfvS70vT7827+n9RFw7mMiJk3y5A2m26STP\nezMivv/6e0pldxlk4lTwdlW1YT61rqnCs1Wka0eZDlgrM7z/Lrkp7/EpF3jfOFTEPXW4KXzvQJpL\nHEhzE0IyeYH3zQ+vyXumT6HUjHA6P2m2zKF0VzuUbqfkU+t1Og8d9EO6dpTp4L0gw/vvtDG+v035\nmfdDpxc0SIf5DqS51IE0N/BRX3tF3vMfXeJ9dRu/1Ds9A1+mzHUoXVeVvAPUua3aJG0X5S0weBvz\nlO/WqQd4v+wOgbsO+M6BdB0J3oVUV16fN37Gmd5JO3sleYgTecgip4K3q0refqnbYkveGb1qxaOh\nMnKo6sRLQ/0b/vD0bhK4Ab4lknCi50c8m4n1pHztGN8Dk+cGLqo5O++dw7yS3Dqb6TugnEhikUNp\nO1Vd0yl+6t3WVrU4XTvKxgf/BnByuS4A/NTVvBv4w6zBsqo7TTTvRJUJwKdAPRn+/WzN6uV3+f4x\n/yDP3OEiW9TqLvMcTPsTB9PuMB8NTg1m6qzZ6dpRNuqLnPwhAlBEVfkHgcu/GCyrutuqPo4E73g0\ntB6Yk6n97yBLF73kv2Ha+4Hf9T7YO/cwEYozlVaOcqrKBOBzoNzB9DskjwY3lbwTWIXZtMhG8J6W\nhTRa1ZvE6o8Cv13cR8r3cTIfGZK2q3gnpH0Qyd6yYP67/j+8/7b/j4P38iw8RGSLHYbvWPCOR0NJ\n4GOn0u8oLw1u+o18SiRh0rWzbFy1JmchjZQGsXLZu4Fr1gekbjen8pBB9cDrDqb/AXBZOnbUjQbW\npMtMh9N/nzSvt5gpHoybgveMdO4s4yXveDS0GAdu73eSJfEpgd83BKRuh2ynnSXTiCTWOpj+FKBL\npYiTPdM+mRW4eM5j/r/uqYF7g6VYwdNJHzicfrt5SAaczkMHvJ3OnWWrj+RrWUoHgH3km/lv+q8t\nypOkU3MiZ8NLTiZuX5Q7XHUiJJMXel9//4vABfPv848d2Y1GRKbLBCKJpMN5+ABwOg/t4qKSdwKY\nlM4dZit4v5yldBjlmTXnef/NAzxiHO/hkmGOBm/b0+3d0Ed97TV5E6bND5y3+CbfEwcWSs3QTGbM\nxZ50OgPxaGgdDlZ3dpBb5pZ/nUgirdOFZKuldgrwLZDRKoyTPdM+udc3dncR1xzQzppDJBF3OhPA\nROA+2igEFFFVcX3e+Jm/9E4a6hXT3QfWdNXXRBJO13c3Gg8c4XQm2iIkkyL4a+oNl8Wqefu7etZU\nGXbo5eGOIwMct5OP+LokQx6ooKhJh8JrDwpw42Gpa1tGPVbJh0sayLN/0YN6eJh/udXZaXEiyWkT\nq/h6dQMX7O3nnmM2zmN33PhKbjk8nxEDWx0zlPbCVlaCdzwaMsFw7J9ANFNpXOR97f0b8saNFHFs\nEvtsesHpDADEo6FlwXBsKmzaB7sn5Wtv8z06+zjPR3t5hO4yKCrTnnI6A008BzwEuTvTZj51NUBB\nfRK2KRWmnF/EtqXCa9/Uc/qzVXz+m42BdF24hDxP+6YwGnN8Pr/ad9PamDveq+G8vXyM3qOQfR+p\n4Mw9fIwY6GXC3DqG9PS0FbgrgFc7/AE3I5vzAjwK1GZix9fljZ96Q964/beQwF0NjHU6E038s+l/\nBrJq2XjfbVM+C1ziD3k/GuURejmVMRdyvMqkUTwaSmDdWeWsALXVAEV+ITIqn2BPDx4RfrazjyE9\nPcz8Ib2Dj79bl+SIIV5K84WRA70sXJukrMYQnV7D7Ue2eY0bTySRttkEG2UteMejoZVkoMR4n++h\nyZfkxQ4V2WKWdPsXkcSPTmeiiWeAxTvI0kUv+/88bXrgij4HeecdJuLI2ppu9gGRxNdOZ6KFh53O\nQFsKqUk5be6PFUm+Xp1k2FYbQ8J291cw+N5yLnipilXr226Lve6dGvr+tZyDHq1kcrx+w/O79/Py\n1sIG1lUbZi5LMqyfhxvfreGq/fz0zG+zVJ+R7zHbAW9M+nZlzDjf7VNO9k4flb595rw64K9OZ6Kp\neDRU/6Tv1lve9v9x8J6e77bkgTVddb3TGWgpHg19iDUVQk7Kl9pNgnddg+Gs56s4by8fu/T10rdQ\n+OTiIhZdVczMXxdRXmO93po7jwqw8Ipill5dzK/39XHCU+v5do0V7K87JMC07+s57LFKLhvho7YB\n5qxo4IReN4ufAAARXUlEQVShPkY/t55D/1PJmI83qVx4h0giI6ORsxq849HQe6RhYImHZEPMf/30\ng71zO1WXevbzVQy4u5wed5Sx84MV/OvTjV/4M/Pq2PWhCkruKGO3hyp48avWG4jHfFzLiEcqCNxa\nxvkvNv9BLE4k2f9flfS+s4w/vNl84Znjxlcyo3O3dE8QSaRtYpt0OdD7xTgRVjidDxd7hUhistOZ\naMVNTmegNYVUNzs5k8ZwzgtV+L3CmOOtaoxivzBioJc8j9C/2MOY4/P537cNlNekHqKw3+A8SgJC\nIE84b28/B23j5bVvrNJ37wJhwqmFzL60mCv39/O716t58Lh8ou/VsPtWXt4+t5C/z6jly5XNzu2M\ntfM5UdXwR7qwBqKfuprJ/qs/GeZZ1OkJpsIH+1l4ZTFl1/Xg5TMLuOHdGmb+0MDSsiRnP1/FvUcH\nKAuXcNdPA4x+rooVlalvswaWCDccGuDCvTetam9s3PjuyhJenF+3IVi3o3GjNQ1k8IfQJZFEFXCH\nE0l/s7qB/FvLONsuTX2xsoERj1TQ684yet1ZxlGPV/LFytZ/bmuqDCdPWE/R7WVsd385T36+MR5k\n6ALcUj3wp3TsKBPi0VAMeMvpfKRSSM2GUpcxhoterubHSsNzpxfg86auxmh8NtnO4WUiYFKMRXtk\nZh37D/ay+1ZePl/RwIiBXvxeYY/+Hj5fsSFevEkkkdaBOU1lPXjHo6F5WI2XHVZEVcX7gd/N29az\nYv+u5GH3rbwU+qzDKFgH6Nu1SZaUJemZLxy3kw8RIbSzjyK/bLhtaukXu/o4aRcffQo3/aF0oXGj\nNROIJNI2qU0GPIIDq/r89rVqRg7aeCEcWOJhwqmFrPpjCav+WMLPh+bxy2dbv03+7WtV+L3w4zUl\njP9FAb+JVTFvhRWUM3ABTuVfRBJfpWNHGfQHcnDR6SKp3lAh/ZtYNV+uTPLKmYUU+Daejx8tqWf+\nqgaSxrB6fZIr3qhmVNA6L1taV214c0E91fWG+qRh/Jw6pi5q4Ngdm3fKW1GZ5KFPaomMsrobDunp\nYdJ39VTUGmb8kGT7Xh6wBjll9KLsVCPfTXRwLbeelK/9MHD5or5Stm86MnBZrIrC28rY5aFKti4W\njt8pjxEDvezaz8Mr8+toSBpe/KqOgBf27N/xk7QLjRuprAOu6XAmsimSqAEuootD5jvi6bl19MwX\njhyy8fj0zBd26O3B6xEM4BVhQSsX38paw3Nf1HPL4QGK/cLB2+Zx4lAfT8yxSt8ZuAC3VA7cnI4d\nZVI8Gvoc+LfT+WipkJo6gEXrkvxjZh2zljcw4O5yim8vo/j2MsbPqWPhWsOx49dTckc5u4+tJOAV\nnjpl4zCQ26fVcNz4SsCqL79hUjX97iqn71/LefDjWl48o5Cd+zQ//6/5Xw03HWr9ZgCuOzjAu/F6\ntrmvnBN2zmu8qD+eqbruRo5MpxiPhpYHw7EIcHd7tt+a1csnBa6uyJe6YenKw8OhAh48Lp8PljQw\nOd5AwAtej3Dunj7OfK6K6nrwe2HiaQUU+Tu+xOV1hwT4TayKf35a26xx4+ZRAUY/t54lZYbTh/m4\n/Cftat+7hkgi9yfJjyTeJlL6N+DKTCdVVmO4aVIN755X2KzNolHPaBkVtdbt8f8dnnpAxterk+R5\naHZy7tXfw+RFViGz8QLcv9jDzGVJbjy0SxfgVO4kknBLW8GNwJlAidMZaVQkVQ0A2/X0YG7u0ep2\nZ+7Reg/i6w8JANbvo1+Rh08u3vzsw4+f3HwM4DalHj76VbP3rSELDdBOdq+7l3bUpW0vPyyaGriq\nPl/qdkx3Brweq7S1pCzJ2Bm1vL2wnj+9XcPk84uovbGEKecX8atXqpm1vON3jJ1s3EjlbSKJnCv1\ntCFMFuZwv/HdGi7ax8fgHql/wuvCPUiESxhzfD77DEi9TUWtoUegeRDuEZANjVmd7F3QXrOBezr7\n5myLR0MryLG7hCKqc64qx3ZZNgpbjgXveDRkgHNpY5m0vWTB12/5/1jgk4aMTjBVn4Rv1xhmLW/g\n0O28jBjoxSPCyEFe9hvk5e2F9ZvfSRva2biRymrgvC4lnm2RRDVwFhkakAUwa3kDb39Xz+8PaPuu\npcgvXDrCx7kvVqdsdC72C2Uteh0kagwldkBP4wW4pXLgNPu7cpP7yY05dQAoojoXJ8+aQCQxIRsJ\nOTqwJR4NLQfOJ0U96SGeOZ+/6L+pv1fMVulMc0Vlkqfn1lFRa2hIWg0UT82t48jtvYwc6OW97xs2\nlLQ/W9bAtO8bWq3zrk8aqusNDUloMGxo6GiZXjsaN1pzEZFE2hYszZpIYjZwQ6Z2PzleT3xdkm3v\nq2DA3eXc/X4tz31Zx77/2LQZJWlgfZ1hadmmVfE79/FQn7R6rDSa/aPVPtFSFy7AqVyc443PKdkF\nrvPIkYXFiyTnSt4/kKY57tvD8VGJ8WjoNazJjTY4wfP+jMd90R1EKE13egKMnVHL4HvL6XVnOde8\nVc39x+Tz86E+DgvmcfNhAU59Zj0ld5RxyjPruf5gP0fvYDUNNG3cALh1ag0Ft5UTnV7LuDl1FNxW\nzq1Tm48baGfjRiq3E0nkTCmnE+4hQzPT/Xq4n2+vKGbWpUXMurSIS0f4Ce2Ux5tnF/LWt/V8tqyB\nhqShrMZw9Zs19MoXdk0RkIv8wi92zeOmyTVU1hre+76el+fXcc6ezetIu3gBbum+bJXMMsEeNn8K\nsN7pvBSSUzcuVcDJRBJrspWgGJO1zgGtCoZjXqyJcE68wPv6BzflPTF8Cx+p93ciid84nYkui5T2\nBt4F9spoMpOrWbDGMO4XBUycV8eNk2pYUpakwCf8ZJCXO44MbLh7un1aDdO+r+f1s6zR+2uqDBe+\nVMVbC+vpUyBEj8pndIsGrnNfqOKEnfM4bZj1/OJEklMnrufr1Uku2NvPvce0u+fJK8BJOTBfd5cF\nw7FzgMedzMPdvr9PPtU7dZSTebAZ4AwiiazOBZMTwRsgGI4VXJX37MNXep8/dwuapySVCcDo7nCC\nAxAp7YMVwLf0RRdmAwdnYoIipwTDsQeAK5xK/2Hf/VOO936cCzNW3kgkcWu2E82ZIBmPhqquynv+\nShHH1+9z0hvAOd0mcANEEqux1kN0ckV0p80EjulOgdt2FfA3pxIvoJUx7tn199YCt4hEROQaEfk/\nETmqvTsUkaCIbPZ8yZngDUAkUQYcDXzmdFYc8D5wSrpX28gJkcQqrAD+hdNZccAbwKgcmwkyLeLR\nkIlHQ1fi0LQNhVLjdPwaSzsaKI0xNxlj0j5M3ukPv6lIYh3WCh5vOJ2VLHoJOJZIwvFGoIyxBqMc\nAXzpdFay6DHghG5Y4m4mHg1dhwMTWOVTm5aRUp10F5HEZUQSzUr/IvJnEflaRN4DhtrPPSYip9p/\nDxeRKSIyU0TeFJGtmzw/W0RmA79tTwZyL3hDYwAPAbeTxeHWDjBYP/qTiSTKnc5Mxlmlz8OBd5zO\nShbcSiRxAZFE1wYJuEQ8GrqFLE/hkE+tE/GrAbiaSGKTeUtEZDjwS2Bv4HhgZIvXfcCDwKnGmOFY\nczzdZr/8H+B3xph2N+7nZvAGiCSSRBJ/xuqW1B0DWwKrVHZLy6t3t2YF8KOBa7HmJ+9uGoBLiCRu\ndDoj2RaPhu7BCl5ZOV8D1GU7fq0EfkokcV8rrx8CvGCMWW+MKWPThdeHArsDb4nILKyxEINFpCfQ\n0xgz1d7uifZkJneDd6NI4gVgPyDXVhnpii+AkUQSMacz4gjrwvxX4EDAdYNV2jAXOJRI4hGnM+KU\neDQ0AdgXmJXptAJSl825mT4BhhNJTOrCPgSYZ4zZ237sYYw5urM7y/3gDRBJfIl1CzIGa/5jt6rD\nGryynxtH2KVdJDED2AfrltHNqoDrgH2JJN53OjNOi0dDC4D9sVZ9yljPKR/1aZmTdzNqgf/D6ua5\nuYVQpgIniUiBiJQAJ7R4fT7QT0QOAKsaRUSGGWPWAetEpHGNgrPak7Gc6efdbpHSYVgjMn/qdFY6\nKIZVV9ad7iDSJ1J6Gla3swFOZ6WD3sSaiGih0xnJRcFw7ACsi/PQdO97duBXn5fK+j3Svd8mPgB+\nRSTR7l5SIvJnrCkEVmDNb/8pVlXJq8aYZ0Vkb6zfeSnWrK73G2P+adeXP4rVDvY/4HhjzO5tpuW6\n4N0oUvpzrFJs2mcbTLOvgN8TSWxJvWc6J1JaiDXo40+Q86vOL8c6rk87nZFcFwzH8rDmMLoR2DZd\n+50XuODLIqnZNV37a2IZVmn7kVwec+He4A0QKfUDlwO/A4LOZmYTi7CmvR3bLftuZ1KktBT4PVaX\nqb4O56alOVj9e8d19y6A6RYMx/xYC3b8GRjU1f3ND5z3bUDqduhyxjZaDtyJNfAmpyZOScXdwbtR\npFSAo4CLgRPBsXlRDPA28BDWorIbrtoiEsFaPagHMLW9nfZFJIh1y9XmLVS3FCktwLoFvRrYycGc\n1ADPAg9rnXbXBcOxAHAJ1nq2nZ7u+ZvAOYt90rBNGrL0NdYF+R/2eqyu0D2Cd1OR0n5Y84RfBGTi\nliqVL7G694wnkki5jmNj8DbGtGv1oCbvC7KlBu9G1sV5H6y+/8cDPyE7je0Lsdbm/Lc9SlSlUTAc\nE+Bg4HTgVDrY3vFt4KwVXZgyugJ4Hmsg1WQ3dtftfsG7qUjpTliDQkZh/UjScZVuwLp1no41pH16\nGwG7aePFYqw5Lpo2XgzHqlopBlYB5xtjljVpvACr8eK4LTp4t2RdoI/FCubHAD3TtOelwDRgEjBJ\newRlTzAc8wCHYgXyE4GBm3vPd4HRiQ5OG/0F1vn0P6yA7ZpSdirdO3i3FCndChiONcNdP6BPiodg\nrWCzyv638e8VWC3HH7WnrtMOwI9h9VHPs9/7d+zgjTUkfgpwojFmpYicARxjjLlQROYAlxtjporI\nXWjwbl2kNA+r0XrbFI9tsEpzVVgDRxofZU3+jmMdm09bm39Eq7yyLxiO9cc6T/dq8u+uwIb5er8L\njK4RIdUCpbVY4wfmNXl8RCSxJNP5zqYtK3hnkYhcBfQ2xtxk//9erJU2GoP3V1gl98YuZl6sVu7T\ngTnGmG3t9+0JPKkBwDla5ZUb7F4rvbG62ZXG80cXYAXzBmDthscW0pDsyOrxCtg42uqAZk9aQ2WV\nw1JVeYnIY3S8ykulSTwaqsc6HiusZxJOZsdx7hhh6U5ZHW2l0ifbEwwp1Rla8s4QY8ynIjIBawWV\nFVhzIzR9vdaeJvJvIrJhtBVW/dwFwKMi0jjaSmXXhgmGAESkrQmGwK7yamWCoeOyk2W1pdHgnUHG\nmNvYWCJL9fosrBb2ls/PpPm6j5tMP6kcpVVeynFabaLUprTKS+U8LXkr1YJWeSk30K6CSinlQlpt\nopRSLqTBWymlXEiDt1JKuZAGb6WUciEN3kop5UIavJVSyoU0eCullAtp8FZKKRfS4K2UUi6kwVsp\npVxIg7dSSrmQBm+llHIhDd5KKeVCGryVUsqFNHgrpZQLafBWSikX0uCtlFIupMFbKaVcSIO3Ukq5\nkAZvpZRyIQ3eSinlQhq8lVLKhTR4K6WUC2nwVkopF9LgrZRSLqTBWymlXEiDt1JKuZAGb6WUciEN\n3kop5UIavJVSyoU0eCullAtp8FZKKRfS4K2UUi6kwVsppVxIg7dSSrmQBm+llHIhDd5KKeVCGryV\nUsqFNHgrpZQLafBWSikX0uCtlFIupMFbKaVcSIO3Ukq5kAZvpZRyIQ3eSinlQhq8lVLKhTR4K6WU\nC2nwVkopF9LgrZRSLqTBWymlXEiDt1JKuZAGb6WUciEN3kop5UIavJVSyoU0eCullAtp8FZKKRfS\n4K2UUi6kwVsppVxIg7dSSrmQBm+llHIhDd5KKeVCGryVUsqFNHgrpZQLafBWSikX0uCtlFIupMFb\nKaVcSIO3Ukq50P8Dl7Pc5+DK9DMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10ea07cf8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class_repartition = []\n",
    "for c in [1,2,3]:\n",
    "    repartition = titanic.query('pclass==@c')['survived'].value_counts()\n",
    "    # We rename the indices \n",
    "    repartition.index = ['survived','died']\n",
    "    n = repartition.sum()\n",
    "    repartition = repartition.map(lambda x: x/n*100)\n",
    "    class_repartition.append(repartition.rename(\"Class {} survivors\".format(c)))\n",
    "    \n",
    "fig1, axarr = plt.subplots(1,3)\n",
    "for i,ax in enumerate(axarr):\n",
    "    ax.pie(class_repartition[i].values,labels=class_repartition[i].index,autopct='%1.1f%%',)\n",
    "    ax.axis('equal')  # Equal aspect ratio ensures that pie is drawn as a circle.\n",
    "    ax.set_title('Class {}'.format(i))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Question 5 : We now calculate the proportion of the passengers that survived by travel class and sex.**\n",
    "We use seaborn ```factorplot``` to do that very simply?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<seaborn.axisgrid.FacetGrid at 0x10ee7cfd0>"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAT4AAAEUCAYAAACyD1pgAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFetJREFUeJzt3X90VeWd7/H3hxCLRYowtlwCitRLiFiKTAS8iFfFdgSp\n0uHKLBQVrTqlV6irsxRcM9rm6u2d3qKX6Sgtd1YXY3W6Rtvljysd1PHa2o4WlaajIL+cjPyOtWZg\nEMvPJN/545y0xzTknITshPB8XmudRfY+z7P3N56sj89z9j7PUURgZpaSPj1dgJlZd3PwmVlyHHxm\nlhwHn5klx8FnZslx8JlZchx8ZpYcB5+ZJcfBZ2bJySz4JK2Q9GtJbx7leUn6a0l1ktZK+sPOnGfa\ntGkB+OGHH93/6LWyHPE9BExr5/npwKj840+B73TmJA0NDZ3pZmYJyyz4IuJnwO52mswEHo6cV4BT\nJQ3Nqh4zsxY9+R7fMGBHwfbO/L6iJNVICklRX1+fSXFmduLqlRc3IqImIhQRqqio6OlyzKyX6cng\n2wWcXrA9PL/PzCxTPRl8TwPX56/ung/sjYh3erAeM0tE36wOLOnvgYuB0yTtBL4GlANExHJgFXA5\nUAfsB27MqhYzs0KZBV9EXF3k+QBuzer8ZmZH0ysvbpiZHQsHn5klJ7Op7vGo+o6He+S8tUuu75Hz\nmlnbPOIzs+Q4+MwsOQ4+M0uOg8/MkuPgM7PkOPjMLDkOPjNLjoPPzJLj4DOz5Dj4zCw5Dj4zS46D\nz8yS4+Azs+Q4+MwsOQ4+M0uOg8/MkpPUQqQ9Zfs9Y7v9nGd8dV23n9Ost/CIz8yS4+Azs+Q4+Mws\nOQ4+M0uOg8/MkuPgM7PkOPjMLDkOPjNLjoPPzJLj4DOz5Dj4zCw5Dj4zS46Dz8yS4+Azs+Q4+Mws\nOZkGn6RpkjZLqpN0ZxvPD5S0UtIbktZLujHLeszMIMPgk1QGLAOmA2OAqyWNadXsVmBDRIwDLgbu\nl3RSVjWZmUG2I76JQF1EvB0Rh4FHgZmt2gQwQJKAU4DdQGOGNZmZZRp8w4AdBds78/sKPQicDdQD\n64DbIqK52IEl1UgKSVFfX99V9ZpZInr64sZlwOtABXAu8KCkjxXrFBE1EaGIUEVFRdY1mtkJJsvg\n2wWcXrA9PL+v0I3AE5FTB2wBqjKsycws0+BbA4ySNDJ/wWIO8HSrNtuBSwEkDQFGA29nWJOZWXZf\nLxkRjZIWAM8BZcCKiFgvaX7++eXAvcBDktYBAhZHRENWNZmZQcbfqxsRq4BVrfYtL/i5HvijLGsw\nM2utpy9umJl1OwefmSXHwWdmyXHwmVlyHHxmlhwHn5klx8FnZslx8JlZchx8ZpYcB5+ZJcfBZ2bJ\ncfCZWXIcfGaWHAefmSXHwWdmyXHwmVlyHHxmlhwHn5klx8FnZslx8JlZchx8ZpYcB5+ZJcfBZ2bJ\ncfCZWXIcfGaWHAefmSXHwWdmyXHwmVlyHHxmlhwHn5klx8FnZslx8JlZchx8ZpYcB5+ZJcfBZ2bJ\nyTT4JE2TtFlSnaQ7j9LmYkmvS1ov6adZ1mNmBtA3qwNLKgOWAZ8FdgJrJD0dERsK2pwKfBuYFhHb\nJX0iq3rMzFpkOeKbCNRFxNsRcRh4FJjZqs01wBMRsR0gIn6dYT1mZkCREZ+ky9t7PiJWtfP0MGBH\nwfZOYFKrNpVAuaQXgQHAtyLi4fbOaWZ2rIpNde/I/9sPmACsy2+PBV4D2gu+Us9fDVwKnAyslvRK\nRLzVXidJNcDXAIYOHXqMJZhZatqd6kbEJRFxCbAVuCAixkfEeGAysKXIsXcBpxdsD8/vK7QTeC4i\nfhMRDcDPgHHFio6ImohQRKiioqJYczOzDyn1Pb5PRcSrLRsR8Rq5UV971gCjJI2UdBIwB3i6VZv/\nB0yR1FfSR8lNhTeWWJOZWaeUGny/kXRty4akucD+9jpERCOwAHiOXJj9ICLWS5ovaX6+zUbgWWAt\nuanzdyPizY7/GmZmpSv1dpYbgUckfRcIcu/1zSvWKX/xY1WrfctbbS8BlpRYh5kdZ2praweRuzh5\nPNpXXV29p/XOkoIvPzI7T9KA/Pa+Li7OzHqhbdu2LaisrLyuvLz85J6upS1Hjhw5sG3btkdGjBjx\nYOH+koJPkoAvAKMi4k5JZwIVEfHzLq/UzHqF2traQZWVldcNGDCgDDjc0/W0pV+/fmXAdbW1td8v\nHPmV+h7f/yF3y8nn89v7gL/q2hLNrJcZcLyO9Arla/zQVLzU4LsEmAscAIiIfyN3b5+ZWa9TavAd\njIho2ZDUB1A2JZlZaoYMGfLp7jxfqcG3Ln8Li/Lv730H+KesijIzy1KpwfdnwMXAUODVfL872utg\nZmnbvHnzSZWVlWOuvPLKkePHj6+65pprRjQ3N/OVr3ylYty4cVWTJk2q/OY3v/nxwj7PPPPMKeed\nd97oCRMmjJ48eXLlrl27+gLceuutw8aNG1d1/vnnV959991DmpubmTVr1pnV1dWjJ02aVLls2bLB\nHamt1NtZ9gG35B9mZiV55513TvrpT3/61tChQxs/97nPffKxxx4buHbt2pN/+ctfbiorK6OxsfFD\n7SdPnrz/F7/4xWaAr3/965+47777PrF06dL6lStXDnrrrbfW9+vXL5qamnjvvffKNm3adPK6des2\nlpWV0dTU1KG6Sr2d5V+BFcD3ImJnh85gZskaMWLEoaFDhzYCTJ48ed+bb7558tSpU/eVlZUB0Lfv\nhyPo9ddf73fXXXcNO3jwYJ/333+/bOzYsfsBlixZsmPu3Lkjmpqa9MUvfvG96dOnf7BgwYJ3r7rq\nqjP79OnD4sWL3504ceKBUusqdap7JTAIeFXS85KukeSrumbWrm3btn3kV7/6VRnA6tWrTznnnHMO\nvPjiiwNaRmitR3z33HPP0EWLFv1qzZo1m6+99tqGiKC5uZkZM2a8//jjj29dsmTJroULF444dOiQ\nbrrppj1PPvnk1ptvvrlh0aJFwzpSV6lT3fXA7ZIWA9OBm4EHgQ7Nq80sLcOGDTv8pS996fStW7f2\nGz169IE5c+bsXb169Snnnnvu2f3792+aNWvWnkWLFr3X0n7OnDm7b7vttjO+9a1vHRwyZMgRgCNH\njmjq1KmVEcHBgwf7zJ8//936+vq+s2fP/mSfPn04ePCgFi9e/E5H6uro0vNV5C5yTABqO9jXzBLT\nv3//pieffHJr4b4HHnhgF62WqHv33XfXAtxyyy17brnllt/7bO1rr722ufW+lvcCO6Okqa6kL0uq\nBZ4AdgPnR8RnO3tSM7OeVOqIbyzw5Yh4OctizOzEMXr06MO1tbWdHpVlqdT3+Hwbi5mdMIp92dAj\nEXGdpDXk1uH7kIiYmFllZmYZKTbia1mB5fasCzEz6y7tBl9EtFy5bY4IfzbXzE4IJa/HJ+lfJN0l\n6fTizc3Musfs2bPPHDx48LhRo0adU2qfUi9uTJA0ltz3bLwiaT3wtxHx952s1TJ2wQMX9Mh5X17o\nC/8pq77j4equPF7tkuuL3i/8hS98oeG222779Y033jiy1OOWOuIjItZFxO3AJ8l9p+7fldrXzCwr\n06dP/+DjH/94Y/GWv1Ny8En6lKT7gH8FRgLXd7A+M7PjQqmrs/wS6A88DPyXiNiRaVVmZhkqGnz5\nZeYX+lMbZnaiKDrVjYhmckvNm5mdEEp9j68u/10bZmbHlSuuuGLklClTqrZs2fKRIUOGfHrp0qWn\nFetT6iIFA4C1kl4CPmjZGRF/0slazewEVMrtJ11t5cqVWzrap9Tg+zt8+4qZnSBKvYH5e1kXYmbW\nXUq9neWHtL06i6e6ZtbrlDrV/VHBz/2Aq4ANXV+OmVn2OjXVlfS3wD9mUpGZWcZK/shaKwF06Ovc\nzMyOF515j68PMA74/1kVZWZWirq6uvK5c+eObGhoKJfEvHnz3rv77rt/XaxfZ97jOwLcFxGvdKpS\nMzthbb9nbJcuS3XGV9e1e19geXk5999//84pU6bs37NnT5/x48ePufzyy9+vrq4+2F6/Uqe604Gn\ngB8ANcAqSUWXo5c0TdJmSXWS7myn3QRJjZKuKrEeMzNGjBhxZMqUKfsBBg0a1HzWWWcd2L59+0nF\n+pUafKMjYi8wA/gxMJwiy1JJKgOWkQvNMcDVksYcpd3/xhdLzOwYbN68+aQNGzZ89KKLLvqgWNtS\ng688/+9FwKqI2A80F+kzEaiLiLcj4jDwKDCzjXYLgceBovNyM7O27N27t8+sWbPO+sY3vrFj8ODB\nxbKp5ODbIOkZ4ArgBUknl9BnGFC4bt9OWl0JljQM+GO8+ouZddKhQ4c0Y8aMs2bPnr173rx5/15K\nn1KDbx7wf4FLIuI3wGDgqO/ZdcBfAYvzS1+VTFKNpJAU9fX1XVCGmfVGzc3NzJkzZ0RlZeXBmpqa\nd0vtV1LwRcSBiHgqIrbkt3dFxLNFuu0CCr+RbXh+X6HzgEclbSX3aZBvS/p8CfXURIQiQhUVFaX8\nCmZ2Anr++edPeeqpp/7gpZdeGlBVVTWmqqpqzGOPPTawWL9Sb2fpjDXAKEkjyQXeHOCawgYR8dtv\nRZL0EPCjiHgqw5rMLEPFbj/papdddtkHBd//XbLMgi8iGiUtAJ4DyoAVEbFe0vz888uzOreZWXuy\nHPEREauAVa32tRl4EXFDlrWYmbXo7Gd1zcx6LQefmSXHwWdmyXHwmVlyMr24YWaWpf3792vSpElV\nhw8fVlNTk6644oo9S5cuLfqpBgefmXWZCx64oEuXpXp54cvt3qPXr1+/eOmllzYPHDiw+dChQ5ow\nYcLoF154Ye+ll176m/b6eaprZr1Wnz59GDhwYDPA4cOH1djYKEnF+2VemZlZhhobG6mqqhozZMiQ\ncRdddNH7U6dObXe0B57qWi9SfcfD3X7O2iXtLjtpx4G+ffuyadOmDQ0NDWUzZsw4a82aNf0mTJjQ\nJSswm5kd10477bSmCy+8cN/KlSuLLlLg4DOzXqu+vr5vQ0NDGcAHH3ygn/zkJx87++yz2x3tgae6\nZtaL7dixo/yGG24Y2dTURERo5syZu6+++uq9xfo5+MysyxS7/aSrTZo06cDGjRs3dLSfp7pmlhwH\nn5klx8FnZslx8JlZZ+07cuTIgZ4uoph8jfsK9/nihpl1SnV19Z5t27Y9AlxXXl5eylfOdrsjR44c\n2L179yPV1dV7Cvc7+Mys00aMGPFgbW3t94EBPV3LUexrHXrg4DOzY5QPlt8Ll+OZ3+Mzs+Q4+Mws\nOQ4+M0uOg8/MkuPgM7PkOPjMLDkOPjNLjoPPzJLj4DOz5Dj4zCw5Dj4zS46Dz8yS4+Azs+Q4+Mws\nOQ4+M0uOg8/MkpNp8EmaJmmzpDpJd7bx/FxJayWtk/RzSeOyrMfMDDIMPkllwDJgOjAGuFrSmFbN\ntgAXRcRY4F7gb7Kqx8ysRZYjvolAXUS8HRGHgUeBmYUNIuLnEdGyZPUrwPAM6zEzA7INvmHAjoLt\nnfl9R3MT8EwpB5ZUIykkRX19/TGUaGYpOi4ubki6hFzwLS6lfUTURIQiQhUVFdkWZ2YnnCy/ZW0X\ncHrB9vD8vg+R9Gngu8D0iPi3DOsxMwOyHfGtAUZJGinpJGAO8HRhA0lnAE8A10XEWxnWYmb2W5mN\n+CKiUdIC4DmgDFgREeslzc8/vxz4KvAHwLclATRGxHlZ1WRmBhl/oXhErAJWtdq3vODnm4Gbs6zB\nzKy14+LihplZd3LwmVlyHHxmlhwHn5klx8FnZsnJ9KqumXXOBQ9c0CPnfXnhyz1y3u7m4DNrx/Z7\nxvbMiQd9rGfOmwhPdc0sOQ4+M0uOg8/MkuPgM7PkOPjMLDkOPjNLjoPPzJLj4DOz5Dj4zCw5Dj4z\nS46Dz8yS4+Azs+Q4+MwsOQ4+M0uOg8/MkuPgM7PkOPjMLDkOPjNLjoPPzJLj4DOz5Dj4zCw5Dj4z\nS46Dz8yS4+Azs+Q4+MwsOQ4+M0uOg8/MkuPgM7PkZBp8kqZJ2iypTtKdbTwvSX+df36tpD/Msh4z\nM8gw+CSVAcuA6cAY4GpJY1o1mw6Myj/+FPhOVvWYmbXIcsQ3EaiLiLcj4jDwKDCzVZuZwMOR8wpw\nqqShGdZkZkbfDI89DNhRsL0TmFRCm2HAO+0dWFIN8LX85n5JG4+p0oyNOLbuFUB9lxTSDfRl9XQJ\nXSql1w46/Po9GxHTsqolS1kGX2Yiogao6eEyuoWkiIiKnq7DOs6v3fEry6nuLuD0gu3h+X0dbWNm\n1qWyDL41wChJIyWdBMwBnm7V5mng+vzV3fOBvRHR7jTXzOxYZTbVjYhGSQuA54AyYEVErJc0P//8\ncmAVcDlQB+wHbsyqnl7sf/R0AdZpfu2OU4qInq7BzKxb+ZMbZpYcB5+ZJcfBZ2bJcfCZWXIcfGaW\nHAdfxiR9WdJGSd/P6Pg1km7P4tjWtSRdLOlHPV2H9dKPrPUy/x34TETs7OlCzCzHI74MSVoOfBJ4\nRtJfSFoh6TVJ/yxpZr7NDZKekvS8pK2SFkj6s3ybVyQNzre7RdIaSW9IelzSR9s431mSnpVUK+mf\nJFV172984pN0pqRNkh6S9Jak70v6jKSXJf2LpIn5x+r8a/hzSaPbOE7/tv4erHs4+DIUEfPJrc5x\nCdAf+HFETMxvL5HUP9/0U8AsYALwdWB/RIwHVgPX59s8ERETImIcsBG4qY1T/g2wMCKqgduBb2fz\nmyXvPwP3A1X5xzXAFHL/zf8c2ARcmH8Nvwr8rzaO8Rcc/e/BMuapbvf5I+DKgvfj+gFn5H/+SUTs\nA/ZJ2guszO9fB3w6//OnJP1P4FTgFHIfBfwtSacAk4EfSr9dWugjWfwixpaIWAcgaT3wQkSEpHXA\nmcBA4HuSRgEBlLdxjKP9PRzXS6ydKBx83UfAf4uIzR/aKU0CDhXsai7YbuZ3r9FDwOcj4g1JNwAX\ntzp+H+DfI+Lcri3b2lDs9bqX3P/M/ljSmcCLbRyjzb8H6x6e6naf54CFyg/HJI3vYP8BwDuSyoG5\nrZ+MiPeBLZJm548vSeOOsWbrnIH8bnm1G47S5lj/HuwYOPi6z73kpjxr89OjezvY/27gVeBlcu8h\ntWUucJOkN4D1/P5S/9Y9vgn8paR/5uizqmP9e7Bj4NVZzCw5HvGZWXIcfGaWHAefmSXHwWdmyXHw\nmVlyHHxmlhwHn5klx8FnmZD0UUk/lLQhv6LMD/L750l6Nb+CzI9bVi6RdJekJwr6rpN0eU/+Dnbi\n8md1LSuXAR+LiDEAkgZJuhD4E+C/RsQhSdOBFcAF5FYweVbSQmA88ExErOqh2u0E5+CzrLwBnC1p\nGbkP6f8DcAUwDni15SOqwCCAiGiWdC3wOrCd3DJPZpnwVNcyERFvA+cAzwOfIReEAlZExLn5x7iI\nOKOg20hyK5ycCpzc3TVbOvxZXcuEpOHA7ojYn18tuh64EngYmBIROyWVAedGRK2kQcAacquZfBYY\nHRFzeqh8O8F5xGdZGQuszq8U8xrwlxHxM3IrDz+d3/8mv1tBZgW50eBLwD3Af5I0vwfqtgR4xGdm\nyfGIz8yS4+Azs+Q4+MwsOQ4+M0uOg8/MkuPgM7PkOPjMLDkOPjNLzn8AxrABilyCmQMAAAAASUVO\nRK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10ee831d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.factorplot(x='sex',y='survived',hue='pclass',kind='bar',data=titanic,ci=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Question 6 : We want to create 2 equally populated age categories and calculate survival proportions by age category, travel class and sex. **\n",
    "First we separate the ages into two equally populated categories that we will name 'young' and 'old' for the sake of simplicity. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pclass</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>survived</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>female</td>\n",
       "      <td>old</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>male</td>\n",
       "      <td>young</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>female</td>\n",
       "      <td>young</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>male</td>\n",
       "      <td>old</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>female</td>\n",
       "      <td>young</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  pclass     sex    age  survived\n",
       "0      1  female    old         1\n",
       "1      1    male  young         1\n",
       "2      1  female  young         0\n",
       "3      1    male    old         0\n",
       "4      1  female  young         0"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = titanic[['pclass','sex','age','survived']]\n",
    "df['age'] = pd.qcut(titanic.age,2,labels=['young','old'])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we group those according to the the categories that interwst us, compute the sum (will sume the 1s corresponding to survivors and then compute the totals thanks to the ```count``` aggregation function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>survived</th>\n",
       "      <th>totals</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pclass</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">female</th>\n",
       "      <th>old</th>\n",
       "      <td>89</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>39</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">male</th>\n",
       "      <th>old</th>\n",
       "      <td>36</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>17</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">2</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">female</th>\n",
       "      <th>old</th>\n",
       "      <td>41</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>51</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">male</th>\n",
       "      <th>old</th>\n",
       "      <td>7</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>16</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">female</th>\n",
       "      <th>old</th>\n",
       "      <td>16</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>56</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">male</th>\n",
       "      <th>old</th>\n",
       "      <td>18</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>41</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     survived  totals\n",
       "pclass sex    age                    \n",
       "1      female old          89      92\n",
       "              young        39      41\n",
       "       male   old          36     119\n",
       "              young        17      32\n",
       "2      female old          41      47\n",
       "              young        51      56\n",
       "       male   old           7      84\n",
       "              young        16      74\n",
       "3      female old          16      43\n",
       "              young        56     109\n",
       "       male   old          18     125\n",
       "              young        41     224"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped = df.groupby(['pclass','sex','age']).agg({'survived':'sum'})\n",
    "grouped['totals'] = df.groupby(['pclass','sex','age']).agg({'survived':'count'})\n",
    "grouped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the number of survivors as well as the count of the population of each group, we can very simply compute the proportion of survivors.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>survived (%)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pclass</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">female</th>\n",
       "      <th>old</th>\n",
       "      <td>96.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>95.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">male</th>\n",
       "      <th>old</th>\n",
       "      <td>30.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>53.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">2</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">female</th>\n",
       "      <th>old</th>\n",
       "      <td>87.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>91.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">male</th>\n",
       "      <th>old</th>\n",
       "      <td>8.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>21.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">female</th>\n",
       "      <th>old</th>\n",
       "      <td>37.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>51.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">male</th>\n",
       "      <th>old</th>\n",
       "      <td>14.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>young</th>\n",
       "      <td>18.30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     survived (%)\n",
       "pclass sex    age                \n",
       "1      female old           96.74\n",
       "              young         95.12\n",
       "       male   old           30.25\n",
       "              young         53.12\n",
       "2      female old           87.23\n",
       "              young         91.07\n",
       "       male   old            8.33\n",
       "              young         21.62\n",
       "3      female old           37.21\n",
       "              young         51.38\n",
       "       male   old           14.40\n",
       "              young         18.30"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped['survived (%)'] = grouped['survived']/grouped['totals']*100\n",
    "grouped[['survived (%)']].round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": false,
   "skip_h1_title": false,
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
